{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build Excel stoich for GAMS (originally A01_build_excel_stoich_for_GAMS.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "import cobra,json,os,sys,openpyxl,re,pandas as pd\n",
    "from collections import OrderedDict\n",
    "from copy import deepcopy\n",
    "\n",
    "model_root_path = '../'\n",
    "sys.path.append(model_root_path)\n",
    "from pycore.gsm_custom_functions import *\n",
    "from pycore.cobrapy_custom_extras import *\n",
    "kribo_default = 22*3600 #(22 aa/ribo/s converted to aa/ribo/h)\n",
    "mito_compartments = ['m', 'mm'] # IDs for mitochondria compartments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reactions to ignore when checking if enzymes must be added:\n",
      "['ABTD2Dy_c', 'ABTD4Dy_c', 'ABTLD_c', 'ACALDt_c_m', 'ACOAO100_x', 'ACOAO120_x', 'ACOAO40_x', 'ACOAO60_x', 'ACOAO80_x', 'ALCD2i2_c', 'ASNtps_v', 'ASPtps_v', 'BIOMASS_RBA', 'BIOMASS_MFA', 'BIOMASS_MFA_NO_GAM', 'BIOMASS_PinheiroEtAl2020', 'BIOMASS_xyl_hybrid', 'CTPS1_c', 'CYSItps_v', 'DDPA_m', 'EX_fald_e', 'FER_c', 'G3PD1i_m', 'GLNtps_v', 'GLUtps_v', 'GLYAT_c', 'HMGCOAS_m', 'ILEtps_v', 'LEUtps_v', 'NITRy_c', 'PHCHGS_m', 'PKETF_c', 'TYRtps_v', 'XU5PFGT_x', 'XYLI1_c', 'XYLI2_c', 'XYLK_c', 'XYLURx_c', 'XYLURy_c', 'BIOMASS']\n"
     ]
    }
   ],
   "source": [
    "# Metabolic model (COBRApy json)\n",
    "# If you have to use multiple stoichiometric models (SMs), \n",
    "#   it's best to keep 1 as the main copy and then use transaction files to apply differences to the others. Helps improve consistency in annotations, stoich. etc.\n",
    "chosen_model = 'Batch-Rabinowitz' # c: carbon, n: nitrogen\n",
    "\n",
    "# List parameters for your model version(s), such as:\n",
    "\t# max_allowed_mito_proteome_allo_fraction: maximum fraction of proteome that can be allocated to mitochondria\n",
    "\t# nonmodeled_proteome_allocation: fraction of proteome not accounted for in the model\n",
    "\t# predicted_but_unused_proteins_path: path to file with predicted proteins that are not used in the model\n",
    "\t# sheet_name: name of sheet in relevant excel files (e.g., biomass, dummy protein) to use\n",
    "# for 'Batch-Rabinowitz' before including nonmodeled proteins into model: 'nonmodeled_proteome_allocation':0.49,\n",
    "modelVersions = {\n",
    "\t'Batch-Rabinowitz':{\n",
    "\t\t'max_allowed_mito_proteome_allo_fraction':0.13953,\n",
    "\t\t'nonmodeled_proteome_allocation':0,\n",
    "\t\t'predicted_but_unused_proteins_path':'%model_root_path%prosyn_constraints_predicted_but_unused_Rabinowitz2023_batchGlc_using_MFA_and_scRBA_methods.txt',\n",
    "\t\t'sheet_name':'batch-Rabinowitz'\n",
    "\t},\n",
    "\t'C':{\n",
    "\t\t'max_allowed_mito_proteome_allo_fraction':0.13464,\n",
    "\t\t'nonmodeled_proteome_allocation':0.54,\n",
    "\t\t'sheet_name':'chemo-clim-Rabinowitz'\n",
    "\t},\n",
    "\t'N':{\n",
    "\t\t'max_allowed_mito_proteome_allo_fraction':0.13033,\n",
    "\t\t'nonmodeled_proteome_allocation':0.53,\n",
    "\t\t'sheet_name':'chemo-nlim-Rabinowitz'\n",
    "\t}\n",
    "\t}\n",
    "modelname = 'iRhto'+chosen_model\n",
    "model = load_cobra_model('./input/'+modelname+'.json')\n",
    "# if not choosing between nutrient-limited models, uncomment the following lines:\n",
    "# model = cobra.io.load_json_model('./input/iRhtoC.json')\n",
    "# chosen_model = ''\n",
    "\n",
    "# Update GAMS settings file to reflect model version, \n",
    "#   and add NGAM, mito capacity, etc.\n",
    "#   If not using multiple models, replace everything in the curly braces with the values you want\n",
    "with open('./model/model-version-settings.txt', 'w') as f:\n",
    "\tnpa = modelVersions[chosen_model].get('nonmodeled_proteome_allocation', 0)\n",
    "\tmampaf = modelVersions[chosen_model]['max_allowed_mito_proteome_allo_fraction'] if 'max_allowed_mito_proteome_allo_fraction' in modelVersions[chosen_model] else 1\n",
    "\tf.write(f\"* file with settings updated by build_model scripts, to help with managing multiple SMs\\n\"\n",
    "\t\tf\"$setGlobal max_allowed_mito_proteome_allo_fraction {mampaf}\\n\"\n",
    "\t\tf\"$setGlobal nonmodeled_proteome_allocation {npa}\\n\")\n",
    "\tif 'measured_but_unused_proteins_path' in modelVersions[chosen_model]:\n",
    "\t\tf.write('* unless specified otherwise (e.g., in test_kapp.gms), include prosyn constraints to reflect production of unused proteins; comment out if using dummy protein instead\\n')\n",
    "\t\tif modelVersions[chosen_model]['predicted_but_unused_proteins_path']:\n",
    "\t\t\tif os.path.exists(modelVersions[chosen_model]['predicted_but_unused_proteins_path']):\n",
    "\t\t\t\tf.write(f\"$if not %ignore_measured_unused_constraints%==1 $include {modelVersions[chosen_model]['predicted_but_unused_proteins_path']}\\n\")\n",
    "\n",
    "# Protein\n",
    "df_pro = read_spreadsheet('./input/PROTEIN_stoich_curation.xlsx')\n",
    "## OPTIONAL: add \"special cases only\" sheet to df_pro. \n",
    "##   Good for distinguishing proteins in the organism from ones added for testing purposes.\n",
    "include_special_case_proteins = False # affects protein and enzyme stoichiometry\n",
    "if include_special_case_proteins:\n",
    "\tdf_pro_special = read_spreadsheet('./input/PROTEIN_stoich_curation.xlsx',sheet_name='special cases only',header=1)\n",
    "\tdf_pro = pd.concat([df_pro,df_pro_special],ignore_index=True)\n",
    "\n",
    "# Amino acid info\n",
    "df_aamap = pd.read_csv('./input/PROTEIN_amino_acid_map.txt', sep='\\t')\n",
    "df_aamap.index = df_aamap.aa_abbv.to_list()\n",
    "\n",
    "# Dummy protein\n",
    "dummy_path = './input/PROTEIN_dummy_prot_calc.xlsx'\n",
    "# df_aa_dummy = read_spreadsheet(dummy_path)\n",
    "# df_aa_dummy2 = read_spreadsheet(dummy_path,sheet_name='unidentified')\n",
    "# 1st entry is default dummy protein; remove 2nd entry unless needed\n",
    "dummy_list = [{\"df\":read_spreadsheet(dummy_path,sheet_name=modelVersions[chosen_model]['sheet_name']),\"name\":\"PROSYN-PROTDUMMY\"},\n",
    "\t\t\t  {\"df\":read_spreadsheet(dummy_path,sheet_name='unidentified'),\"name\":\"PROSYN-PROTDUMMYUNIDENTIFIED\"}]\n",
    "for prot in dummy_list:\n",
    "\tprot[\"df\"].index = prot[\"df\"].aa_abbv.to_list()\n",
    "\tprot[\"medianL\"] = int(round(prot[\"df\"].loc['A', 'Unnamed: 5'], 0))\n",
    "\t# Add 1e-5 g/mmol to prevent number round lost of protein MW\n",
    "\t#dummy_MW\n",
    "\tprot[\"MW\"] = round(prot[\"df\"].loc['C', 'Unnamed: 5'], 5) + 1e-5\n",
    "\tprot[\"biosyn\"] = prot[\"name\"].replace('PROSYN-','BIOSYN-')\n",
    "\tprot[\"met\"] = prot[\"name\"].lower().replace('prosyn-','BIO-')\n",
    "\tprot[\"translation_loc\"] = 'unknown' if not 'translation_loc' in prot else prot['translation_loc']\n",
    "\n",
    "# Enzyme\n",
    "df_enz = read_spreadsheet('./input/ENZYME_stoich_curation.tsv')\n",
    "# add 'id' column to df_enz, which is 'RXN-' + rxn_src + '_' + dir + '-' + enz\n",
    "if include_special_case_proteins:\n",
    "\tdf_enz_special = read_spreadsheet('./input/ENZYME_stoich_curation_special_cases.tsv')\n",
    "\tdf_enz = pd.concat([df_enz,df_enz_special],ignore_index=True)\n",
    "\n",
    "# RNA\n",
    "df_rnas = read_spreadsheet('./input/RNA_stoich.xlsx')\n",
    "df_rnas.index = df_rnas.RNAid.to_list()\n",
    "\n",
    "# Ribosome ('_c' for nucleus or nucleoid, '_m' for mitochondria)\n",
    "ribo_dict = {'_c':read_spreadsheet('./input/RIBOSOME_nucleus.xlsx'), \n",
    "\t\t\t '_m':read_spreadsheet('./input/RIBOSOME_mitochondria.xlsx')}\n",
    "\n",
    "# Biomass\n",
    "# heading start at cell X4\n",
    "df_biom = read_spreadsheet('./input/BIOMASS_RBA.xlsx', sheet_name=modelVersions[chosen_model]['sheet_name'], header=4)\n",
    "# concat with the other biomass sheet if needed (WARNING: should find a way to avoid having elements be redefined)\n",
    "# df_biom = pd.concat([df_biom,read_spreadsheet('./input/BIOMASS_RBA.xlsx', sheet_name='batch-Rabinowitz+Pinheiro-macro', header=4)],ignore_index=True)\n",
    "# df_biom = read_spreadsheet('./input/BIOMASS_RBA.xlsx', sheet_name='batch-Pinheiro', header=4)\n",
    "# list of biomass rxns to turn off by default and only turn on when needed; any rxns not found will be excluded from list\n",
    "whole_biomass_rxns = ['BIOSYN-' + i for i in ['BIODILAERO', 'BIODILAERO-NOGAM','BIOXYL-PINHEIRO','BIOXYL-PINHEIRO-NOGAM']] + ['BIOMASS_xyl_hybrid']\n",
    "\n",
    "# checks which rxns to ignore when seeing if all necessary enzymes are modeled. Ignore blocked rxns (to avoid unnecessary work from adding them)\n",
    "rxns_not_needing_enzymes = list(dict.fromkeys([r.id for r in model.reactions if r.bounds == (0,0)] + [rxn.id for rxn in find_biomass_reactions(model)] + ['BIOMASS_MFA','BIOMASS_MFA_NO_GAM','BIOMASS_RBA']))\n",
    "# Gene-Protein Reaction associations (GPRs) representing spontaneous reactions (i.e., no enzyme needed in model)\n",
    "spont_GPRs = ['SPONT', 'UNKNOWN']\n",
    "# take IDs of all rxns\n",
    "print('Reactions to ignore when checking if enzymes must be added:')\n",
    "print(rxns_not_needing_enzymes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Expanding df_pro to include sublocations as separate proteins\n",
    "## remove the 'id' column\n",
    "df_pro = df_pro.drop(columns=['id'])\n",
    "# if the 'gene_src' column is empty, fill it with the 'name' column\n",
    "df_pro['gene_src'] = df_pro['gene_src'].fillna(df_pro['name'])\n",
    "## for each protein, if there are commas in subloc_assigned, split them into separate entries with 'id' modified to reflect sublocation\n",
    "df_pro['subloc_assigned'] = df_pro['subloc_assigned'].apply(lambda x: x.split(',') if type(x) == str and ',' in x else x)\n",
    "# add the 'id' column back in, assigning it to \"gene_src\" plus \"subloc_assigned\" if the latter is not 'unknown'\n",
    "# modify IDs to reflect sublocation\n",
    "# add location only if there are multiple proteins with the same gene source\n",
    "df_pro = df_pro.explode('subloc_assigned').reset_index(drop=True)\n",
    "if 'id' not in df_pro.columns:\n",
    "\tdf_pro['id'] = df_pro.apply(\n",
    "\t\tlambda row: row['id_manual_override (clear when using new sheet)']\n",
    "\t\tif pd.notna(row['id_manual_override (clear when using new sheet)'])\n",
    "\t\telse (\n",
    "\t\t\tstr(row['gene_src'])\n",
    "\t\t\t+ (\n",
    "\t\t\t\t'_' + row['subloc_assigned']\n",
    "\t\t\t\tif row['subloc_assigned'] != 'unknown'\n",
    "\t\t\t\tand df_pro[df_pro['gene_src'] == row['gene_src']].shape[0] > 1\n",
    "\t\t\t\telse ''\n",
    "\t\t\t)\n",
    "\t\t),\n",
    "\t\taxis=1\n",
    "\t)\n",
    "df_pro.index = df_pro.id.to_list()\n",
    "# save df_pro to model folder\n",
    "df_pro[['id'] + [col for col in df_pro.columns if col != 'id']].to_csv('./model/PROTEIN_stoich_curation.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enzyme stoichiometry has empty \"dir\" values. Creating FWD and REV entries for those enzymes.\n"
     ]
    }
   ],
   "source": [
    "# calculate MW of enzymes = sum of MW of all proteins in protein_stoich (MW from df_pro) * number of protein_stoich\n",
    "\n",
    "# if there are any rows in df_enz with 'dir' that's empty, create 2 new rows for each of those, one with 'dir' as 'FWD' and one with 'dir' as 'REV' (for faster writing to model)\n",
    "if df_enz['dir'].isnull().any():\n",
    "\tprint('Enzyme stoichiometry has empty \"dir\" values. Creating FWD and REV entries for those enzymes.')\n",
    "\tdf_enz_fwd = df_enz.copy()\n",
    "\tdf_enz_fwd['dir'] = df_enz_fwd['dir'].fillna('FWD')\n",
    "\tdf_enz_rev = df_enz.copy()\n",
    "\tdf_enz_rev['dir'] = df_enz_rev['dir'].fillna('REV')\n",
    "\tdf_enz = pd.concat([df_enz_fwd[df_enz_fwd['dir'] == 'FWD'], df_enz_rev[df_enz_rev['dir'] == 'REV']], ignore_index=True)\n",
    "\t# sort df_enz by 'rxn_src' and 'dir'\n",
    "\tdf_enz = df_enz.sort_values(by=['rxn_src', 'dir']).reset_index(drop=True)\n",
    "\t# Now df_enz_new contains all original rows, plus FWD and REV for those missing 'dir'\n",
    "\n",
    "df_enz['id'] = df_enz.apply(lambda x: 'RXN-' + x['rxn_src'] + '_' + x['dir'] + '-' + x['enz'], axis=1)\n",
    "df_enz.index = df_enz.id.to_list()\n",
    "mw = 'MW (g/mmol)'\n",
    "for enz in df_enz.index:\n",
    "\tenz_mw = 0\n",
    "\tfor pro in df_enz.loc[enz, 'protein_stoich'].split(','):\n",
    "\t\t# find protein stoichiometry in df_pro\n",
    "\t\tdelim = ':'\n",
    "\t\tif delim in pro:\n",
    "\t\t\tprot_id = pro.split(delim)[0].strip()\n",
    "\t\t\tprot_stoich = float(pro.split(delim)[1].strip())\n",
    "\t\t\tpro_matches = df_pro.loc[df_pro.gene_src == prot_id]\n",
    "\t\t\t# if no matches, check if prot_id is in df_pro.index\n",
    "\t\t\tif len(pro_matches) == 0:\n",
    "\t\t\t\tif prot_id in df_pro.index:\n",
    "\t\t\t\t\tpro_matches = df_pro.loc[prot_id]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tprint(f'Warning: {prot_id} not found in protein stoichiometry file')\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\tpro_mw = pro_matches[mw]\n",
    "\t\t\tif isinstance(pro_matches[mw], pd.Series):\n",
    "\t\t\t\tif pro_matches[mw].nunique() > 1:\n",
    "\t\t\t\t\tprint(f'Warning: {prot_id} has multiple MWs in df_pro')\n",
    "\t\t\t\tpro_mw = pro_matches[mw].iloc[0]\n",
    "\t\t\tif float(pro_mw) == 0:\n",
    "\t\t\t\tprint(f'Warning: MW for {prot_id} is 0 in protein stoichiometry file')\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tenz_mw += pro_mw * prot_stoich\n",
    "\t# set enz_mw to NaN if it is 0\n",
    "\tif enz_mw == 0:\n",
    "\t\tenz_mw = float('nan')\n",
    "\tdf_enz.loc[enz, mw] = enz_mw\n",
    "# move 'id' column to the front\n",
    "df_enz[['id'] + [col for col in df_enz.columns if col != 'id']].to_csv('./model/ENZYME_stoich_curation.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list all genes, to map them to translation (PROSYN) rxns and other rxns making their products where needed\n",
    "# combine all genes from df_pro and model\n",
    "# combine all genes from df_pro and model, ensuring all are strings and not NaN\n",
    "genes_model = [str(gene.id) for gene in model.genes if pd.notna(gene.id)]\n",
    "genes_pro = [str(g) for g in df_pro.gene_src.to_list() if pd.notna(g)]\n",
    "genes_rna = [str(r) for r in df_rnas.RNAid.to_list() if pd.notna(r)]\n",
    "genes = sorted(list(dict.fromkeys(genes_model + genes_pro + genes_rna)))\n",
    "gene_expression_dict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "# testing only\n",
    "# # tag_c formula weight\n",
    "# # print(model.metabolites.get_by_id('tag_c').formula_weight-227.000000) # 227.000000 from Ac element accidentally listed in it\n",
    "# # TAG MW is 89.07 w/o Acyl3, 882.40 w/ Acyl3 according to https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6838544/. Thus, Acyl MW is (882.40-89.07)/3 = 264.4433333333\n",
    "# # calculate_molecular_weight('C28H43O')\n",
    "# # (12.010700*28)+(1.007940*43)+15.999400 = 395.64042\n",
    "# # print(model.metabolites.get_by_id('ergstest_c').formula_weight-227.000000+264.4433333333) # 227.000000 from Ac element accidentally listed in it\n",
    "# print(model.metabolites.get_by_id('o2_c').formula_weight)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DM_5mta_c\n",
      "5mta_c --> \n",
      "5mta_c\tC11H15N5O3S1\t0\tC00170\t5'-S-methyl-5'-thioadenosine\tbiocyc\n",
      "C:-11, H:-15, N:-5, O:-3, S:-1, charge:0\n",
      "\n",
      "DM_gcald_c\n",
      "gcald_c --> \n",
      "gcald_c\tC2H4O2\t0\tC00266\tglycolaldehyde\tbiocyc\n",
      "C:-2, H:-4, O:-2, charge:0\n",
      "\n",
      "TORULNOXi_c\n",
      "o2_c + toruln_c --> 2.0 h_c + torularho_c\n",
      "h_c\tH1\t1\tC00080\tH+\tbiocyc\n",
      "o2_c\tO2\t0\tC00007\toxygen\tbiocyc\n",
      "torularho_c\tC40H52O2\t0\t      \tTorularhodin\tmanual\n",
      "toruln_c\tC40H54\t0\tC08613\tTorulene\tseed.compound\n",
      "H:0, O:0, C:0, charge:2\n",
      "\n",
      "compACYLCOA_c\n",
      "0.01875 docoscoa_c + 0.355664 linocoa_c + 0.087612 linolncoa_c + 0.207129 odecoa_c + 0.1694 pmtcoa_c + 0.112806 stcoa_c + 0.048639 ttccoa_c <=> acylcoa_c\n",
      "acylcoa_c\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\tmanual\n",
      "docoscoa_c\tC43H74N7O17P3S1\t-4\tC16528\tdocosanoyl-CoA\tbiocyc\n",
      "linocoa_c\tC39H62N7O17P3S\t-4\tC02050\tlinoleoyl-CoA\tseed.compound\n",
      "linolncoa_c\tC39H60N7O17P3S\t-4\tC16162\talpha-linolenoyl-CoA\tseed.compound\n",
      "odecoa_c\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA\tbiocyc\n",
      "pmtcoa_c\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA\tbiocyc\n",
      "stcoa_c\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA\tbiocyc\n",
      "ttccoa_c\tC45H78N7O17P3S\t-4\tC16529\ttetracosanoyl-CoA\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693481999999996, N:3.3306690738754696e-16, O:-0.999999999999999, P:5.551115123125783e-17, S:-5.551115123125783e-17, Acyl:1, charge:2.220446049250313e-16\n",
      "\n",
      "compACYLCOA_l\n",
      "0.01875 docoscoa_l + 0.355664 linocoa_l + 0.087612 linolncoa_l + 0.207129 odecoa_l + 0.1694 pmtcoa_l + 0.112806 stcoa_l + 0.048639 ttccoa_l <=> acylcoa_l\n",
      "acylcoa_l\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\tmanual\n",
      "docoscoa_l\tC43H74N7O17P3S1\t-4\tC16528\tdocosanoyl-CoA\tbiocyc\n",
      "linocoa_l\tC39H62N7O17P3S\t-4\tC02050\tlinoleoyl-CoA\tseed.compound\n",
      "linolncoa_l\tC39H60N7O17P3S\t-4\tC16162\talpha-linolenoyl-CoA\tseed.compound\n",
      "odecoa_l\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA\tbiocyc\n",
      "pmtcoa_l\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA\tbiocyc\n",
      "stcoa_l\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA\tbiocyc\n",
      "ttccoa_l\tC45H78N7O17P3S\t-4\tC16529\ttetracosanoyl-CoA\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693481999999996, N:3.3306690738754696e-16, O:-0.999999999999999, P:5.551115123125783e-17, S:-5.551115123125783e-17, Acyl:1, charge:2.220446049250313e-16\n",
      "\n",
      "compACYLCOA_m\n",
      "0.01875 docoscoa_m + 0.355664 linocoa_m + 0.087612 linolncoa_m + 0.207129 odecoa_m + 0.1694 pmtcoa_m + 0.112806 stcoa_m + 0.048639 ttccoa_m <=> acylcoa_m\n",
      "acylcoa_m\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\tmanual\n",
      "docoscoa_m\tC43H74N7O17P3S1\t-4\tC16528\tdocosanoyl-CoA\tbiocyc\n",
      "linocoa_m\tC39H62N7O17P3S\t-4\tC02050\tlinoleoyl-CoA\tseed.compound\n",
      "linolncoa_m\tC39H60N7O17P3S\t-4\tC16162\talpha-linolenoyl-CoA\tseed.compound\n",
      "odecoa_m\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA\tbiocyc\n",
      "pmtcoa_m\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA\tbiocyc\n",
      "stcoa_m\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA\tbiocyc\n",
      "ttccoa_m\tC45H78N7O17P3S\t-4\tC16529\ttetracosanoyl-CoA\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693481999999996, N:3.3306690738754696e-16, O:-0.999999999999999, P:5.551115123125783e-17, S:-5.551115123125783e-17, Acyl:1, charge:2.220446049250313e-16\n",
      "\n",
      "compACYLCOA_r\n",
      "0.01875 docoscoa_r + 0.355664 linocoa_r + 0.087612 linolncoa_r + 0.207129 odecoa_r + 0.1694 pmtcoa_r + 0.112806 stcoa_r + 0.048639 ttccoa_r <=> acylcoa_r\n",
      "acylcoa_r\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\tmanual\n",
      "docoscoa_r\tC43H74N7O17P3S1\t-4\tC16528\tdocosanoyl-CoA\tbiocyc\n",
      "linocoa_r\tC39H62N7O17P3S\t-4\tC02050\tlinoleoyl-CoA\tseed.compound\n",
      "linolncoa_r\tC39H60N7O17P3S\t-4\tC16162\talpha-linolenoyl-CoA\tseed.compound\n",
      "odecoa_r\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA\tbiocyc\n",
      "pmtcoa_r\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA\tbiocyc\n",
      "stcoa_r\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA\tbiocyc\n",
      "ttccoa_r\tC45H78N7O17P3S\t-4\tC16529\ttetracosanoyl-CoA\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693481999999996, N:3.3306690738754696e-16, O:-0.999999999999999, P:5.551115123125783e-17, S:-5.551115123125783e-17, Acyl:1, charge:2.220446049250313e-16\n",
      "\n",
      "compACYLCOA_rm\n",
      "0.01875 docoscoa_rm + 0.355664 linocoa_rm + 0.087612 linolncoa_rm + 0.207129 odecoa_rm + 0.1694 pmtcoa_rm + 0.112806 stcoa_rm + 0.048639 ttccoa_rm <=> acylcoa_rm\n",
      "acylcoa_rm\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\tmanual\n",
      "docoscoa_rm\tC43H74N7O17P3S1\t-4\tC16528\tdocosanoyl-CoA\tbiocyc\n",
      "linocoa_rm\tC39H62N7O17P3S\t-4\tC02050\tlinoleoyl-CoA\tseed.compound\n",
      "linolncoa_rm\tC39H60N7O17P3S\t-4\tC16162\talpha-linolenoyl-CoA\tseed.compound\n",
      "odecoa_rm\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA\tbiocyc\n",
      "pmtcoa_rm\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA\tbiocyc\n",
      "stcoa_rm\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA\tbiocyc\n",
      "ttccoa_rm\tC45H78N7O17P3S\t-4\tC16529\ttetracosanoyl-CoA\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693481999999996, N:3.3306690738754696e-16, O:-0.999999999999999, P:5.551115123125783e-17, S:-5.551115123125783e-17, Acyl:1, charge:2.220446049250313e-16\n",
      "\n",
      "compACYLCOA_x\n",
      "0.01875 docoscoa_x + 0.355664 linocoa_x + 0.087612 linolncoa_x + 0.207129 odecoa_x + 0.1694 pmtcoa_x + 0.112806 stcoa_x + 0.048639 ttccoa_x <=> acylcoa_x\n",
      "acylcoa_x\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\tmanual\n",
      "docoscoa_x\tC43H74N7O17P3S1\t-4\tC16528\tdocosanoyl-CoA\tbiocyc\n",
      "linocoa_x\tC39H62N7O17P3S\t-4\tC02050\tlinoleoyl-CoA\tseed.compound\n",
      "linolncoa_x\tC39H60N7O17P3S\t-4\tC16162\talpha-linolenoyl-CoA\tseed.compound\n",
      "odecoa_x\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA\tbiocyc\n",
      "pmtcoa_x\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA\tbiocyc\n",
      "stcoa_x\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA\tbiocyc\n",
      "ttccoa_x\tC45H78N7O17P3S\t-4\tC16529\ttetracosanoyl-CoA\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693481999999996, N:3.3306690738754696e-16, O:-0.999999999999999, P:5.551115123125783e-17, S:-5.551115123125783e-17, Acyl:1, charge:2.220446049250313e-16\n",
      "\n",
      "compCER_g\n",
      "0.25 cer1_24_g + 0.25 cer1_26_g + 0.25 cer2A_24_g + 0.25 cer2A_26_g <=> cer_g\n",
      "cer1_24_g\tC42H85N1O3\t0\t      \tceramide-1 (C24)\tbiocyc\n",
      "cer1_26_g\tC44H89N1O3\t0\t      \tceramide-1 (C26)\tbiocyc\n",
      "cer2A_24_g\tC42H85N1O4\t0\t      \tceramide-2 (C24)\tbiocyc\n",
      "cer2A_26_g\tC44H89N1O4\t0\t      \tceramide-2 (C26)\tbiocyc\n",
      "cer_g\tCer\t0\tC00195\tCeramide\tmanual\n",
      "C:-43, H:-87, N:-1, O:-3.5, Cer:1, charge:0\n",
      "\n",
      "compCER_m\n",
      "0.25 cer1_24_m + 0.25 cer1_26_m + 0.25 cer2A_24_m + 0.25 cer2A_26_m <=> cer_m\n",
      "cer1_24_m\tC42H85N1O3\t0\t      \tceramide-1 (C24)\tbiocyc\n",
      "cer1_26_m\tC44H89N1O3\t0\t      \tceramide-1 (C26)\tbiocyc\n",
      "cer2A_24_m\tC42H85N1O4\t0\t      \tceramide-2 (C24)\tbiocyc\n",
      "cer2A_26_m\tC44H89N1O4\t0\t      \tceramide-2 (C26)\tbiocyc\n",
      "cer_m\tCer\t0\tC00195\tCeramide\tmanual\n",
      "C:-43, H:-87, N:-1, O:-3.5, Cer:1, charge:0\n",
      "\n",
      "compCER_r\n",
      "0.25 cer1_24_r + 0.25 cer1_26_r + 0.25 cer2A_24_r + 0.25 cer2A_26_r <=> cer_r\n",
      "cer1_24_r\tC42H85N1O3\t0\t      \tceramide-1 (C24)\tbiocyc\n",
      "cer1_26_r\tC44H89N1O3\t0\t      \tceramide-1 (C26)\tbiocyc\n",
      "cer2A_24_r\tC42H85N1O4\t0\t      \tceramide-2 (C24)\tbiocyc\n",
      "cer2A_26_r\tC44H89N1O4\t0\t      \tceramide-2 (C26)\tbiocyc\n",
      "cer_r\tCer\t0\tC00195\tCeramide\tmanual\n",
      "C:-43, H:-87, N:-1, O:-3.5, Cer:1, charge:0\n",
      "\n",
      "compFALPD_c\n",
      "0.01875 docosa_c + 0.1694 hdca_c + 0.355664 linoea_c + 0.087612 linolen_c + 0.112806 ocdca_c + 0.207129 ocdcea_c + 0.048639 ttcosa_c <=> falpd_c\n",
      "docosa_c\tC22H43O2\t-1\tC08281\tdocosanoate\tseed.compound\n",
      "falpd_c\tAcylO\t-1\tC00162\tFatty acid (from lipid)\tmanual\n",
      "hdca_c\tC16H31O2\t-1\tC00249\tpalmitate\tbiocyc\n",
      "linoea_c\tC18H31O2\t-1\tC01595\tlinoleate\tseed.compound\n",
      "linolen_c\tC18H29O2\t-1\tC06427\talpha-linolenate\tseed.compound\n",
      "ocdca_c\tC18H35O2\t-1\tC01530\tstearate\tbiocyc\n",
      "ocdcea_c\tC18H33O2\t-1\tC00712\toleate\tbiocyc\n",
      "ttcosa_c\tC24H47O2\t-1\tC08320\ttetracosanoate\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693482, O:-0.9999999999999999, Acyl:1, charge:6.938893903907228e-17\n",
      "\n",
      "compFALPD_en\n",
      "0.01875 docosa_en + 0.1694 hdca_en + 0.355664 linoea_en + 0.087612 linolen_en + 0.112806 ocdca_en + 0.207129 ocdcea_en + 0.048639 ttcosa_en <=> falpd_en\n",
      "docosa_en\tC22H43O2\t-1\tC08281\tdocosanoate\tseed.compound\n",
      "falpd_en\tAcylO\t-1\tC00162\tFatty acid (from lipid)\tmanual\n",
      "hdca_en\tC16H31O2\t-1\tC00249\tpalmitate\tbiocyc\n",
      "linoea_en\tC18H31O2\t-1\tC01595\tlinoleate\tseed.compound\n",
      "linolen_en\tC18H29O2\t-1\tC06427\talpha-linolenate\tseed.compound\n",
      "ocdca_en\tC18H35O2\t-1\tC01530\tstearate\tbiocyc\n",
      "ocdcea_en\tC18H33O2\t-1\tC00712\toleate\tbiocyc\n",
      "ttcosa_en\tC24H47O2\t-1\tC08320\ttetracosanoate\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693482, O:-0.9999999999999999, Acyl:1, charge:6.938893903907228e-17\n",
      "\n",
      "compFALPD_m\n",
      "0.01875 docosa_m + 0.1694 hdca_m + 0.355664 linoea_m + 0.087612 linolen_m + 0.112806 ocdca_m + 0.207129 ocdcea_m + 0.048639 ttcosa_m <=> falpd_m\n",
      "docosa_m\tC22H43O2\t-1\tC08281\tdocosanoate\tseed.compound\n",
      "falpd_m\tAcylO\t-1\tC00162\tFatty acid (from lipid)\tmanual\n",
      "hdca_m\tC16H31O2\t-1\tC00249\tpalmitate\tbiocyc\n",
      "linoea_m\tC18H31O2\t-1\tC01595\tlinoleate\tseed.compound\n",
      "linolen_m\tC18H29O2\t-1\tC06427\talpha-linolenate\tseed.compound\n",
      "ocdca_m\tC18H35O2\t-1\tC01530\tstearate\tbiocyc\n",
      "ocdcea_m\tC18H33O2\t-1\tC00712\toleate\tbiocyc\n",
      "ttcosa_m\tC24H47O2\t-1\tC08320\ttetracosanoate\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693482, O:-0.9999999999999999, Acyl:1, charge:6.938893903907228e-17\n",
      "\n",
      "compFALPD_mm\n",
      "0.01875 docosa_mm + 0.1694 hdca_mm + 0.355664 linoea_mm + 0.087612 linolen_mm + 0.112806 ocdca_mm + 0.207129 ocdcea_mm + 0.048639 ttcosa_mm <=> falpd_mm\n",
      "docosa_mm\tC22H43O2\t-1\tC08281\tdocosanoate\tseed.compound\n",
      "falpd_mm\tAcylO\t-1\tC00162\tFatty acid (from lipid)\tmanual\n",
      "hdca_mm\tC16H31O2\t-1\tC00249\tpalmitate\tbiocyc\n",
      "linoea_mm\tC18H31O2\t-1\tC01595\tlinoleate\tseed.compound\n",
      "linolen_mm\tC18H29O2\t-1\tC06427\talpha-linolenate\tseed.compound\n",
      "ocdca_mm\tC18H35O2\t-1\tC01530\tstearate\tbiocyc\n",
      "ocdcea_mm\tC18H33O2\t-1\tC00712\toleate\tbiocyc\n",
      "ttcosa_mm\tC24H47O2\t-1\tC08320\ttetracosanoate\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693482, O:-0.9999999999999999, Acyl:1, charge:6.938893903907228e-17\n",
      "\n",
      "compFALPD_rm\n",
      "0.01875 docosa_rm + 0.1694 hdca_rm + 0.355664 linoea_rm + 0.087612 linolen_rm + 0.112806 ocdca_rm + 0.207129 ocdcea_rm + 0.048639 ttcosa_rm <=> falpd_rm\n",
      "docosa_rm\tC22H43O2\t-1\tC08281\tdocosanoate\tseed.compound\n",
      "falpd_rm\tAcylO\t-1\tC00162\tFatty acid (from lipid)\tmanual\n",
      "hdca_rm\tC16H31O2\t-1\tC00249\tpalmitate\tbiocyc\n",
      "linoea_rm\tC18H31O2\t-1\tC01595\tlinoleate\tseed.compound\n",
      "linolen_rm\tC18H29O2\t-1\tC06427\talpha-linolenate\tseed.compound\n",
      "ocdca_rm\tC18H35O2\t-1\tC01530\tstearate\tbiocyc\n",
      "ocdcea_rm\tC18H33O2\t-1\tC00712\toleate\tbiocyc\n",
      "ttcosa_rm\tC24H47O2\t-1\tC08320\ttetracosanoate\tseed.compound\n",
      "C:-18.028033999999998, H:-32.693482, O:-0.9999999999999999, Acyl:1, charge:6.938893903907228e-17\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Make version of model with no generic cofactors, for publication and for MEMOTE testing\n",
    "model_no_generics = deepcopy(model)\n",
    "# remove rxns with names beginning with \"GENERIC_\"\n",
    "for rxn in model.reactions:\n",
    "\tif rxn.id.startswith(\"GENERIC_\"):\n",
    "\t\t# print(rxn.id)\n",
    "\t\tmodel_no_generics.reactions.get_by_id(rxn.id).remove_from_model()\n",
    "report_mass_balance(model_no_generics)\n",
    "# remove metabolites with names beginning with \"pseudometabolite \"\n",
    "for met in model.metabolites:\n",
    "\tif met.name.startswith(\"pseudometabolite \"):\n",
    "\t\t# print(met.name)\n",
    "\t\tmodel_no_generics.metabolites.get_by_id(met.id).remove_from_model()\n",
    "# write JSON version of the model w/o these additions (for future use)\n",
    "cobra.io.save_json_model(model_no_generics, \"./input/\"+modelname+\"-no-generic-cofactors.json\")\n",
    "# convert to xml file\n",
    "cobra.io.write_sbml_model(model_no_generics, \"./input/\"+modelname+\"-no-generic-cofactors.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assemble reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'rt04': True, 'rt012': False, 'rt05': True, 'rt01': False, 'rt0123': False}, {'rt04': False, 'rt012': True, 'rt05': True, 'rt01': True, 'rt0123': False}, {'rt04': False, 'rt012': False, 'rt05': True, 'rt01': True, 'rt0123': True}]\n"
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "import cobra.manipulation\n",
    "\n",
    "def find_true_combinations(expression, use_knockout=False, variables:list=[]):\n",
    "    \"\"\"\n",
    "    Finds all combinations of boolean values that make the given expression true.\n",
    "\n",
    "    Args:\n",
    "        expression: The boolean expression as a string.\n",
    "        use_knockout: If True, the function will test combinations where only one variable is False, and use those to eliminate redundant combinations, saving time. Will return an error if the expression contains 'not' operators unless set to 'ignore warning'.\n",
    "        variables: A list of strings (variable names in expression) to consider. If none provided, the function will extract variable names from the expression. Added in case consistent RegEx parsing proves too unreliable.\n",
    "    Returns:\n",
    "        A list of dictionaries, where each dictionary represents a true combination.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Extract variables from the expression\n",
    "    if variables == []:\n",
    "        variables = set(re.findall(r'\\b\\w+\\b', expression))\n",
    "        if use_knockout and any([var in ['not','!'] for var in variables]):\n",
    "            raise ValueError(\"The expression contains 'not' operators, which may cause issues with the 'use_knockout' option. If you still want to use this option, set it to 'ignore warning'.\")\n",
    "        # remove 'and', 'or', 'not' from variables\n",
    "        variables = [var for var in variables if var not in ['and', 'or', 'not']]\n",
    "    # print(variables)\n",
    "\n",
    "    # Generate all possible combinations of True/False values for variables\n",
    "    combinations = itertools.product([True, False], repeat=len(variables))\n",
    "    # sort combinations from fewest to most True values\n",
    "    combinations = sorted(combinations, key=lambda x: sum(x))\n",
    "    combos_to_check = combinations.copy()\n",
    "    # print(combinations)\n",
    "    # find all combinations where only one variable is False\n",
    "    knockout_combos = []\n",
    "    essential_vars = [] # variables that must be True\n",
    "    # test if combo where all variables are True makes the expression True\n",
    "    all_true = dict(zip(variables, [True]*len(variables)))\n",
    "    expr = expression\n",
    "    for var in sorted(variables, key=len, reverse=True):\n",
    "        expr = expr.replace(var, str(all_true[var]))\n",
    "    if use_knockout != False:\n",
    "        if eval(expr):\n",
    "            # for each variable, find the combination where only that variable is False\n",
    "            for var in variables:\n",
    "                combo = [True]*len(variables)\n",
    "                combo[variables.index(var)] = False\n",
    "                knockout_combos.append(combo)\n",
    "                expr = expression\n",
    "                for var2 in sorted(variables, key=len, reverse=True):\n",
    "                    expr = expr.replace(var2, str(combo[variables.index(var2)]))\n",
    "                # add variable to essential_vars if it must be True\n",
    "                if not eval(expr):\n",
    "                    essential_vars.append(var)\n",
    "                    # remove all combinations where this variable is False from combos_to_check\n",
    "                    combos_to_check = [combo for combo in combos_to_check if combo[variables.index(var)]]\n",
    "    # if len(combinations) != len(combos_to_check):\n",
    "    #     print(f'Reduced combinations to check from {len(combinations)} to {len(combos_to_check)} by eliminating redundant combinations.')\n",
    "\n",
    "    true_combinations = []\n",
    "    for combo in combos_to_check:\n",
    "        # Create a dictionary mapping variables to their assigned values\n",
    "        values = dict(zip(variables, combo))\n",
    "        # print(values)\n",
    "        # print(combo)\n",
    "        # replace variables with their values in 'values'\n",
    "        # cobra.manipulation.knock_out_model_genes\n",
    "\n",
    "        expr = expression\n",
    "        # Replace variables with their values in the expression, starting with the longest variable names to avoid the risk of replacing substrings\n",
    "        for var in sorted(variables, key=len, reverse=True):\n",
    "            expr = expr.replace(var, str(values[var]))\n",
    "\n",
    "        # print(expression)\n",
    "        # print(expr,type(expr))\n",
    "        # Evaluate the expression\n",
    "        if eval(expr):\n",
    "            # check if the combination's True values are a subset of any previous ones; if so, don't add it\n",
    "            # print('\\t',values)\n",
    "            redundant_combos = False\n",
    "            true_vals = [var for var in values if values[var]]\n",
    "            for prev in true_combinations:\n",
    "                prev_vals = [var for var in prev if prev[var]]\n",
    "                if set(true_vals).issubset(set(prev_vals)) or set(prev_vals).issubset(set(true_vals)):\n",
    "                    redundant_combos = True\n",
    "                    # print('redundant:',true_vals)\n",
    "                    break\n",
    "            if not redundant_combos:\n",
    "                true_combinations.append(values)\n",
    "\n",
    "    return true_combinations\n",
    "\n",
    "# Example usage:\n",
    "expression = \"rt05 and ((rt01 and (rt012 or rt0123)) or rt04)\"\n",
    "true_combos = find_true_combinations(expression, use_knockout=True)\n",
    "print(true_combos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: SPONT not found in df_pro\n",
      "Warning: TRUE not found in df_pro\n",
      "Warning: TRUE not found in df_pro\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/fx/q_x2rm651s91fj2m6_tb19pxvpxbmb/T/ipykernel_64752/1784060244.py:98: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_enz_rxns = pd.concat([df_enz_rxns, pd.DataFrame(new_rows)], sort=False, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# Automatically create placeholders for enzymes based on GPRs in model.reactions\n",
    "rxns_to_skip = set(rxns_not_needing_enzymes)\n",
    "enz_rxn_dict = {}\n",
    "spont_rxn_dict = {}\n",
    "rxn_updates = {rxn.id: {} for rxn in model.reactions}\n",
    "GPR_enz_dict = {}\n",
    "try:\n",
    "\tdf_enz_rxns = read_spreadsheet('./input/ENZYME_stoich.tsv')\n",
    "\t# make set of rxns to skip, from combining the 'rxns' column cells in df_enz_rxns (to avoid reviewing existing rxns) and separating by commas. However, relying on this will prevent updates to their GPRs from being added to the model.\n",
    "\t# rxns_to_skip.update(set(df_enz_rxns.rxns.str.split(',').sum()))\n",
    "\t# make enz_rxn_dict from df_enz_rxns, with 'enz' as keys and 'rxns' and 'protein_stoich' as values\n",
    "\tfor index, row in df_enz_rxns.iterrows():\n",
    "\t\tenz_rxn_dict[row['enz']] = {'rxns': sorted(list(dict.fromkeys(row['rxns'].split(',')))),\n",
    "\t\t\t\t\t\t\t\t\t\t'protein_stoich': {prot.split(':')[0]: float(prot.split(':')[1]) for prot in row['protein_stoich'].split(',')}}\n",
    "\t\tif not pd.isna(row[mw]):\n",
    "\t\t\tenz_rxn_dict[row['enz']][mw] = row[mw]\n",
    "except:\n",
    "\tdf_enz_rxns = pd.DataFrame(columns=['enz','rxns','protein_stoich',mw])\n",
    "# add dict of protein_stoich to enz_rxn_dict, with each protein as a key and its stoichiometry as the value\n",
    "\n",
    "# find all unique GPRs in model.reactions, and their associated reactions\n",
    "for rxn in model.reactions:\n",
    "\t# print(rxn.id, rxn.gene_reaction_rule)\n",
    "\tif rxn.id in rxns_to_skip:\n",
    "\t\tcontinue\n",
    "\tif rxn.gene_reaction_rule not in spont_GPRs + ['']:\n",
    "\t\tif rxn.gene_reaction_rule in GPR_enz_dict.keys():\n",
    "\t\t\t# add rxn to corresponding enz in enz_rxn_dict\n",
    "\t\t\tenz_ids = GPR_enz_dict[rxn.gene_reaction_rule]\n",
    "\t\t\tfor enz_id in enz_ids:\n",
    "\t\t\t\tenz_rxn_dict[enz_id]['rxns'].append(rxn.id)\n",
    "\t\t\t\trxn_updates[rxn.id].setdefault('enz',enz_id)\n",
    "\t\telse:\n",
    "\t\t\tGPR_enz_dict[rxn.gene_reaction_rule] = set()\n",
    "\t\t\tenzs = sorted(find_true_combinations(rxn.gene_reaction_rule, use_knockout=True, variables=[g.id for g in rxn.genes]), key=lambda x: sorted(x.keys()))\n",
    "\t\t\tfor enz in enzs:\n",
    "\t\t\t\t# print('\\t',enz)\n",
    "\t\t\t\tenz_found = None\n",
    "\t\t\t\t# default: 1 copy of each protein is used\n",
    "\t\t\t\t# prot_stoich = ','.join([gene+':1' for gene in sorted(enz.keys()) if enz[gene]])\n",
    "\t\t\t\tprot_stoich = {gene:1 for gene in sorted(enz.keys()) if enz[gene]}\n",
    "\t\t\t\t# check if any enzymes with the same stoichiometry are already in enz_rxn_dict\n",
    "\t\t\t\tenz_id = ''.join(sorted(prot_stoich.keys()))\n",
    "\n",
    "\t\t\t\tfor old_enz_id in enz_rxn_dict:\n",
    "\t\t\t\t\t# print(enz_id,enz_rxn_dict[enz_id])\n",
    "\t\t\t\t\t# automatically assumes enzymes are the same if they have the same proteins, regardless of order/count\n",
    "\t\t\t\t\tif sorted(enz_rxn_dict[old_enz_id]['protein_stoich'].keys()) == sorted(prot_stoich.keys()):\n",
    "\t\t\t\t\t\tenz_rxn_dict[old_enz_id]['rxns'].append(rxn.id)\n",
    "\t\t\t\t\t\tenz_found = old_enz_id\n",
    "\t\t\t\t\t\tbreak\n",
    "\t\t\t\tif not enz_found:\n",
    "\t\t\t\t\t# enz_rxn_dict[rxn.id] = enz\n",
    "\t\t\t\t\tenz_rxn_dict[enz_id] = {'rxns':[rxn.id],\n",
    "\t\t\t\t\t\t\t\t\t\t\t'protein_stoich':prot_stoich}\n",
    "\t\t\t\t\trxn_updates[rxn.id].setdefault('enz',enz_id)\n",
    "\t\t\t\t\tGPR_enz_dict[rxn.gene_reaction_rule].add(enz_id)\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tGPR_enz_dict[rxn.gene_reaction_rule].add(enz_found)\n",
    "\telse:\n",
    "\t\t# add to spont_rxn_dict under its GPR as a list of rxn IDs, if it is not already there\n",
    "\t\tspont_rxn_dict[rxn.gene_reaction_rule] = spont_rxn_dict.get(rxn.gene_reaction_rule, {'rxns':[],'protein_stoich':{'zeroCost':1}})\n",
    "\t\tspont_rxn_dict[rxn.gene_reaction_rule]['rxns'].append(rxn.id)\n",
    "\t\trxn_updates[rxn.id].setdefault('gene_reaction_rule',rxn.gene_reaction_rule)\n",
    "\t\t# spont_rxn_dict[rxn.gene_reaction_rule] = spont_rxn_dict.get(rxn.gene_reaction_rule, [rxn.id])\n",
    "\t\t# spont_rxn_dict[rxn.gene_reaction_rule] = rxn.id\n",
    "\t\t\t\t\t\n",
    "# give each enz in enz_rxn_dict a row in df_enz_rxns\n",
    "new_rows = []\n",
    "for enz in enz_rxn_dict:\n",
    "\t# only add enzymes that are not already in df_enz_rxns\n",
    "\tif df_enz_rxns[df_enz_rxns.enz == enz].empty:\n",
    "\t\t# calculate MW of each enzyme based on its protein stoichiometry\n",
    "\t\tif mw in enz_rxn_dict[enz] and enz_rxn_dict[enz][mw]:\n",
    "\t\t\tenz_mw = enz_rxn_dict[enz][mw]\n",
    "\t\telse:\n",
    "\t\t\tenz_mw = 0\n",
    "\t\t\tfor prot in enz_rxn_dict[enz]['protein_stoich']:\n",
    "\t\t\t\tpro_mw = df_pro.loc[prot,mw] if prot in df_pro.index else 0\n",
    "\t\t\t\t# check in gene_src column\n",
    "\t\t\t\tif pro_mw == 0:\n",
    "\t\t\t\t\tif prot in df_pro.gene_src.to_list():\n",
    "\t\t\t\t\t\t# print(prot,df_pro.loc[df_pro.gene_src == prot,mw])\n",
    "\t\t\t\t\t\t# if there are multiple proteins with the same gene_src, check if their MW is the same\n",
    "\t\t\t\t\t\t# if not, print a warning\n",
    "\t\t\t\t\t\t# pick the first one\n",
    "\t\t\t\t\t\tpro_match = df_pro.loc[df_pro.gene_src == prot,mw].to_list()\n",
    "\t\t\t\t\t\tif df_pro.loc[df_pro.gene_src == prot,mw].nunique() > 1:\n",
    "\t\t\t\t\t\t\tprint(f'Warning: {prot} has multiple MWs in df_pro')\n",
    "\t\t\t\t\t\tpro_mw = pro_match[0]\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tprint(f'Warning: {prot} not found in df_pro')\n",
    "\t\t\t\tenz_mw += pro_mw*enz_rxn_dict[enz]['protein_stoich'][prot]\n",
    "\t\tnew_rows.append({'enz':enz,'rxns':','.join(enz_rxn_dict[enz]['rxns']),\n",
    "\t\t\t\t\t\t'protein_stoich':','.join([prot+':'+str(enz_rxn_dict[enz]['protein_stoich'][prot]) for prot in enz_rxn_dict[enz]['protein_stoich']]),\n",
    "\t\t\t\t\t\tmw:enz_mw,\n",
    "\t\t\t\t\t\t'status':'automatically added'})\n",
    "df_enz_rxns = pd.concat([df_enz_rxns, pd.DataFrame(new_rows)], sort=False, ignore_index=True)\n",
    "# sort by id, MW, and then by rxns\n",
    "df_enz_rxns = df_enz_rxns.sort_values(by=['rxns',mw,'enz'])\n",
    "# save automatically updated df_enz_rxns as separate file\n",
    "df_enz_rxns.to_csv('./model/ENZYME_stoich.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating 12AMANTF_g with {'enz': 'rt2093'}\n",
      "Updating 13BDGLUCANt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating 13BGH_e with {'enz': 'rt5185'}\n",
      "Updating 13GS_c with {'enz': 'rt7616'}\n",
      "Updating 14BMANTF_c with {'enz': 'rt6298'}\n",
      "Updating 14DMLANOSTt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 16GS_c with {'enz': 'rt0150'}\n",
      "Updating 1AGPCt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating 1AGPEt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating 1MLCLAT_mm with {'enz': 'rt8106'}\n",
      "Updating 2DDA7Pt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 2DHPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 2DOXG6PP_c with {'enz': 'rt5045'}\n",
      "Updating 2MBACt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 2MBALDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 2MBALDt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 2MBTOHt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 2MBTOHt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 2OBUTt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 2OGMAH_c with {'enz': 'rt6557'}\n",
      "Updating 2OSUCAH_c with {'enz': 'rt6557'}\n",
      "Updating 2OXOADPt_c_m with {'enz': 'rt2267'}\n",
      "Updating 2PHETOHt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 2PHETOHt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 3C3HMPt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 3C3HMPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 3C4MOPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 3DHSKD_c with {'enz': 'rt3472'}\n",
      "Updating 3DSPHR_r with {'enz': 'rt1611'}\n",
      "Updating 3HACD200_rm with {'enz': 'rt2309'}\n",
      "Updating 3HACD220_rm with {'enz': 'rt2309'}\n",
      "Updating 3HACD240_rm with {'enz': 'rt2309'}\n",
      "Updating 3HACD260_rm with {'enz': 'rt2309'}\n",
      "Updating 3HAD100_c with {'enz': 'rt0302'}\n",
      "Updating 3HAD120_c with {'enz': 'rt0302'}\n",
      "Updating 3HAD140_c with {'enz': 'rt0302'}\n",
      "Updating 3HAD160_c with {'enz': 'rt0302'}\n",
      "Updating 3HAD180_c with {'enz': 'rt0302'}\n",
      "Updating 3HAD40_c with {'enz': 'rt0302'}\n",
      "Updating 3HAD40_m with {'enz': 'rt7207'}\n",
      "Updating 3HAD60_c with {'enz': 'rt0302'}\n",
      "Updating 3HAD60_m with {'enz': 'rt7207'}\n",
      "Updating 3HAD80_c with {'enz': 'rt0302'}\n",
      "Updating 3HAD80_m with {'enz': 'rt7207'}\n",
      "Updating 3HAO_c with {'enz': 'rt0234'}\n",
      "Updating 3HBHYOX_c with {'enz': 'rt5813'}\n",
      "Updating 3HBRDH_c with {'enz': 'rt4427'}\n",
      "Updating 3HCINNMH_c with {'enz': 'rt6518'}\n",
      "Updating 3HPPPNH_c with {'enz': 'rt6518'}\n",
      "Updating 3HXPHBt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 3INDACNLAH_c with {'enz': 'rt4042'}\n",
      "Updating 3IPM3MT_c with {'enz': 'rt6691'}\n",
      "Updating 3MBALDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 3MBALDt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 3MLCLAT_mm with {'enz': 'rt8106'}\n",
      "Updating 3MOBDC_c with {'enz': 'rt7423'}\n",
      "Updating 3MOBR_m with {'enz': 'rt4198rt7068'}\n",
      "Updating 3MOPDC_c with {'enz': 'rt7423'}\n",
      "Updating 3MOPR_m with {'enz': 'rt4198'}\n",
      "Updating 3MOPt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 3NP4H5MOBCL_m with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating 3NPDHBZH_m with {'enz': 'rt7829'}\n",
      "Updating 3NPDHMT_m with {'enz': 'rt3097'}\n",
      "Updating 3OACE200_rm with {'enz': 'rt2975'}\n",
      "Updating 3OACE220_rm with {'enz': 'rt8287'}\n",
      "Updating 3OACE240_rm with {'enz': 'rt8287'}\n",
      "Updating 3OACE260_rm with {'enz': 'rt8287'}\n",
      "Updating 3OACR200_rm with {'enz': 'rt8327'}\n",
      "Updating 3OACR220_rm with {'enz': 'rt8327'}\n",
      "Updating 3OACR240_rm with {'enz': 'rt8327'}\n",
      "Updating 3OACR260_rm with {'enz': 'rt8327'}\n",
      "Updating 3OADPCOAT_m with {'enz': 'rt4722'}\n",
      "Updating 3OAR100_c with {'enz': 'rt2190'}\n",
      "Updating 3OAR120_c with {'enz': 'rt1363'}\n",
      "Updating 3OAR140_c with {'enz': 'rt1363'}\n",
      "Updating 3OAR160_c with {'enz': 'rt1363'}\n",
      "Updating 3OAR180_c with {'enz': 'rt1363'}\n",
      "Updating 3OAR40_c with {'enz': 'rt1363'}\n",
      "Updating 3OAR40_m with {'enz': 'rt1622'}\n",
      "Updating 3OAR60_c with {'enz': 'rt1363'}\n",
      "Updating 3OAR60_m with {'enz': 'rt1622'}\n",
      "Updating 3OAR80_c with {'enz': 'rt1363'}\n",
      "Updating 3OAR80_m with {'enz': 'rt1622'}\n",
      "Updating 3OAS120_c with {'enz': 'rt0409'}\n",
      "Updating 3OAS140_c with {'enz': 'rt0409'}\n",
      "Updating 3OAS160_c with {'enz': 'rt0409'}\n",
      "Updating 3OAS180_c with {'enz': 'rt0409'}\n",
      "Updating 3OAS40_c with {'enz': 'rt0409'}\n",
      "Updating 3OAS60_c with {'enz': 'rt0409'}\n",
      "Updating 3OAS80_c with {'enz': 'rt0409'}\n",
      "Updating 4ABUTNt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 4ABUTt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 4ABUTtps_e with {'enz': 'rt4390'}\n",
      "Updating 4ABZt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 4ABZt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 4AMPMS_c with {'enz': 'rt2779'}\n",
      "Updating 4CMCOAL_c with {'enz': 'rt8267'}\n",
      "Updating 4CMCOAL_x with {'enz': 'rt3465'}\n",
      "Updating 4DPMIPPH_c with {'enz': 'rt1973'}\n",
      "Updating 4H2OGLTt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 4H2OGLTt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating 4HBZCOAS_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating 4HBZF_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating 4HBZNPT_m with {'enz': 'rt6869'}\n",
      "Updating 4HBZt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 4HGLSD_m with {'enz': 'rt0201'}\n",
      "Updating 4HPEOR_c with {'gene_reaction_rule': ''}\n",
      "Updating 4HPLH_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating 4HPPCL_c with {'gene_reaction_rule': ''}\n",
      "Updating 4HPPR_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating 4HTHRS_c with {'enz': 'rt1374'}\n",
      "Updating 4MOPR_m with {'enz': 'rt4198'}\n",
      "Updating 5AOPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating 5FTHFt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating 6DPMIPPH_c with {'enz': 'rt1395'}\n",
      "Updating AACTOOR_c with {'gene_reaction_rule': ''}\n",
      "Updating AASADy_c with {'enz': 'rt1127rt1852'}\n",
      "Updating AATA_c with {'enz': 'rt7471'}\n",
      "Updating ABTA_c with {'enz': 'rt7537'}\n",
      "Updating ABTD2Dx_c with {'enz': 'rt1622'}\n",
      "Updating ABTD2Dy_c with {'enz': 'rt8084'}\n",
      "Updating ABTD4Dx_c with {'enz': 'rt0620'}\n",
      "Updating ABTD4Dy_c with {'enz': 'rt8084'}\n",
      "Updating ABTDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating ABTLD_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating ABTt_c_e with {'enz': 'rt8486'}\n",
      "Updating ABUTD_m with {'enz': 'rt2756'}\n",
      "Updating ACACT100i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT100i_x with {'enz': 'rt6860'}\n",
      "Updating ACACT101ei_x with {'enz': 'rt5445'}\n",
      "Updating ACACT120i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT120i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT121ai_m with {'enz': 'rt0517'}\n",
      "Updating ACACT121ai_x with {'enz': 'rt5445'}\n",
      "Updating ACACT121di_x with {'enz': 'rt5445'}\n",
      "Updating ACACT122ei_x with {'enz': 'rt5445'}\n",
      "Updating ACACT140i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT140i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT141ai_m with {'enz': 'rt0517'}\n",
      "Updating ACACT141ai_x with {'enz': 'rt5445'}\n",
      "Updating ACACT141bi_m with {'enz': 'rt0517'}\n",
      "Updating ACACT141bi_x with {'enz': 'rt5445'}\n",
      "Updating ACACT142i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT143i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT160i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT160i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT161ai_m with {'enz': 'rt0517'}\n",
      "Updating ACACT161ai_x with {'enz': 'rt5445'}\n",
      "Updating ACACT161bi_m with {'enz': 'rt0517'}\n",
      "Updating ACACT161bi_x with {'enz': 'rt5445'}\n",
      "Updating ACACT162i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT163i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT180i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT180i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT181ai_m with {'enz': 'rt0517'}\n",
      "Updating ACACT181ai_x with {'enz': 'rt5445'}\n",
      "Updating ACACT182i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT183i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT200i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT200i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT220i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT220i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT240i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT240i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT260i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT260i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT40i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT40i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT40ir_c with {'enz': 'rt0310'}\n",
      "Updating ACACT60i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT60i_x with {'enz': 'rt5445'}\n",
      "Updating ACACT80i_m with {'enz': 'rt0517'}\n",
      "Updating ACACT80i_x with {'enz': 'rt5445'}\n",
      "Updating ACALDCD_c with {'enz': 'rt7423'}\n",
      "Updating ACALDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating ACALDt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating ACCOAC_c with {'enz': 'rt0271rt8036'}\n",
      "Updating ACCOATsucac_m with {'enz': 'rt5429'}\n",
      "Updating ACCOATsucpp_m with {'enz': 'rt5429'}\n",
      "Updating ACCOAt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating ACCOAt_c_r with {'enz': 'rt2578'}\n",
      "Updating ACEST1_m with {'enz': 'rt1634'}\n",
      "Updating ACEST1_x with {'enz': 'rt0018rt2393'}\n",
      "Updating ACGAM6PS_c with {'enz': 'rt4107'}\n",
      "Updating ACGAM6S1_r with {'enz': 'rt5121'}\n",
      "Updating ACGAM6S2_r with {'enz': 'rt5121'}\n",
      "Updating ACGAMPM_c with {'enz': 'rt4597'}\n",
      "Updating ACGK_m with {'enz': 'rt1009'}\n",
      "Updating ACGS_m with {'enz': 'rt7899'}\n",
      "Updating ACHAH_c with {'enz': 'rt3378'}\n",
      "Updating ACHAH_e with {'enz': 'rt1203'}\n",
      "Updating ACHBS_m with {'enz': 'rt7317'}\n",
      "Updating ACHLE1_c with {'enz': 'rt6872'}\n",
      "Updating ACHLE2_c with {'enz': 'rt6872'}\n",
      "Updating ACHLE3_c with {'enz': 'rt6872'}\n",
      "Updating ACITL_c with {'enz': 'rt1358'}\n",
      "Updating ACKr_c with {'enz': 'rt5015'}\n",
      "Updating ACLS_m with {'enz': 'rt7317'}\n",
      "Updating ACOAD100f_m with {'enz': 'rt4202'}\n",
      "Updating ACOAD100f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD101df_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD102ef_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD120f_m with {'enz': 'rt4202'}\n",
      "Updating ACOAD120f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD121af_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD140f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD141af_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD141bf_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD142f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD143f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD160f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD161af_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD161bf_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD162f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD163f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD180f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD181af_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD182f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD183f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD40f_m with {'enz': 'rt5702'}\n",
      "Updating ACOAD40f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD60f_m with {'enz': 'rt5702'}\n",
      "Updating ACOAD60f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD80f_m with {'enz': 'rt4202'}\n",
      "Updating ACOAD80f_x with {'enz': 'rt0595'}\n",
      "Updating ACOAD82ef_x with {'enz': 'rt0595'}\n",
      "Updating ACOADS160_rm with {'enz': 'rt1362'}\n",
      "Updating ACOADS180_rm with {'enz': 'rt1362'}\n",
      "Updating ACOADS181_rm with {'enz': 'rt0477'}\n",
      "Updating ACOADS182_3_rm with {'enz': 'rt8487'}\n",
      "Updating ACOADS182_rm with {'enz': 'rt0477'}\n",
      "Updating ACOADm30f_m with {'enz': 'rt5702'}\n",
      "Updating ACOADm40f_m with {'enz': 'rt5702'}\n",
      "Updating ACOADm50f_m with {'enz': 'rt1644'}\n",
      "Updating ACOAO100_x with {'enz': 'rt4374'}\n",
      "Updating ACOAO101d_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO102e_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO120_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO121a_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO140_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO141a_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO141b_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO142_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO143_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO160_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO161a_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO161b_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO162_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO163_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO180_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO181a_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO182_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO183_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO200_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO220_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO240_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO260_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO40_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO60_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO80_x with {'enz': 'rt4384'}\n",
      "Updating ACOAO82e_x with {'enz': 'rt4384'}\n",
      "Updating ACOATA_c with {'enz': 'rt0302'}\n",
      "Updating ACOATA_m with {'enz': 'rt5893'}\n",
      "Updating ACONIs_c with {'gene_reaction_rule': ''}\n",
      "Updating ACONIs_m with {'gene_reaction_rule': ''}\n",
      "Updating ACONMT_c with {'enz': 'rt6691'}\n",
      "Updating ACONT2_m with {'enz': 'rt4274'}\n",
      "Updating ACONTa_m with {'enz': 'rt6232'}\n",
      "Updating ACONTb_m with {'enz': 'rt3256'}\n",
      "Updating ACOTAi_m with {'enz': 'rt6510'}\n",
      "Updating ACP1_e with {'enz': 'rt1163'}\n",
      "Updating ACPCS_c with {'enz': 'rt0245'}\n",
      "Updating ACRNAH_x with {'enz': 'rt5418'}\n",
      "Updating ACRNt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating ACRNt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating ACS_c with {'enz': 'rt7532'}\n",
      "Updating ACS_n with {'enz': 'rt7532'}\n",
      "Updating ACt_c_e with {'enz': 'rt7285'}\n",
      "Updating ACt_c_m with {'enz': 'rt5456'}\n",
      "Updating ACt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating ADA_c with {'enz': 'rt8364'}\n",
      "Updating ADCL_c with {'enz': 'rt4108'}\n",
      "Updating ADCS_c with {'enz': 'rt5301'}\n",
      "Updating ADD_c with {'enz': 'rt3300'}\n",
      "Updating ADEt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating ADEtps_e with {'enz': 'rt6190'}\n",
      "Updating ADK1_c with {'enz': 'rt7128'}\n",
      "Updating ADK1_m with {'enz': 'rt6761'}\n",
      "Updating ADK3_m with {'enz': 'rt7128'}\n",
      "Updating ADMDC_c with {'enz': 'rt2849'}\n",
      "Updating ADNCYC_c with {'enz': 'rt4919'}\n",
      "Updating ADNK1_c with {'enz': 'rt0017'}\n",
      "Updating ADNtps_e with {'enz': 'rt5601'}\n",
      "Updating ADPATPt_c_m with {'enz': 'rt4336'}\n",
      "Updating ADPATPt_c_x with {'enz': 'rt7777'}\n",
      "Updating ADPRDP_m with {'enz': 'rt8143'}\n",
      "Updating ADPT_c with {'enz': 'rt7825'}\n",
      "Updating ADPh_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ADPh_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ADPt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating ADPt_c_gm with {'gene_reaction_rule': ''}\n",
      "Updating ADPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating ADPt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating ADPt_c_vm with {'gene_reaction_rule': ''}\n",
      "Updating ADSK_c with {'enz': 'rt0341'}\n",
      "Updating ADSL1r_c with {'enz': 'rt7437'}\n",
      "Updating ADSL2i_c with {'enz': 'rt7437'}\n",
      "Updating ADSS_c with {'enz': 'rt3802'}\n",
      "Updating AEPI1_c with {'enz': 'rt0136'}\n",
      "Updating AEPI2_c with {'enz': 'rt0136'}\n",
      "Updating AFAT_c with {'enz': 'rt3174'}\n",
      "Updating AFLTXNB1S_c with {'enz': 'rt2689'}\n",
      "Updating AFLTXNB2S_c with {'enz': 'rt2689'}\n",
      "Updating AGMT_e with {'enz': 'rt1772'}\n",
      "Updating AGNPR_l with {'enz': 'rt7207'}\n",
      "Updating AGNPR_rm with {'enz': 'rt7207'}\n",
      "Updating AGPAT_l with {'enz': 'rt2059'}\n",
      "Updating AGPAT_rm with {'enz': 'rt7662'}\n",
      "Updating AGPRi_m with {'enz': 'rt1009'}\n",
      "Updating AGTi_c with {'enz': 'rt0451'}\n",
      "Updating AHCYSt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating AHCYSt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating AHCi_c with {'enz': 'rt4544'}\n",
      "Updating AHGDf_m with {'enz': 'rt1883'}\n",
      "Updating AHMMPS4_c with {'enz': 'rt2779'}\n",
      "Updating AHSERL2_c with {'enz': 'rt8250'}\n",
      "Updating AHSERL_c with {'enz': 'rt8250'}\n",
      "Updating AICART_c with {'enz': 'rt4053'}\n",
      "Updating AIRC1_c with {'enz': 'rt3764'}\n",
      "Updating AKGCITta_m with {'enz': 'rt8378'}\n",
      "Updating AKGDH_m with {'enz': 'rt1639rt1672rt3748'}\n",
      "Updating AKGMALta_m with {'enz': 'rt2267'}\n",
      "Updating AKGt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating AKGt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating AKGt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating AKP1_c with {'enz': 'rt5041'}\n",
      "Updating ALAS_m with {'enz': 'rt0309'}\n",
      "Updating ALATA_L_m with {'enz': 'rt1267'}\n",
      "Updating ALATRS_c with {'enz': 'rt3605'}\n",
      "Updating ALAt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating ALAtps_e with {'enz': 'rt5861'}\n",
      "Updating ALCD19y_c with {'enz': 'rt3514'}\n",
      "Updating ALCD22xi_c with {'enz': 'rt8149'}\n",
      "Updating ALCD22xi_m with {'enz': 'rt3939'}\n",
      "Updating ALCD22yi_c with {'enz': 'rt4416'}\n",
      "Updating ALCD23xi_c with {'enz': 'rt8149'}\n",
      "Updating ALCD23xi_m with {'enz': 'rt3939'}\n",
      "Updating ALCD24xi_c with {'enz': 'rt8149'}\n",
      "Updating ALCD24xi_m with {'enz': 'rt3939'}\n",
      "Updating ALCD24yi_c with {'enz': 'rt5194'}\n",
      "Updating ALCD25xi_c with {'enz': 'rt8149'}\n",
      "Updating ALCD25xi_m with {'enz': 'rt3939'}\n",
      "Updating ALCD25yi_c with {'enz': 'rt5194'}\n",
      "Updating ALCD26xi_c with {'enz': 'rt8149'}\n",
      "Updating ALCD26xi_m with {'enz': 'rt3939'}\n",
      "Updating ALCD2i1_c with {'enz': 'rt3282'}\n",
      "Updating ALCD2i2_c with {'enz': 'rt7070'}\n",
      "Updating ALCD2i2_m with {'enz': 'rt3939'}\n",
      "Updating ALDD20x_m with {'enz': 'rt2756'}\n",
      "Updating ALDD20y_c with {'enz': 'rt7041'}\n",
      "Updating ALDD20y_m with {'enz': 'rt3674'}\n",
      "Updating ALDD22x_c with {'enz': 'rt7955'}\n",
      "Updating ALDD2x_c with {'enz': 'rt2756'}\n",
      "Updating ALDD2y_c with {'enz': 'rt3674'}\n",
      "Updating ALLTN_c with {'enz': 'rt3516'}\n",
      "Updating ALLTNt_c with {'enz': 'rt4534'}\n",
      "Updating ALLTTt_c with {'enz': 'rt6702'}\n",
      "Updating ALPHNH_c with {'enz': 'rt0958'}\n",
      "Updating AMAOTr_c with {'enz': 'rt4363'}\n",
      "Updating AMETt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating AMETt_c_m with {'enz': 'rt3664'}\n",
      "Updating AMETt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating AMID_c with {'enz': 'rt3257'}\n",
      "Updating AMPATPt_c_x with {'enz': 'rt7777'}\n",
      "Updating AMPDA_c with {'enz': 'rt3101'}\n",
      "Updating AMPt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating AMPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating AMPt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating ANNAT_c with {'enz': 'rt7589'}\n",
      "Updating ANPRT_c with {'enz': 'rt1532'}\n",
      "Updating ANS_c with {'enz': 'rt6741rt8196'}\n",
      "Updating AOBUTDs_c with {'gene_reaction_rule': ''}\n",
      "Updating AOXS_c with {'enz': 'rt4368'}\n",
      "Updating AP4AH_c with {'enz': 'rt0084'}\n",
      "Updating ARAB14LO_c with {'enz': 'rt5365'}\n",
      "Updating ARAB14LO_x with {'enz': 'rt2122'}\n",
      "Updating ARAB1Dx_c with {'enz': 'rt8345'}\n",
      "Updating ARABDt_c with {'gene_reaction_rule': ''}\n",
      "Updating ARABLt_c with {'gene_reaction_rule': ''}\n",
      "Updating ARABRy_c with {'enz': 'rt5845'}\n",
      "Updating ARACHt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating ARACHt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating ARAI2_c with {'enz': 'rt6112'}\n",
      "Updating ARAK_c with {'enz': 'rt4526'}\n",
      "Updating ARGN_c with {'enz': 'rt7813'}\n",
      "Updating ARGSL_c with {'enz': 'rt3236'}\n",
      "Updating ARGSS_c with {'enz': 'rt7828'}\n",
      "Updating ARGTRS_c with {'enz': 'rt1711'}\n",
      "Updating ARGtpa_v with {'enz': 'rt0111'}\n",
      "Updating ARGtps_e with {'enz': 'rt1689'}\n",
      "Updating ARGtps_m with {'gene_reaction_rule': ''}\n",
      "Updating ARMT_c with {'enz': 'rt1648'}\n",
      "Updating AROHL_c with {'enz': 'rt5827'}\n",
      "Updating ARSO4H_c with {'enz': 'rt6899'}\n",
      "Updating ASAD_c with {'enz': 'rt1962'}\n",
      "Updating ASCBOX_c with {'enz': 'rt5489'}\n",
      "Updating ASNN_e with {'enz': 'rt1675'}\n",
      "Updating ASNS1_c with {'enz': 'rt6769'}\n",
      "Updating ASNTRS_c with {'enz': 'rt0799'}\n",
      "Updating ASNTRS_m with {'enz': 'rt3755'}\n",
      "Updating ASNt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating ASNtpa_v with {'enz': 'rt7819'}\n",
      "Updating ASNtps_v with {'enz': 'rt4375'}\n",
      "Updating ASPCT_c with {'enz': 'rt8313'}\n",
      "Updating ASPGLUt_c_m with {'enz': 'rt8431'}\n",
      "Updating ASPGLUt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating ASPK_c with {'enz': 'rt6294'}\n",
      "Updating ASPTA4_c with {'enz': 'rt1784'}\n",
      "Updating ASPTA_c with {'enz': 'rt0568'}\n",
      "Updating ASPTA_x with {'enz': 'rt5913'}\n",
      "Updating ASPTAi_m with {'enz': 'rt5562'}\n",
      "Updating ASPTRS_c with {'enz': 'rt5890'}\n",
      "Updating ASPtps_v with {'enz': 'rt0398'}\n",
      "Updating ATAH_c with {'enz': 'rt6356'}\n",
      "Updating ATHRDHr_c with {'enz': 'rt1844'}\n",
      "Updating ATPASEP2e_c with {'enz': 'rt0602'}\n",
      "Updating ATPASEV_v with {'enz': 'rt0842rt1254rt1380rt2434rt2657rt2684rt2696rt2749rt3035rt6603rt6941rt6983rt7595rt8029'}\n",
      "Updating ATPATF1_c with {'enz': 'rt5328'}\n",
      "Updating ATPHs_c with {'gene_reaction_rule': ''}\n",
      "Updating ATPM_c with {'gene_reaction_rule': ''}\n",
      "Updating ATPPRT_c with {'enz': 'rt1520'}\n",
      "Updating ATPS_m with {'enz': 'rt0712rt2306rt3590rt3599rt5056rt5391rt5474rt6418rt6544rt7221rt7991rtmATP6rtmATP8rtmATP9'}\n",
      "Updating ATPt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating ATPt_c_gm with {'gene_reaction_rule': ''}\n",
      "Updating ATPt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating ATPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating ATPt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating ATPt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating ATPt_c_vm with {'gene_reaction_rule': ''}\n",
      "Updating AVNORy_c with {'enz': 'rt7877'}\n",
      "Updating BACCL_c with {'enz': 'rt8036'}\n",
      "Updating BBOX_m with {'enz': 'rt6692'}\n",
      "Updating BETALDHx_x with {'enz': 'rt2053'}\n",
      "Updating BIOMASS_PinheiroEtAl2020 with {'gene_reaction_rule': ''}\n",
      "Updating BIOMASS_xyl_hybrid with {'gene_reaction_rule': ''}\n",
      "Updating BPNT_c with {'enz': 'rt2305'}\n",
      "Updating BTDDH_c with {'enz': 'rt5273'}\n",
      "Updating BTDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating BTS1_c with {'enz': 'rt7540'}\n",
      "Updating BTS1_m with {'enz': 'rt7540'}\n",
      "Updating BUTt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating BUTt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating BZHYOX_c with {'enz': 'rt5813'}\n",
      "Updating BZHYOX_r with {'enz': 'rt1936'}\n",
      "Updating C14STR_c with {'enz': 'rt0303'}\n",
      "Updating C22STDSy_c with {'enz': 'rt3348'}\n",
      "Updating C24STR_r with {'enz': 'rt2613'}\n",
      "Updating C3STDH1_c with {'enz': 'rt0467'}\n",
      "Updating C3STDH2_c with {'enz': 'rt0467'}\n",
      "Updating C3STKR1_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating C3STKR2_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating C4STMO1_c with {'enz': 'rt8272'}\n",
      "Updating C4STMO2_c with {'enz': 'rt8272'}\n",
      "Updating C4STMO3_c with {'enz': 'rt8272'}\n",
      "Updating C4STMO4_c with {'enz': 'rt8272'}\n",
      "Updating C5STDS_c with {'enz': 'rt0413'}\n",
      "Updating C8STI_c with {'enz': 'rt1745'}\n",
      "Updating CA2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating CAROL1_c with {'enz': 'rt0221'}\n",
      "Updating CAROL2_c with {'enz': 'rt0221'}\n",
      "Updating CAROL3_c with {'enz': 'rt0221'}\n",
      "Updating CAROMOX1_c with {'enz': 'rt0220'}\n",
      "Updating CAROMOX2_c with {'enz': 'rt0220'}\n",
      "Updating CATDOX_c with {'enz': 'rt0267'}\n",
      "Updating CATOX_e with {'enz': 'rt4763'}\n",
      "Updating CAT_c with {'enz': 'rt0986'}\n",
      "Updating CAT_x with {'enz': 'rt6937'}\n",
      "Updating CBL_c with {'enz': 'rt0391'}\n",
      "Updating CBMHL_c with {'enz': 'rt2726'}\n",
      "Updating CBPS_c with {'enz': 'rt3934rt4929rt8313'}\n",
      "Updating CDPCHOLt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating CDPDAGS_mm with {'enz': 'rt4513'}\n",
      "Updating CDPDAGS_rm with {'enz': 'rt4513'}\n",
      "Updating CDPEAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating CDPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating CDPt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating CELLBH_e with {'enz': 'rt8348'}\n",
      "Updating CEPHSCOX_x with {'enz': 'rt3508'}\n",
      "Updating CERH124A_r with {'enz': 'rt6946'}\n",
      "Updating CERH124B_r with {'enz': 'rt1296'}\n",
      "Updating CERH126A_r with {'enz': 'rt6946'}\n",
      "Updating CERH126B_r with {'enz': 'rt1296'}\n",
      "Updating CERH2A24_r with {'enz': 'rt1296'}\n",
      "Updating CERH2A26_r with {'enz': 'rt1296'}\n",
      "Updating CERH324_r with {'enz': 'rt1296'}\n",
      "Updating CERH326_r with {'enz': 'rt1296'}\n",
      "Updating CERS124_c with {'enz': 'rt3023'}\n",
      "Updating CERS126_c with {'enz': 'rt6800'}\n",
      "Updating CERS2A24_r with {'enz': 'rt6800'}\n",
      "Updating CERS2A26_r with {'enz': 'rt6800'}\n",
      "Updating CERt_g_r with {'gene_reaction_rule': ''}\n",
      "Updating CFCOAMT_c with {'enz': 'rt5076'}\n",
      "Updating CHITOSNSE_c with {'enz': 'rt2528'}\n",
      "Updating CHLPCTD_c with {'enz': 'rt2172'}\n",
      "Updating CHLSTI2_c with {'enz': 'rt5191'}\n",
      "Updating CHOLD_c with {'enz': 'rt3274'}\n",
      "Updating CHOLK_c with {'enz': 'rt7614'}\n",
      "Updating CHOLPT_rm with {'enz': 'rt2983'}\n",
      "Updating CHOLt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating CHOLtps_e with {'enz': 'rt7623'}\n",
      "Updating CHORM_c with {'enz': 'rt1336'}\n",
      "Updating CHORS_c with {'enz': 'rt5669'}\n",
      "Updating CHRPL_c with {'gene_reaction_rule': ''}\n",
      "Updating CHTNDA_c with {'enz': 'rt0086'}\n",
      "Updating CHTNH_e with {'enz': 'rt6409'}\n",
      "Updating CHTNH_v with {'enz': 'rt4284'}\n",
      "Updating CHTNS_c with {'enz': 'rt6592'}\n",
      "Updating CITCOAL_c with {'enz': 'rt1454'}\n",
      "Updating CITICITta_m with {'enz': 'rt2146'}\n",
      "Updating CITICTta_x with {'gene_reaction_rule': ''}\n",
      "Updating CITMALta_m with {'enz': 'rt2146'}\n",
      "Updating CITMALta_x with {'gene_reaction_rule': ''}\n",
      "Updating CITt3_m with {'enz': 'rt2146'}\n",
      "Updating CITt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating CLAH1_mm with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating CLAH3_mm with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating CLPNS_mm with {'enz': 'rt8047'}\n",
      "Updating CLt_c_e with {'enz': 'rt6468'}\n",
      "Updating CMBOLHcis_c with {'enz': 'rt6386'}\n",
      "Updating CMBOLHtrans_c with {'enz': 'rt6386'}\n",
      "Updating CMPA_c with {'enz': 'rt3507'}\n",
      "Updating CMPN_c with {'gene_reaction_rule': ''}\n",
      "Updating CMPt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating CMPt_m_mm with {'gene_reaction_rule': ''}\n",
      "Updating CMUCOC_c with {'enz': 'rt5764'}\n",
      "Updating CO2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_gm with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_m with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_n with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_r with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_vm with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_x with {'gene_reaction_rule': ''}\n",
      "Updating COAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating COAt_c_m with {'enz': 'rt0548'}\n",
      "Updating COAt_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating COAt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating COAt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating COAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating COAt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating COBALT2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating CPLSPNEST_e with {'enz': 'rt3377'}\n",
      "Updating CPPPGO_c with {'enz': 'rt5829'}\n",
      "Updating CRN100t_c_m with {'enz': 'rt0963'}\n",
      "Updating CRN100t_c_x with {'enz': 'rt7415'}\n",
      "Updating CRN120t_c_m with {'enz': 'rt0963'}\n",
      "Updating CRN120t_c_x with {'enz': 'rt7415'}\n",
      "Updating CRN40t_c_m with {'enz': 'rt0963'}\n",
      "Updating CRN40t_c_x with {'enz': 'rt7415'}\n",
      "Updating CRN60t_c_m with {'enz': 'rt0963'}\n",
      "Updating CRN60t_c_x with {'enz': 'rt7415'}\n",
      "Updating CRN80t_c_m with {'enz': 'rt0963'}\n",
      "Updating CRN80t_c_x with {'enz': 'rt7415'}\n",
      "Updating CRNACRNt_c_m with {'enz': 'rt0963'}\n",
      "Updating CRNAT100_m with {'enz': 'rt7415'}\n",
      "Updating CRNAT100_x with {'enz': 'rt5212'}\n",
      "Updating CRNAT120_m with {'enz': 'rt7415'}\n",
      "Updating CRNAT120_x with {'enz': 'rt5212'}\n",
      "Updating CRNAT40_m with {'enz': 'rt7415'}\n",
      "Updating CRNAT40_x with {'enz': 'rt5212'}\n",
      "Updating CRNAT60_m with {'enz': 'rt7415'}\n",
      "Updating CRNAT60_x with {'enz': 'rt5212'}\n",
      "Updating CRNAT80_m with {'enz': 'rt7415'}\n",
      "Updating CRNAT80_x with {'enz': 'rt5212'}\n",
      "Updating CRNCARt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating CRNOAT_m with {'enz': 'rt0947'}\n",
      "Updating CRNt_c_m with {'enz': 'rt0963'}\n",
      "Updating CRNt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating CSND_c with {'enz': 'rt6191'}\n",
      "Updating CSNPURt_c_e with {'enz': 'rt6644'}\n",
      "Updating CSNtps_e with {'enz': 'rt6190'}\n",
      "Updating CS_m with {'enz': 'rt2963'}\n",
      "Updating CS_x with {'enz': 'rt0180'}\n",
      "Updating CTPCMPt_c_m with {'enz': 'rt4019'}\n",
      "Updating CTPS1_c with {'enz': 'rt8107'}\n",
      "Updating CTPS2_c with {'enz': 'rt8107'}\n",
      "Updating CTPt2_m with {'enz': 'rt4019'}\n",
      "Updating CTPt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating CTPt_m_mm with {'gene_reaction_rule': ''}\n",
      "Updating CU2t_c_e with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating CYSItps_v with {'enz': 'rt6589'}\n",
      "Updating CYSS_c with {'enz': 'rt3663'}\n",
      "Updating CYSTGL_c with {'enz': 'rt1131'}\n",
      "Updating CYSTRS_c with {'enz': 'rt1487'}\n",
      "Updating CYSTS_c with {'enz': 'rt7344'}\n",
      "Updating CYTDK2_c with {'enz': 'rt0265'}\n",
      "Updating CYTD_c with {'enz': 'rt7919'}\n",
      "Updating CYTDtps_e with {'enz': 'rt5601'}\n",
      "Updating CYTK1_c with {'gene_reaction_rule': ''}\n",
      "Updating CYTK2_c with {'gene_reaction_rule': ''}\n",
      "Updating DADA_c with {'enz': 'rt8364'}\n",
      "Updating DADPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating DADt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating DAG3Pabct_c_m with {'enz': 'rt1060'}\n",
      "Updating DAGK_rm with {'enz': 'rt1788'}\n",
      "Updating DAGL_m with {'enz': 'rt5949'}\n",
      "Updating DAGt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating DAGt_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating DAGt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating DAGt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating DAGt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating DAGt_gm_rm with {'gene_reaction_rule': ''}\n",
      "Updating DANNt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating DAO4hpro_x with {'enz': 'rt7626'}\n",
      "Updating DAOala_x with {'enz': 'rt7626'}\n",
      "Updating DAOarg_x with {'enz': 'rt7626'}\n",
      "Updating DAOgly_x with {'enz': 'rt7626'}\n",
      "Updating DAOorn_x with {'enz': 'rt7626'}\n",
      "Updating DAOpro_x with {'enz': 'rt7626'}\n",
      "Updating DB4PS_c with {'enz': 'rt8298'}\n",
      "Updating DBTS_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating DCAt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating DCAt_c_x with {'enz': 'SPONT'}\n",
      "Updating DCDPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating DCMPDA_c with {'enz': 'rt0757'}\n",
      "Updating DCYTD_c with {'enz': 'rt7919'}\n",
      "Updating DCYTt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating DDCACOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating DDCACOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating DDCAt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating DDCAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating DDCAt_c_x with {'enz': 'rt5693'}\n",
      "Updating DDCAt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating DDPA_c with {'enz': 'rt2234'}\n",
      "Updating DDPA_m with {'enz': 'rt2234'}\n",
      "Updating DECOAI82e_x with {'enz': 'rt3539'}\n",
      "Updating DECOAR102d_x with {'enz': 'rt4860'}\n",
      "Updating DECOAR103e_x with {'enz': 'rt4860'}\n",
      "Updating DECOAR82e_x with {'enz': 'rt4860'}\n",
      "Updating DESAT18_2_rm with {'enz': 'rt0477'}\n",
      "Updating DGAT_l with {'enz': 'rt8092'}\n",
      "Updating DGAT_rm with {'enz': 'rt8092'}\n",
      "Updating DGDPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating DGK1_c with {'enz': 'rt3157'}\n",
      "Updating DGLCND3H_c with {'enz': 'rt2544'}\n",
      "Updating DGLCND5H_c with {'enz': 'rt2544'}\n",
      "Updating DGPP_vm with {'enz': 'rt4719'}\n",
      "Updating DGSNt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating DHAD1_c with {'enz': 'rt6258'}\n",
      "Updating DHAD1_m with {'enz': 'rt1321'}\n",
      "Updating DHAD2_c with {'enz': 'rt6258'}\n",
      "Updating DHAD2_m with {'enz': 'rt1321'}\n",
      "Updating DHAK_c with {'enz': 'rt3901'}\n",
      "Updating DHAPAT_l with {'enz': 'rt5001'}\n",
      "Updating DHAPAT_rm with {'enz': 'rt5001'}\n",
      "Updating DHAPt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating DHAPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating DHAPt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating DHAPt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating DHFRi_c with {'enz': 'rt3791'}\n",
      "Updating DHFRi_m with {'enz': 'rt3791'}\n",
      "Updating DHFS_1_c with {'enz': 'rt2092'}\n",
      "Updating DHFt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating DHNPA2i_m with {'enz': 'rt6009'}\n",
      "Updating DHNPTt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating DHORDfum_c with {'enz': 'rt6498'}\n",
      "Updating DHORTS_c with {'enz': 'rt3923'}\n",
      "Updating DHPM2_c with {'enz': 'rt3509'}\n",
      "Updating DHPPDA2_c with {'enz': 'rt7138'}\n",
      "Updating DHPTS_m with {'enz': 'rt6009'}\n",
      "Updating DHPTt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating DHQS_c with {'enz': 'rt5884'}\n",
      "Updating DHQTi_c with {'enz': 'rt2204'}\n",
      "Updating DINt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating DIPS_c with {'enz': 'rt8137'}\n",
      "Updating DKMEOX_c with {'enz': 'rt7962'}\n",
      "Updating DKMPPH_c with {'enz': 'rt7962'}\n",
      "Updating DMATT_c with {'enz': 'rt4576'}\n",
      "Updating DM_5mta_c with {'gene_reaction_rule': ''}\n",
      "Updating DM_gcald_c with {'gene_reaction_rule': ''}\n",
      "Updating DNADt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating DNMPPA_c with {'enz': 'rt6247'}\n",
      "Updating DNTPPA_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating DNTP_c with {'enz': 'rt4736'}\n",
      "Updating DOCOSAt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating DOCOSAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating DOCOSAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating DOCOSCOAt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating DOCOSCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating DOCOSCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating DOLDPP_c with {'enz': 'rt5681'}\n",
      "Updating DOLDPt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating DOLDPt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating DOLGLCPt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating DOLK_c with {'enz': 'rt0544'}\n",
      "Updating DOLPMMT_r with {'enz': 'rt3753'}\n",
      "Updating DOLPMT_c with {'enz': 'rt5116'}\n",
      "Updating DOLPt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating DPCOAK_c with {'enz': 'rt2746'}\n",
      "Updating DPGM_c with {'enz': 'rt1542'}\n",
      "Updating DPGT_c with {'enz': 'rt3441'}\n",
      "Updating DPMVD_c with {'enz': 'rt3542'}\n",
      "Updating DPR_c with {'enz': 'rt8154'}\n",
      "Updating DRAPPRy_c with {'enz': 'rt8469'}\n",
      "Updating DRBK_c with {'enz': 'rt2501'}\n",
      "Updating DTMPK_c with {'enz': 'rt6884'}\n",
      "Updating DTTPt_c with {'gene_reaction_rule': ''}\n",
      "Updating DUDPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating DUMPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating DURIK1_c with {'gene_reaction_rule': ''}\n",
      "Updating DURIt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating DUTPDP_c with {'enz': 'rt5532'}\n",
      "Updating DXHPSq9_c with {'enz': 'rt1830'}\n",
      "Updating E4HGLUt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating E4HGLUt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating E4Pt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating EAR100y_c with {'enz': 'rt0302'}\n",
      "Updating EAR120y_c with {'enz': 'rt0302'}\n",
      "Updating EAR140y_c with {'enz': 'rt0302'}\n",
      "Updating EAR160y_c with {'enz': 'rt0302'}\n",
      "Updating EAR180y_c with {'enz': 'rt0302'}\n",
      "Updating EAR40y_c with {'enz': 'rt0302'}\n",
      "Updating EAR40y_m with {'enz': 'rt3469'}\n",
      "Updating EAR60y_c with {'enz': 'rt0302'}\n",
      "Updating EAR60y_m with {'enz': 'rt3469'}\n",
      "Updating EAR80y_c with {'enz': 'rt0302'}\n",
      "Updating EAR80y_m with {'enz': 'rt3469'}\n",
      "Updating ECOAH100_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH100_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH101e_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH120_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH120_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH121a_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH121a_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH121d_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH122e_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH140_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH140_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH141b_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH141b_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH141c_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH141c_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH142_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH143_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH160_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH160_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH161b_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH161b_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH161c_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH161c_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH162_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH163_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH180_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH180_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH181a_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH181a_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH182_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH183_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH200_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH220_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH240_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH260_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH40_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH40_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH60_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH60_x with {'enz': 'rt2994'}\n",
      "Updating ECOAH80_m with {'enz': 'rt6437'}\n",
      "Updating ECOAH80_x with {'enz': 'rt2994'}\n",
      "Updating ECOAI101a_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI101d_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI102e_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI121a_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI121b_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI122c_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI122d_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI123e_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI141a_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI142c_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI81e_x with {'enz': 'rt1925'}\n",
      "Updating ECOAI82e_x with {'enz': 'rt1925'}\n",
      "Updating ECOAR122_x with {'enz': 'rt4860'}\n",
      "Updating ECOAR142_x with {'enz': 'rt4860'}\n",
      "Updating ECOAR200_rm with {'enz': 'rt7873'}\n",
      "Updating ECOAR220_rm with {'enz': 'rt7873'}\n",
      "Updating ECOAR240_rm with {'enz': 'rt7873'}\n",
      "Updating ECOAR260_rm with {'enz': 'rt7873'}\n",
      "Updating ENO_c with {'enz': 'rt0669'}\n",
      "Updating EPGALURS_e with {'enz': 'rt2116'}\n",
      "Updating EPISTt_c_e with {'enz': 'rt6543'}\n",
      "Updating EPISTt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating EPISTt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating ERGSTESTt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating ERGSTESTt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating ERGSTGLCT_c with {'enz': 'rt5466'}\n",
      "Updating ERGSTt_c_e with {'enz': 'rt6543'}\n",
      "Updating ERGSTt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating ERGSTt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating ERGSTt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating ERGSTt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating ERGTETROLt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating ERGTETROLt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating ETACt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating ETFOXRq9_c with {'enz': 'rt1518rt1701rt1850'}\n",
      "Updating ETFOXRq9_m with {'enz': 'rt1518rt1701rt1850'}\n",
      "Updating ETHAK_c with {'enz': 'rt7614'}\n",
      "Updating ETHAt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating ETNO2OX_c with {'enz': 'rt3897'}\n",
      "Updating ETOHt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating ETOHt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating EX_13BDglucan_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_14dmlanost_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_2mbac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_2mbald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_2mbtoh_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_2phetoh_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_3c3hmp_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_3mbald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_3mop_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_4abut_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_4abz_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_5aop_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_5fthf_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_8aonn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_Nfortyr2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_abt__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_abt__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_acald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ade_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_adn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_akg_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ala__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_alltn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_alltt_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_amet_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_arab__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_arab__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_arach_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_arg__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_asn__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_asp__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_btd_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_btn_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_but_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ca2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_caro_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_cellb_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_chol_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_cit_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_cl_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_co2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_cobalt2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_crn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_csn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_cu2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_cys__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_cytd_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_dad_2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_dann_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_dca_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_dcyt_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ddca_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_dgsn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_din_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_docosa_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_dttp_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_duri_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_epist_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ergst_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ergtetrol_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_etac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_etha_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_etoh_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fe2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fe3_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fecost_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fmn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fol_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_for_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fru_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fum_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_g3pc_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_g3pi_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gal_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_galur_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gam6p_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gcald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_glc__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gln__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_glu__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_glx_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gly_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_glyc_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gsn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gthox_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gthrd_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gua_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_h2o_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_h_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hco3_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hdca_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hdcea_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_his__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hom__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hxa_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hxan_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hxca_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hxdcal_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_iamac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_iamoh_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ibtald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ibutac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ibutoh_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_id3acald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ile__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ind3eth_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_inost_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ins_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_k_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_lac__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_lac__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_lanost_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_leu__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_lys__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mal__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_malt_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_man_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_melib_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_met__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mev__R_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mg2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mmet_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mn2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mobd_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_na1_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_nac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_nad_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_nadp_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_nh4_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ni2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_nmn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_o2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_oaa_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ocdca_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ocdcea_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_octa_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_orn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pacald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pap_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pectin_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_phe__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pheac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pi_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pnto__R_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ppa_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pro__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ptrc_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pydx5p_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pydxn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pyr_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_q9h2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_raffin_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_rib__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ribflv_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_sbt__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_sbt__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ser__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_so3_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_so4_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_spmd_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_sprm_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_srb__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_succ_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_sucr_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_tag_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_taur_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_thm_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_thmmp_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_thmpp_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_thr__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_thymd_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_tre_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_trp__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ttcosa_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ttdca_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_tyr__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ura_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_urea_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_uri_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_val__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_xan_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_xtsn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_xyl__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_xylt_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_zn2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_zymst_e with {'gene_reaction_rule': ''}\n",
      "Updating F1PP_c with {'enz': 'rt0208'}\n",
      "Updating FACOA120tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA140tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA160tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA161tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA180tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA181tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA182tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA183tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA200tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA220tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA240tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOA260tabc_x with {'enz': 'rt1269rt4799'}\n",
      "Updating FACOAE100_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE120_c with {'enz': 'rt4680'}\n",
      "Updating FACOAE120_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE140_c with {'enz': 'rt4680'}\n",
      "Updating FACOAE140_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE160_c with {'enz': 'rt0088'}\n",
      "Updating FACOAE160_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE161_c with {'enz': 'rt0088'}\n",
      "Updating FACOAE161_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE180_c with {'enz': 'rt0088'}\n",
      "Updating FACOAE180_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE181_c with {'enz': 'rt0088'}\n",
      "Updating FACOAE181_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE182_c with {'enz': 'rt0088'}\n",
      "Updating FACOAE182_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE183_c with {'enz': 'rt0088'}\n",
      "Updating FACOAE183_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE40_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE60_x with {'enz': 'rt7680'}\n",
      "Updating FACOAE80_x with {'enz': 'rt7680'}\n",
      "Updating FACOAL100_x with {'enz': 'rt4187'}\n",
      "Updating FACOAL120_c with {'enz': 'rt7532'}\n",
      "Updating FACOAL120_l with {'enz': 'rt2799'}\n",
      "Updating FACOAL120_rm with {'enz': 'rt2799'}\n",
      "Updating FACOAL120_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL140_l with {'enz': 'rt2799'}\n",
      "Updating FACOAL140_rm with {'enz': 'rt2799'}\n",
      "Updating FACOAL140_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL160_l with {'enz': 'rt2799'}\n",
      "Updating FACOAL160_rm with {'enz': 'rt2799'}\n",
      "Updating FACOAL160_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL161_l with {'enz': 'rt2799'}\n",
      "Updating FACOAL161_rm with {'enz': 'rt2799'}\n",
      "Updating FACOAL161_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL180_l with {'enz': 'rt2799'}\n",
      "Updating FACOAL180_rm with {'enz': 'rt2799'}\n",
      "Updating FACOAL180_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL181_l with {'enz': 'rt2799'}\n",
      "Updating FACOAL181_rm with {'enz': 'rt2799'}\n",
      "Updating FACOAL181_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL182_l with {'enz': 'rt2799'}\n",
      "Updating FACOAL182_rm with {'enz': 'rt2799'}\n",
      "Updating FACOAL182_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL183_l with {'enz': 'rt2799'}\n",
      "Updating FACOAL183_rm with {'enz': 'rt2799'}\n",
      "Updating FACOAL183_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL200_en with {'enz': 'rt7380'}\n",
      "Updating FACOAL200_l with {'enz': 'rt1544'}\n",
      "Updating FACOAL200_rm with {'enz': 'rt1544'}\n",
      "Updating FACOAL220_en with {'enz': 'rt7380'}\n",
      "Updating FACOAL220_l with {'enz': 'rt1544'}\n",
      "Updating FACOAL220_rm with {'enz': 'rt1544'}\n",
      "Updating FACOAL220_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL240_en with {'enz': 'rt7380'}\n",
      "Updating FACOAL240_l with {'enz': 'rt1544'}\n",
      "Updating FACOAL240_rm with {'enz': 'rt1544'}\n",
      "Updating FACOAL240_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL260_en with {'enz': 'rt7380'}\n",
      "Updating FACOAL260_l with {'enz': 'rt1544'}\n",
      "Updating FACOAL260_rm with {'enz': 'rt1544'}\n",
      "Updating FACOAL260_x with {'enz': 'rt7761'}\n",
      "Updating FACOAL40_c with {'enz': 'rt7532'}\n",
      "Updating FACOAL40o_x with {'enz': 'rt6908'}\n",
      "Updating FACOAL60_c with {'enz': 'rt7532'}\n",
      "Updating FACOAL80_c with {'enz': 'rt7532'}\n",
      "Updating FACOAL80_x with {'enz': 'rt7761'}\n",
      "Updating FADH2t_c_m with {'enz': 'rt5791'}\n",
      "Updating FADt_c_m with {'enz': 'rt5791'}\n",
      "Updating FALPDt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating FALPDt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating FALPDt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating FALPDt_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating FALPDt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating FBA2_c with {'enz': 'rt7052'}\n",
      "Updating FBA3_c with {'enz': 'rt7052'}\n",
      "Updating FBA_c with {'enz': 'rt7052'}\n",
      "Updating FBP2_c with {'enz': 'rt7376'}\n",
      "Updating FBP_c with {'enz': 'rt7055'}\n",
      "Updating FCLT_m with {'enz': 'rt7907'}\n",
      "Updating FDH_c with {'enz': 'rt3584'}\n",
      "Updating FE2t_c_e with {'enz': 'rt0996'}\n",
      "Updating FE2t_c_m with {'gene_reaction_rule': ''}\n",
      "Updating FE2t_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating FE3t_c_e with {'enz': 'rt0996rt7477'}\n",
      "Updating FE3t_c_m with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating FE3t_c_rm with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating FECOOR_m with {'enz': 'rt0353rt1326rt2577rt3401rt3794rt5312rtmCOX1rtmCOX2rtmCOX3'}\n",
      "Updating FECOSTt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating FECOSTt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating FECPOR_m with {'enz': 'rt0434rt4793'}\n",
      "Updating FECRq9_m with {'enz': 'rt0434rt0830rt1337rt3250rt4598rt5240rt5246rt6863rt7313rt7390rtmCOB'}\n",
      "Updating FER_c with {'enz': 'rt7329'}\n",
      "Updating FKYNH_c with {'enz': 'rt6420'}\n",
      "Updating FMETTRS_m with {'enz': 'rt3531'}\n",
      "Updating FMNAT_m with {'enz': 'rt3174'}\n",
      "Updating FMNRx2_c with {'enz': 'rt8485'}\n",
      "Updating FMNRx_c with {'enz': 'rt8485'}\n",
      "Updating FOLD3_m with {'enz': 'rt6009'}\n",
      "Updating FOLR_c with {'enz': 'rt3791'}\n",
      "Updating FOLR_m with {'enz': 'rt3791'}\n",
      "Updating FOLt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating FORt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating FORt_c_m with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating FRDPt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating FRDPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating FRD_c with {'enz': 'rt3052'}\n",
      "Updating FRMDAH_c with {'enz': 'rt7438'}\n",
      "Updating FRMDH_c with {'enz': 'rt6471'}\n",
      "Updating FRTT_c with {'enz': 'rt8135'}\n",
      "Updating FRTT_l with {'enz': 'rt4281'}\n",
      "Updating FRUK_c with {'gene_reaction_rule': ''}\n",
      "Updating FRUtps_e with {'enz': 'rt7817'}\n",
      "Updating FTHFCL_c with {'enz': 'rt4194'}\n",
      "Updating FTHFL_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating FTHFL_m with {'enz': 'rt6998'}\n",
      "Updating FUMACACH_c with {'enz': 'rt3594'}\n",
      "Updating FUMACACH_m with {'enz': 'rt3874'}\n",
      "Updating FUM_c with {'enz': 'rt5633'}\n",
      "Updating FUM_m with {'enz': 'rt5633'}\n",
      "Updating FUMt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating G00006t_c_r with {'gene_reaction_rule': ''}\n",
      "Updating G00011t_g_r with {'gene_reaction_rule': ''}\n",
      "Updating G13027t_g_r with {'gene_reaction_rule': ''}\n",
      "Updating G3PAT_l with {'enz': 'rt2297'}\n",
      "Updating G3PCt_c_e with {'enz': 'rt4413'}\n",
      "Updating G3PCt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating G3PCt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating G3PD1i_m with {'enz': 'rt6208'}\n",
      "Updating G3PD1i_x with {'enz': 'rt3786'}\n",
      "Updating G3PDf_m with {'enz': 'rt2671'}\n",
      "Updating G3PIt_c_e with {'enz': 'rt4413'}\n",
      "Updating G3PIt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating G3PT_c with {'enz': 'rt0092'}\n",
      "Updating G5SADr_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating G5SADr_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating G5SDy_c with {'enz': 'rt7905'}\n",
      "Updating G6PDA_c with {'gene_reaction_rule': ''}\n",
      "Updating G6PDH2i_c with {'enz': 'rt1632'}\n",
      "Updating G6PE_c with {'enz': 'rt8255'}\n",
      "Updating GALKi_c with {'enz': 'rt5622'}\n",
      "Updating GALOX_e with {'enz': 'rt0174'}\n",
      "Updating GALT_c with {'enz': 'rt2964'}\n",
      "Updating GALUi_c with {'enz': 'rt3384'}\n",
      "Updating GALtps_e with {'enz': 'rt2336'}\n",
      "Updating GAM6Pt_c with {'gene_reaction_rule': ''}\n",
      "Updating GAPD_c with {'enz': 'rt3311'}\n",
      "Updating GARFT_c with {'enz': 'rt5227'}\n",
      "Updating GBBOX_c with {'enz': 'rt6692'}\n",
      "Updating GCALDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating GCALDt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating GDPMANNt_c_g with {'gene_reaction_rule': ''}\n",
      "Updating GDPMANt_c_g with {'enz': 'rt4177'}\n",
      "Updating GDPt_c_g with {'gene_reaction_rule': ''}\n",
      "Updating GDPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_ca2metal2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_ca2metal2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_cobalt2metal2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_cobalt2metal2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_cu2metal2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_cu2metal2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_fe2feCATION_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_fe2feCATION_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_fe2feCATION_m with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_fe2metal2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_fe2metal2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_fe2metal2_m with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_fe3feCATION_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_fe3feCATION_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_en with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_g with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_gm with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_l with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_m with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_mm with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_n with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_r with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_rm with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_v with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_vm with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hCATION1_x with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hemeAheme_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hemeAheme_m with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_hemeOheme_m with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_kCATION1_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_kCATION1_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_kCATION1_m with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_mg2metal2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_mg2metal2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_mg2mg2ormn2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_mg2mg2ormn2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_mn2metal2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_mn2metal2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_mn2mg2ormn2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_mn2mg2ormn2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_na1CATION1_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_na1CATION1_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_ni2metal2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_ni2metal2_e with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_phemeheme_m with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_shemeheme_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_zn2metal2_c with {'gene_reaction_rule': ''}\n",
      "Updating GENERIC_zn2metal2_e with {'gene_reaction_rule': ''}\n",
      "Updating GF6PTA_c with {'enz': 'rt3730rt3731'}\n",
      "Updating GGTT_c with {'gene_reaction_rule': ''}\n",
      "Updating GGTT_l with {'enz': 'rt4281'}\n",
      "Updating GHMT2r_c with {'enz': 'rt1299'}\n",
      "Updating GHMT2r_m with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating GHMT3_m with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating GK1_c with {'enz': 'rt3157'}\n",
      "Updating GLCDx_c with {'enz': 'rt5828'}\n",
      "Updating GLCGSD_c with {'enz': 'rt4271'}\n",
      "Updating GLCGSD_v with {'enz': 'rt6375'}\n",
      "Updating GLCNtps_e with {'enz': 'rt0518'}\n",
      "Updating GLCP_c with {'enz': 'rt5888'}\n",
      "Updating GLCt_c_e with {'enz': 'rt4674'}\n",
      "Updating GLCt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating GLCt_c_v with {'enz': 'rt7394'}\n",
      "Updating GLCtps_e with {'enz': 'rt3525'}\n",
      "Updating GLNS_c with {'enz': 'rt3476'}\n",
      "Updating GLNTRS_c with {'enz': 'rt5012'}\n",
      "Updating GLNt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating GLNtpa_v with {'enz': 'rt7819'}\n",
      "Updating GLNtps_v with {'enz': 'rt4375'}\n",
      "Updating GLU5K_c with {'enz': 'rt6820'}\n",
      "Updating GLUCYS_c with {'enz': 'rt3639'}\n",
      "Updating GLUDC_c with {'enz': 'rt1066'}\n",
      "Updating GLUDxi_c with {'enz': 'rt1488'}\n",
      "Updating GLUDy_c with {'enz': 'rt3880'}\n",
      "Updating GLUN_c with {'enz': 'rt5611'}\n",
      "Updating GLUN_e with {'enz': 'rt5259'}\n",
      "Updating GLUPRT_c with {'enz': 'rt6932'}\n",
      "Updating GLUR_c with {'enz': 'rt4701'}\n",
      "Updating GLUSx_c with {'enz': 'rt7345'}\n",
      "Updating GLUTCOADH_m with {'enz': 'rt3621'}\n",
      "Updating GLUTRS_c with {'enz': 'rt2385'}\n",
      "Updating GLUTRS_m with {'enz': 'rt6197'}\n",
      "Updating GLUt_c_m with {'enz': 'rt4150'}\n",
      "Updating GLUt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating GLUtps_e with {'enz': 'rt0594'}\n",
      "Updating GLUtps_v with {'enz': 'rt0398'}\n",
      "Updating GLXt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating GLXt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating GLYAT_c with {'gene_reaction_rule': ''}\n",
      "Updating GLYC3Pt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating GLYC3Pt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating GLYC3Pt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating GLYC3Pt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating GLYC3Pt_m_mm with {'gene_reaction_rule': ''}\n",
      "Updating GLYCDy_c with {'enz': 'rt3388'}\n",
      "Updating GLYCK_c with {'enz': 'rt1670'}\n",
      "Updating GLYCL_m with {'enz': 'rt1672rt1837rt4530rt6816'}\n",
      "Updating GLYCOGENt_c_v with {'gene_reaction_rule': ''}\n",
      "Updating GLYCt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating GLYCt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating GLYGS_c with {'enz': 'rt1228rt5634'}\n",
      "Updating GLYK_c with {'enz': 'rt2654'}\n",
      "Updating GLYOX_c with {'enz': 'rt4673'}\n",
      "Updating GLYOX_m with {'enz': 'rt5195'}\n",
      "Updating GLYTRS_c with {'enz': 'rt7136'}\n",
      "Updating GLYt_c_m with {'enz': 'rt0066'}\n",
      "Updating GMAND_c with {'enz': 'rt2621'}\n",
      "Updating GMPS2_c with {'enz': 'rt7388'}\n",
      "Updating GND_c with {'enz': 'rt6799'}\n",
      "Updating GNK_c with {'enz': 'rt8291'}\n",
      "Updating GPIA10a_r with {'enz': 'rt0563'}\n",
      "Updating GPIA1_c with {'enz': 'rt3458rt5371rt5395rt6785'}\n",
      "Updating GPIA2_c with {'enz': 'rt1519'}\n",
      "Updating GPIA3_c_r with {'enz': 'rt3333'}\n",
      "Updating GPIA4_r with {'enz': 'rt3077'}\n",
      "Updating GPIA5_r with {'enz': 'rt1537'}\n",
      "Updating GPIA6_r with {'enz': 'rt3077'}\n",
      "Updating GPIA7_r with {'enz': 'rt4593'}\n",
      "Updating GPIA8a_r with {'enz': 'rt0237'}\n",
      "Updating GPIA8b_r with {'enz': 'rt4069'}\n",
      "Updating GRTT_c with {'enz': 'rt4576'}\n",
      "Updating GSNK_c with {'gene_reaction_rule': ''}\n",
      "Updating GSNt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating GSNt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating GTHOXt_c with {'gene_reaction_rule': ''}\n",
      "Updating GTHO_c with {'enz': 'rt7114rt8483'}\n",
      "Updating GTHO_m with {'enz': 'rt0882rt7114'}\n",
      "Updating GTHP_c with {'enz': 'rt2565'}\n",
      "Updating GTHP_m with {'enz': 'rt0422'}\n",
      "Updating GTHRDt_c_e with {'enz': 'rt5171'}\n",
      "Updating GTHRDt_c_v with {'enz': 'rt4896'}\n",
      "Updating GTHRDtabc_v with {'enz': 'rt4896'}\n",
      "Updating GTHS_c with {'enz': 'rt1561'}\n",
      "Updating GTMLT_c with {'enz': 'rt1470'}\n",
      "Updating GTPCII2_c with {'enz': 'rt4183'}\n",
      "Updating GTPCI_c with {'enz': 'rt1964'}\n",
      "Updating GTPDPDP_c with {'enz': 'rt1567'}\n",
      "Updating GTPGDPtp_m with {'enz': 'rt5389'}\n",
      "Updating GUAD_1_c with {'enz': 'rt1340'}\n",
      "Updating GUAPRT_c with {'enz': 'rt4077'}\n",
      "Updating GUAt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating GUAtps_e with {'enz': 'rt6190'}\n",
      "Updating H2O2t_c_n with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_e with {'enz': 'rt0646'}\n",
      "Updating H2Ot_c_en with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_g with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_gm with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_l with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_m with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_n with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_r with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_v with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_vm with {'gene_reaction_rule': ''}\n",
      "Updating H2Ot_c_x with {'gene_reaction_rule': ''}\n",
      "Updating HACD100i_m with {'enz': 'rt2835'}\n",
      "Updating HACD100i_x with {'enz': 'rt7916'}\n",
      "Updating HACD101ei_x with {'enz': 'rt7916'}\n",
      "Updating HACD120i_m with {'enz': 'rt2835'}\n",
      "Updating HACD120i_x with {'enz': 'rt7916'}\n",
      "Updating HACD121ai_m with {'enz': 'rt2835'}\n",
      "Updating HACD121ai_x with {'enz': 'rt7916'}\n",
      "Updating HACD121di_x with {'enz': 'rt7916'}\n",
      "Updating HACD122ei_x with {'enz': 'rt7916'}\n",
      "Updating HACD140i_m with {'enz': 'rt2835'}\n",
      "Updating HACD140i_x with {'enz': 'rt7916'}\n",
      "Updating HACD141ai_m with {'enz': 'rt2835'}\n",
      "Updating HACD141ai_x with {'enz': 'rt7916'}\n",
      "Updating HACD141bi_m with {'enz': 'rt2835'}\n",
      "Updating HACD141bi_x with {'enz': 'rt7916'}\n",
      "Updating HACD142i_x with {'enz': 'rt7916'}\n",
      "Updating HACD143i_x with {'enz': 'rt7916'}\n",
      "Updating HACD160i_m with {'enz': 'rt2835'}\n",
      "Updating HACD160i_x with {'enz': 'rt7916'}\n",
      "Updating HACD161ai_m with {'enz': 'rt2835'}\n",
      "Updating HACD161ai_x with {'enz': 'rt7916'}\n",
      "Updating HACD161bi_m with {'enz': 'rt2835'}\n",
      "Updating HACD161bi_x with {'enz': 'rt7916'}\n",
      "Updating HACD162i_x with {'enz': 'rt7916'}\n",
      "Updating HACD163i_x with {'enz': 'rt7916'}\n",
      "Updating HACD180i_m with {'enz': 'rt2835'}\n",
      "Updating HACD180i_x with {'enz': 'rt7916'}\n",
      "Updating HACD181ai_m with {'enz': 'rt2835'}\n",
      "Updating HACD181ai_x with {'enz': 'rt7916'}\n",
      "Updating HACD182i_x with {'enz': 'rt7916'}\n",
      "Updating HACD183i_x with {'enz': 'rt7916'}\n",
      "Updating HACD200i_x with {'enz': 'rt7916'}\n",
      "Updating HACD220i_x with {'enz': 'rt7916'}\n",
      "Updating HACD240i_x with {'enz': 'rt7916'}\n",
      "Updating HACD260i_x with {'enz': 'rt7916'}\n",
      "Updating HACD40i_m with {'enz': 'rt2835'}\n",
      "Updating HACD40i_x with {'enz': 'rt7916'}\n",
      "Updating HACD40y_c with {'enz': 'rt6812'}\n",
      "Updating HACD60i_m with {'enz': 'rt2835'}\n",
      "Updating HACD60i_x with {'enz': 'rt7916'}\n",
      "Updating HACD80i_m with {'enz': 'rt2835'}\n",
      "Updating HACD80i_x with {'enz': 'rt7916'}\n",
      "Updating HACNH_m with {'enz': 'rt7648'}\n",
      "Updating HCITS_m with {'enz': 'rt6488'}\n",
      "Updating HCITS_n with {'enz': 'rt6488'}\n",
      "Updating HCO3E_c with {'enz': 'rt4213'}\n",
      "Updating HCO3E_e with {'gene_reaction_rule': ''}\n",
      "Updating HCO3E_m with {'gene_reaction_rule': ''}\n",
      "Updating HCO3E_n with {'gene_reaction_rule': ''}\n",
      "Updating HCO3t_c_n with {'gene_reaction_rule': ''}\n",
      "Updating HCYSMT_c with {'enz': 'rt7391'}\n",
      "Updating HDCAt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating HDCAt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating HDCAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating HDCAt_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating HDCAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating HDCEAt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating HDCEAt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating HDCEAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating HDCEAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating HDCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating HDCOAt_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating HDCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating HEMEAt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating HEMEOS_m with {'enz': 'rt2735'}\n",
      "Updating HETZK_c with {'enz': 'rt7845'}\n",
      "Updating HEX1_c with {'enz': 'rt1896'}\n",
      "Updating HEX7_c with {'enz': 'rt3614'}\n",
      "Updating HEX7b_c with {'enz': 'rt3614'}\n",
      "Updating HGENDO_c with {'enz': 'rt7338'}\n",
      "Updating HIBH_m with {'enz': 'rt6850'}\n",
      "Updating HIBTD_c with {'enz': 'rt5326'}\n",
      "Updating HICITD_m with {'enz': 'rt2060'}\n",
      "Updating HISTD_c with {'enz': 'rt3278'}\n",
      "Updating HISTP_c with {'enz': 'rt1838'}\n",
      "Updating HISTRS_c with {'enz': 'rt0207'}\n",
      "Updating HISTRS_m with {'enz': 'rt0207'}\n",
      "Updating HISt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating HIStpa_v with {'enz': 'rt1949'}\n",
      "Updating HKYNH_c with {'enz': 'rt0357'}\n",
      "Updating HLYSMT_m with {'enz': 'rt2181'}\n",
      "Updating HMGCOAR_c with {'enz': 'rt1206'}\n",
      "Updating HMGCOAS_c with {'enz': 'rt3754'}\n",
      "Updating HMGCOAS_m with {'enz': 'rt3754'}\n",
      "Updating HMGCOAt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating HMPK1_c with {'enz': 'rt4410'}\n",
      "Updating HOMOX_m with {'enz': 'rt1860rt7217rt8407'}\n",
      "Updating HOMt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating HPPK2_m with {'enz': 'rt6009'}\n",
      "Updating HPRORy_c with {'enz': 'rt5064'}\n",
      "Updating HQLOR_c with {'enz': 'rt3938'}\n",
      "Updating HSDx_c with {'enz': 'rt3712'}\n",
      "Updating HSDy_c with {'enz': 'rt3712'}\n",
      "Updating HSERTA_c with {'enz': 'rt4145'}\n",
      "Updating HSK_c with {'enz': 'rt0283'}\n",
      "Updating HSTPT_c with {'enz': 'rt4862'}\n",
      "Updating HXAND_c with {'enz': 'rt7594'}\n",
      "Updating HXANt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating HXAt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating HXAt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating HXCAt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating HXCAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating HXCAt_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating HXCAt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating HXCAt_r_rm with {'gene_reaction_rule': ''}\n",
      "Updating HXCCOAt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating HXCCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating HXCCOAt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating HXCCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating HYPOE_c with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_e with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_en with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_g with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_gm with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_l with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_m with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_n with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_r with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_v with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_vm with {'gene_reaction_rule': ''}\n",
      "Updating Ht_c_x with {'gene_reaction_rule': ''}\n",
      "Updating Ht_m_mm with {'gene_reaction_rule': ''}\n",
      "Updating IAMACt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating IAMOHt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating IAMOHt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating IBTALDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating IBTALDt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating IBUTACt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating IBUTOHt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating IBUTOHt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating ICDHx_m with {'enz': 'rt3313rt3314'}\n",
      "Updating ICDHyi_m with {'enz': 'rt2761'}\n",
      "Updating ICL_1_c with {'enz': 'rt5654'}\n",
      "Updating ICOSCOAt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating ICOSCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating ID3ACALDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating ID3ACALDt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating IG3PS_c with {'enz': 'rt0716'}\n",
      "Updating IGPDH_c with {'enz': 'rt0339'}\n",
      "Updating IGPS_c with {'enz': 'rt8196'}\n",
      "Updating ILETA_c with {'enz': 'rt6485'}\n",
      "Updating ILETA_m with {'enz': 'rt6242'}\n",
      "Updating ILETRS_c with {'enz': 'rt5078'}\n",
      "Updating ILETRS_m with {'enz': 'rt7188'}\n",
      "Updating ILEt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating ILEt_c_v with {'enz': 'rt6501'}\n",
      "Updating ILEtpa_v with {'enz': 'rt7819'}\n",
      "Updating ILEtps_v with {'enz': 'rt4375'}\n",
      "Updating IMPC_c with {'enz': 'rt4053'}\n",
      "Updating IMPD_c with {'enz': 'rt0027'}\n",
      "Updating IND3ACt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating IND3ETHt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating IND3ETHt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating INDPYRD_c with {'enz': 'rt7423'}\n",
      "Updating INO3GALT_c with {'enz': 'rt5966'}\n",
      "Updating INOSHP3H_v with {'enz': 'rt2517'}\n",
      "Updating INOSHPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating INOSTt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating INOSTtps_e with {'enz': 'rt0734'}\n",
      "Updating INSK_c with {'gene_reaction_rule': ''}\n",
      "Updating INSt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating IPCS_g with {'enz': 'rt4559'}\n",
      "Updating IPCt_g_r with {'gene_reaction_rule': ''}\n",
      "Updating IPDDI_c with {'enz': 'rt7835'}\n",
      "Updating IPDPt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating IPDPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating IPK2_n with {'enz': 'rt4892'}\n",
      "Updating IPK3_n with {'enz': 'rt4892'}\n",
      "Updating IPK4_n with {'enz': 'rt4892'}\n",
      "Updating IPK5_n with {'enz': 'rt4892'}\n",
      "Updating IPMD_c with {'enz': 'rt5526'}\n",
      "Updating IPPMIa_c with {'enz': 'rt6546'}\n",
      "Updating IPPMIb_c with {'enz': 'rt6546'}\n",
      "Updating IPPS_c with {'enz': 'rt7120'}\n",
      "Updating IPPYP2_c with {'enz': 'rt1973'}\n",
      "Updating IPPYP3_c with {'enz': 'rt1973'}\n",
      "Updating IPPYP4_c with {'enz': 'rt1395'}\n",
      "Updating IPT_g with {'enz': 'rt4804'}\n",
      "Updating ITCOAL_m with {'enz': 'rt7599rt7776'}\n",
      "Updating KARA1i_m with {'enz': 'rt0808'}\n",
      "Updating KARA2i_m with {'enz': 'rt0808'}\n",
      "Updating KYN3OX_c with {'enz': 'rt0899'}\n",
      "Updating KYN_c with {'enz': 'rt0357'}\n",
      "Updating Kt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating Kt_c_m with {'enz': 'rt7131'}\n",
      "Updating Ktps_e with {'enz': 'rt2101'}\n",
      "Updating LACDtps_c with {'enz': 'rt1816'}\n",
      "Updating LACLtps_c with {'enz': 'rt1816'}\n",
      "Updating LACPYRt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating LACt_c_m with {'enz': 'rt1816'}\n",
      "Updating LACtps_m with {'gene_reaction_rule': ''}\n",
      "Updating LALDO3_c with {'enz': 'rt3783'}\n",
      "Updating LANOSTt_c_e with {'enz': 'rt6543'}\n",
      "Updating LANOSTt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating LANOSTt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating LCADi_c with {'enz': 'rt3783'}\n",
      "Updating LCYSTAT_c with {'enz': 'rt1784'}\n",
      "Updating LDH_L_x with {'enz': 'rt5258'}\n",
      "Updating LEUTA_c with {'enz': 'rt6485'}\n",
      "Updating LEUTA_m with {'enz': 'rt6242'}\n",
      "Updating LEUTRS_c with {'enz': 'rt1776'}\n",
      "Updating LEUt_c_v with {'enz': 'rt6501'}\n",
      "Updating LEUtpa_v with {'enz': 'rt7819'}\n",
      "Updating LEUtps_v with {'enz': 'rt4375'}\n",
      "Updating LGTHL_c with {'enz': 'rt8152'}\n",
      "Updating LINOCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating LINOCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating LINOEAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating LINOLENt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating LINOLNCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating LINOLNCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating LIPOCT_m with {'enz': 'rt8488'}\n",
      "Updating LLFC2O_c with {'enz': 'rt0745'}\n",
      "Updating LNS14DMy_c with {'enz': 'rt4475rt4918'}\n",
      "Updating LNSTLS_c with {'enz': 'rt0872'}\n",
      "Updating LPAP_vm with {'enz': 'rt4719'}\n",
      "Updating LPCAT_mm with {'enz': 'rt8106'}\n",
      "Updating LPIAT_rm with {'enz': 'rt1153'}\n",
      "Updating LSERDHi_c with {'enz': 'rt1844'}\n",
      "Updating LTA4H2_c with {'enz': 'rt2643'}\n",
      "Updating LTA4H2_n with {'enz': 'rt2643'}\n",
      "Updating LTA4H3_c with {'enz': 'rt2643'}\n",
      "Updating LTA4H3_n with {'enz': 'rt2643'}\n",
      "Updating LTA4H_c with {'enz': 'rt2643'}\n",
      "Updating LTA4H_n with {'enz': 'rt2643'}\n",
      "Updating LYSN6MOX_c with {'enz': 'rt2880'}\n",
      "Updating LYSTRS_c with {'enz': 'rt3577'}\n",
      "Updating LYSt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating LYStpa_v with {'enz': 'rt1949'}\n",
      "Updating LYStps_e with {'enz': 'rt5861'}\n",
      "Updating M1PS_c with {'enz': 'rt1768'}\n",
      "Updating MACACI_c with {'enz': 'rt0063'}\n",
      "Updating MAGL_l with {'enz': 'rt1360'}\n",
      "Updating MAGL_m with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating MALCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating MALOAAta_x with {'gene_reaction_rule': ''}\n",
      "Updating MALS_c with {'enz': 'rt1089'}\n",
      "Updating MALT_c with {'enz': 'rt0767'}\n",
      "Updating MALTtps_e with {'enz': 'rt0769'}\n",
      "Updating MALt2_m with {'enz': 'rt5142'}\n",
      "Updating MALt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating MAN1PT_c with {'enz': 'rt6588'}\n",
      "Updating MAN6PI_c with {'enz': 'rt0239'}\n",
      "Updating MANNAN14H_c with {'enz': 'rt0045'}\n",
      "Updating MANNANH_c with {'enz': 'rt2404'}\n",
      "Updating MANNANH_e with {'enz': 'rt2418'}\n",
      "Updating MANNANt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating MANt_c_g with {'gene_reaction_rule': ''}\n",
      "Updating MANtps_e with {'enz': 'rt3556'}\n",
      "Updating MASCBRx_c with {'enz': 'rt1779'}\n",
      "Updating MCCC_m with {'enz': 'rt4499rt8036'}\n",
      "Updating MCITL2_m with {'enz': 'rt5794'}\n",
      "Updating MCITS_m with {'enz': 'rt2960'}\n",
      "Updating MCOATA_m with {'enz': 'rt5893rt8174'}\n",
      "Updating MDH_c with {'enz': 'rt2246'}\n",
      "Updating MDH_m with {'enz': 'rt2810'}\n",
      "Updating MDLDH_x with {'enz': 'rt5416'}\n",
      "Updating MDRPD_c with {'enz': 'rt7461'}\n",
      "Updating ME1_m with {'enz': 'rt5549'}\n",
      "Updating ME2_c with {'enz': 'rt4393'}\n",
      "Updating ME2_m with {'enz': 'rt5549'}\n",
      "Updating MERLACOR_c with {'enz': 'rt5258'}\n",
      "Updating METAT_c with {'enz': 'rt5403'}\n",
      "Updating METB1_c with {'enz': 'rt8374'}\n",
      "Updating METS_c with {'enz': 'rt1457'}\n",
      "Updating METTRS_c with {'enz': 'rt4772'}\n",
      "Updating METTRS_m with {'enz': 'rt7150'}\n",
      "Updating METt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating MEVK1_c with {'enz': 'rt0390'}\n",
      "Updating MEVK2_c with {'enz': 'rt0390'}\n",
      "Updating MEVK3_c with {'enz': 'rt0390'}\n",
      "Updating MEVK4_c with {'enz': 'rt0390'}\n",
      "Updating MEVt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating MG2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating MGCH_m with {'enz': 'rt7760'}\n",
      "Updating MGSA_c with {'gene_reaction_rule': ''}\n",
      "Updating MHPGLUT_c with {'enz': 'rt1457'}\n",
      "Updating MI13456Pt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating MI134PPH_c with {'enz': 'rt6885'}\n",
      "Updating MI1PP_c with {'enz': 'rt5340'}\n",
      "Updating MI1Pt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating MIP2Ct_g_r with {'gene_reaction_rule': ''}\n",
      "Updating MIPCS_g with {'enz': 'rt8085'}\n",
      "Updating MIPCt_g_r with {'gene_reaction_rule': ''}\n",
      "Updating MN2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating MOBDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating MPYRCNST_c with {'enz': 'rt7838'}\n",
      "Updating MTAP_c with {'enz': 'rt0004'}\n",
      "Updating MTHFC_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating MTHFC_m with {'enz': 'rt6998'}\n",
      "Updating MTHFD2i_c with {'enz': 'rt4257'}\n",
      "Updating MTHFD_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating MTHFD_m with {'enz': 'rt6998'}\n",
      "Updating MTHFR3_c with {'enz': 'rt0876rt7115'}\n",
      "Updating MTMOHT_m with {'enz': 'rt0075'}\n",
      "Updating MTRI_c with {'enz': 'rt5017'}\n",
      "Updating MUCOCI_c with {'enz': 'rt5684'}\n",
      "Updating MUCOCI_e with {'enz': 'rt4254'}\n",
      "Updating NACt_c_e with {'enz': 'rt3853'}\n",
      "Updating NADDP_x with {'enz': 'rt4066'}\n",
      "Updating NADHK1_c with {'enz': 'rt1814'}\n",
      "Updating NADHK1_m with {'enz': 'rt1814'}\n",
      "Updating NADHcplxI_c_m with {'enz': 'rt0331rt0449rt3334rt4649rt4846rt6050rt7496rt7605rtmNAD1rtmNAD2rtmNAD3rtmNAD4rtmNAD4LrtmNAD5rtmNAD6'}\n",
      "Updating NADHq9_c with {'enz': 'rt0604'}\n",
      "Updating NADHt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating NADK_c with {'enz': 'rt1814'}\n",
      "Updating NADK_m with {'enz': 'rt1814'}\n",
      "Updating NADN_n with {'enz': 'rt4419'}\n",
      "Updating NADPHt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating NADPHt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating NADPHt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating NADPHt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating NADPt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating NADPt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating NADPt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating NADPt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating NADPt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating NADS2_c with {'enz': 'rt0915'}\n",
      "Updating NADS2_n with {'enz': 'rt0915'}\n",
      "Updating NADt1_m with {'enz': 'rt2593'}\n",
      "Updating NADt2_m with {'enz': 'rt2593'}\n",
      "Updating NADt3_m with {'enz': 'rt2593'}\n",
      "Updating NADt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating NADt_c_m with {'enz': 'rt5934'}\n",
      "Updating NADt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating NAMNPP_c with {'enz': 'rt1924'}\n",
      "Updating NAMNPP_m with {'enz': 'rt1924'}\n",
      "Updating NAtpa_c with {'enz': 'rt2444'}\n",
      "Updating NCAMt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating NDP1_c with {'enz': 'rt0602'}\n",
      "Updating NDP3_c with {'enz': 'rt0602'}\n",
      "Updating NDP3_g with {'enz': 'rt1437'}\n",
      "Updating NDP4_c with {'enz': 'rt0602'}\n",
      "Updating NDP5_c with {'enz': 'rt0602'}\n",
      "Updating NDP7_c with {'enz': 'rt0602'}\n",
      "Updating NDPDA1_c with {'gene_reaction_rule': ''}\n",
      "Updating NDPDA2_c with {'gene_reaction_rule': ''}\n",
      "Updating NDPK10_c with {'enz': 'rt7311'}\n",
      "Updating NDPK1_c with {'enz': 'rt7311'}\n",
      "Updating NDPK2_c with {'enz': 'rt7311'}\n",
      "Updating NDPK3_c with {'enz': 'rt7311'}\n",
      "Updating NDPK4_c with {'enz': 'rt7311'}\n",
      "Updating NDPK5_c with {'enz': 'rt7311'}\n",
      "Updating NDPK6_c with {'enz': 'rt7311'}\n",
      "Updating NDPK7_c with {'enz': 'rt7311'}\n",
      "Updating NDPK8_c with {'enz': 'rt7311'}\n",
      "Updating NDPK9_c with {'enz': 'rt7311'}\n",
      "Updating NFORTYR2t_c_e with {'enz': 'rt2566'}\n",
      "Updating NFTYROX_c with {'enz': 'rt0820'}\n",
      "Updating NGLYCANS10_r with {'enz': 'rt2904'}\n",
      "Updating NGLYCANS11_r with {'enz': 'rt0671'}\n",
      "Updating NGLYCANS12_r with {'enz': 'rt2573'}\n",
      "Updating NGLYCANS13_r with {'enz': 'rt6611'}\n",
      "Updating NGLYCANS14_r with {'enz': 'rt2972'}\n",
      "Updating NGLYCANS15_r with {'enz': 'rt3790'}\n",
      "Updating NGLYCANS16_r with {'enz': 'rt1613'}\n",
      "Updating NGLYCANS17_r with {'enz': 'rt1802'}\n",
      "Updating NGLYCANS18_r with {'enz': 'rt1802'}\n",
      "Updating NGLYCANS19_g with {'enz': 'rt8131'}\n",
      "Updating NGLYCANS1_c with {'enz': 'rt2834'}\n",
      "Updating NGLYCANS2_c with {'enz': 'rt4980'}\n",
      "Updating NGLYCANS3_c with {'enz': 'rt6298'}\n",
      "Updating NGLYCANS4_c with {'enz': 'rt1727'}\n",
      "Updating NGLYCANS5_c with {'enz': 'rt1727'}\n",
      "Updating NGLYCANS6_c with {'enz': 'rt1577'}\n",
      "Updating NGLYCANS7_c with {'enz': 'rt1577'}\n",
      "Updating NGLYCANS8_r with {'enz': 'rt5214'}\n",
      "Updating NGLYCANS9_r with {'enz': 'rt0671'}\n",
      "Updating NH4t_c_e with {'enz': 'rt4558'}\n",
      "Updating NH4t_c_m with {'gene_reaction_rule': ''}\n",
      "Updating NI2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating NITROPP_c with {'enz': 'rt3422'}\n",
      "Updating NITRx_c with {'enz': 'rt6640'}\n",
      "Updating NMNt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating NMNt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating NMNt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating NNAM_c with {'enz': 'rt3293'}\n",
      "Updating NNATi_c with {'enz': 'rt2062'}\n",
      "Updating NNDPR_c with {'enz': 'rt1963'}\n",
      "Updating NNDPR_m with {'enz': 'rt1963'}\n",
      "Updating NNMT_c with {'enz': 'rt5672'}\n",
      "Updating NO3t_c_e with {'enz': 'rt4068'}\n",
      "Updating NOX_c with {'enz': 'rt4914'}\n",
      "Updating NTD10pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD1pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD2pp_c with {'enz': 'rt1627'}\n",
      "Updating NTD3pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD4pp_c with {'enz': 'rt1627'}\n",
      "Updating NTD5_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD6pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD7pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD8_c with {'gene_reaction_rule': ''}\n",
      "Updating NTP10_c with {'enz': 'rt0602'}\n",
      "Updating NTP3_c with {'enz': 'rt7922'}\n",
      "Updating NTP4_c with {'enz': 'rt0602'}\n",
      "Updating NTP5_c with {'enz': 'rt0602'}\n",
      "Updating NTP7_c with {'enz': 'rt0602'}\n",
      "Updating NTPDA2_c with {'gene_reaction_rule': ''}\n",
      "Updating O2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating O2t_c_m with {'gene_reaction_rule': ''}\n",
      "Updating O2t_c_r with {'gene_reaction_rule': ''}\n",
      "Updating O2t_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating O2t_c_x with {'gene_reaction_rule': ''}\n",
      "Updating OAADC_c with {'enz': 'rt7226'}\n",
      "Updating OAAt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating OAAtps_m with {'enz': 'rt0643'}\n",
      "Updating OAR1_c with {'enz': 'rt2815'}\n",
      "Updating OAR2_c with {'enz': 'rt2815'}\n",
      "Updating OAR3_m with {'enz': 'rt0607'}\n",
      "Updating OCBT_c with {'enz': 'rt4934'}\n",
      "Updating OCDCAt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating OCDCAt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating OCDCAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating OCDCAt_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating OCDCAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating OCDCEAt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating OCDCEAt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating OCDCEAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating OCDCEAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating OCOAT1_m with {'enz': 'rt4722'}\n",
      "Updating OCTAt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating OCTAt_c_x with {'enz': 'rt5693'}\n",
      "Updating ODECOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating ODECOAt_c_mm with {'gene_reaction_rule': ''}\n",
      "Updating ODECOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating OGLYCANS1_r with {'enz': 'rt3753'}\n",
      "Updating OGLYCANS2_g with {'enz': 'rt1384'}\n",
      "Updating OGLYCANS3_g with {'enz': 'rt2093'}\n",
      "Updating OGLYCANS4_g with {'enz': 'rt4112'}\n",
      "Updating OGLYCANS5_g with {'enz': 'rt4112'}\n",
      "Updating OGLYCANS_B3GAT1_r with {'enz': 'rt2574'}\n",
      "Updating OMCDC_m with {'enz': 'rt6242'}\n",
      "Updating OMPDC_c with {'enz': 'rt3750'}\n",
      "Updating OPAH_c with {'enz': 'rt7643'}\n",
      "Updating ORNCD_m with {'enz': 'rt6574'}\n",
      "Updating ORNDC_c with {'enz': 'rt4267'}\n",
      "Updating ORNN5MOX_c with {'enz': 'rt2880'}\n",
      "Updating ORNTACi_m with {'enz': 'rt7840'}\n",
      "Updating ORNTA_c with {'enz': 'rt6698'}\n",
      "Updating ORNtpa_m with {'enz': 'rt4951'}\n",
      "Updating ORNtps_e with {'enz': 'rt1689'}\n",
      "Updating ORPT_c with {'enz': 'rt2348'}\n",
      "Updating P5CD2_m with {'enz': 'rt0201'}\n",
      "Updating P5CR_c with {'enz': 'rt5064'}\n",
      "Updating PACALDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating PACALDt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PACCOAE_c with {'enz': 'rt6217'}\n",
      "Updating PACCOAE_x with {'enz': 'rt5558'}\n",
      "Updating PACt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating PAFH_c with {'enz': 'rt3915'}\n",
      "Updating PAIL35Pt_c_vm with {'gene_reaction_rule': ''}\n",
      "Updating PAIL35Pt_en_vm with {'gene_reaction_rule': ''}\n",
      "Updating PAIL3K_vm with {'enz': 'rt0344rt0641'}\n",
      "Updating PAIL3Pt_gm_vm with {'gene_reaction_rule': ''}\n",
      "Updating PAIL3Pt_rm_vm with {'gene_reaction_rule': ''}\n",
      "Updating PAIL45Pt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating PAIL45Pt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating PAIL45Pt_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAIL4K_en with {'enz': 'rt6791'}\n",
      "Updating PAIL4K_gm with {'enz': 'rt1414rt4220'}\n",
      "Updating PAIL4K_n with {'enz': 'rt1414'}\n",
      "Updating PAIL4Pt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating PAIL4Pt_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAILS_rm with {'enz': 'rt6753'}\n",
      "Updating PAILt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAILt_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAILt_gm_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAILt_n_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAILt_rm_vm with {'gene_reaction_rule': ''}\n",
      "Updating PAN4Pt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PAPSR_c with {'enz': 'rt3373rt6971'}\n",
      "Updating PAP_rm with {'enz': 'rt4117'}\n",
      "Updating PAP_vm with {'enz': 'rt4719'}\n",
      "Updating PAPt_c with {'gene_reaction_rule': ''}\n",
      "Updating PAPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PAt_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAt_gm_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAt_mm_rm with {'gene_reaction_rule': ''}\n",
      "Updating PAt_rm_vm with {'gene_reaction_rule': ''}\n",
      "Updating PBAL_c with {'enz': 'rt2107'}\n",
      "Updating PCLAD_c with {'enz': 'rt4548'}\n",
      "Updating PCOATA140_c with {'gene_reaction_rule': ''}\n",
      "Updating PCOATA160_c with {'enz': 'rt0302'}\n",
      "Updating PCOATA180_c with {'enz': 'rt0302'}\n",
      "Updating PC_c with {'enz': 'rt7325'}\n",
      "Updating PCt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating PCt_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating PCt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating PCt_mm_rm with {'gene_reaction_rule': ''}\n",
      "Updating PDAGATpc_rm with {'enz': 'rt8109'}\n",
      "Updating PDAGATpe_rm with {'enz': 'rt8109'}\n",
      "Updating PDE1_c with {'enz': 'rt5322'}\n",
      "Updating PDE2_c with {'enz': 'rt5322'}\n",
      "Updating PDE4_c with {'enz': 'rt3419'}\n",
      "Updating PDE5_c with {'enz': 'rt3419'}\n",
      "Updating PDH_m with {'enz': 'rt1672rt5262rt5354rt5580rt5758'}\n",
      "Updating PDMEMT_rm with {'enz': 'rt4380'}\n",
      "Updating PDX5POi_c with {'enz': 'rt2312'}\n",
      "Updating PEMT_rm with {'enz': 'rt2279'}\n",
      "Updating PENDPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PEt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating PEt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating PEt_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating PEt_gm_rm with {'gene_reaction_rule': ''}\n",
      "Updating PEt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating PEt_mm_rm with {'gene_reaction_rule': ''}\n",
      "Updating PEt_rm_vm with {'gene_reaction_rule': ''}\n",
      "Updating PEtabc_m with {'enz': 'rt4312'}\n",
      "Updating PFK_c with {'enz': 'rt0499'}\n",
      "Updating PGAMT_c with {'enz': 'rt4597'}\n",
      "Updating PGCD_c with {'enz': 'rt3683'}\n",
      "Updating PGCPH_c with {'enz': 'rt4502'}\n",
      "Updating PGI_c with {'enz': 'rt1221'}\n",
      "Updating PGK_c with {'enz': 'rt7353'}\n",
      "Updating PGLYCP_c with {'enz': 'rt3422'}\n",
      "Updating PGL_c with {'enz': 'rt6131'}\n",
      "Updating PGMT_c with {'enz': 'rt1591'}\n",
      "Updating PGM_c with {'enz': 'rt7057'}\n",
      "Updating PGPP_mm with {'enz': 'rt2532'}\n",
      "Updating PGPS_mm with {'enz': 'rt4228'}\n",
      "Updating PGS_x with {'enz': 'rt5040'}\n",
      "Updating PGt_c_m with {'enz': 'rt4312'}\n",
      "Updating PHACRDC1_c with {'enz': 'rt4099'}\n",
      "Updating PHACRDC2_c with {'enz': 'rt4099'}\n",
      "Updating PHACRDC3_c with {'enz': 'rt4099'}\n",
      "Updating PHACRDC4_c with {'enz': 'rt4099'}\n",
      "Updating PHCD_m with {'enz': 'rt0201'}\n",
      "Updating PHCHGS_m with {'gene_reaction_rule': ''}\n",
      "Updating PHEACt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating PHEN2MOXy_c with {'enz': 'rt4454'}\n",
      "Updating PHETA1_c with {'enz': 'rt7697'}\n",
      "Updating PHETA1_m with {'enz': 'rt5562'}\n",
      "Updating PHETRS_c with {'enz': 'rt4102rt4881'}\n",
      "Updating PHETRS_m with {'enz': 'rt5030'}\n",
      "Updating PHEt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PHEtps_e with {'enz': 'rt5861'}\n",
      "Updating PHYTOED1_c with {'enz': 'rt0217'}\n",
      "Updating PHYTOED2_c with {'enz': 'rt0217'}\n",
      "Updating PHYTOED3_c with {'enz': 'rt0217'}\n",
      "Updating PHYTOED4_c with {'enz': 'rt0217'}\n",
      "Updating PHYTOED5_c with {'enz': 'rt0217'}\n",
      "Updating PHYTOES_c with {'enz': 'rt2239'}\n",
      "Updating PI35P5P_c with {'enz': 'rt7001'}\n",
      "Updating PI35P5P_en with {'enz': 'rt7049'}\n",
      "Updating PI35P5P_vm with {'enz': 'rt5535rt8251'}\n",
      "Updating PI3P5K_vm with {'enz': 'rt5771'}\n",
      "Updating PI3PP_c with {'enz': 'rt6512'}\n",
      "Updating PI3PP_en with {'enz': 'rt7049'}\n",
      "Updating PI3PP_gm with {'enz': 'rt3013'}\n",
      "Updating PI3PP_rm with {'enz': 'rt3013'}\n",
      "Updating PI45P5P_c with {'enz': 'rt6972'}\n",
      "Updating PI45P5P_en with {'enz': 'rt7177'}\n",
      "Updating PI4P5K_en with {'enz': 'rt3510'}\n",
      "Updating PI4P5K_n with {'enz': 'rt3510'}\n",
      "Updating PI4PP_c with {'enz': 'rt7049'}\n",
      "Updating PI4PP_en with {'enz': 'rt7049'}\n",
      "Updating PI4PP_gm with {'enz': 'rt3013'}\n",
      "Updating PI4PP_rm with {'enz': 'rt3013'}\n",
      "Updating PIt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating PIt_c_gm with {'gene_reaction_rule': ''}\n",
      "Updating PIt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating PIt_c_v with {'enz': 'rt3050'}\n",
      "Updating PIt_c_vm with {'enz': 'rt3050'}\n",
      "Updating PIt_m_mm with {'gene_reaction_rule': ''}\n",
      "Updating PIt_r_rm with {'gene_reaction_rule': ''}\n",
      "Updating PItps_e with {'enz': 'rt3318'}\n",
      "Updating PItps_m with {'enz': 'rt7506'}\n",
      "Updating PKETF_c with {'enz': 'rt5014'}\n",
      "Updating PKETX_c with {'enz': 'rt5014'}\n",
      "Updating PLAA1lpc_en with {'enz': 'rt4017'}\n",
      "Updating PLAA1lpc_mm with {'enz': 'rt5790'}\n",
      "Updating PLAA1lpc_rm with {'enz': 'rt5941'}\n",
      "Updating PLAA1lpe_en with {'enz': 'rt4017'}\n",
      "Updating PLAA1lpe_mm with {'enz': 'rt5790'}\n",
      "Updating PLAA2pc_en with {'enz': 'rt4017'}\n",
      "Updating PLAA2pc_mm with {'enz': 'rt4445'}\n",
      "Updating PLAA2pc_rm with {'enz': 'rt5941'}\n",
      "Updating PLAA2pe_en with {'enz': 'rt4017'}\n",
      "Updating PLAA2pe_mm with {'enz': 'rt4445'}\n",
      "Updating PLACipc_m with {'enz': 'rt7489'}\n",
      "Updating PLACipc_r with {'enz': 'rt7489'}\n",
      "Updating PLACmip2c_m with {'enz': 'rt7489'}\n",
      "Updating PLACmip2c_r with {'enz': 'rt7489'}\n",
      "Updating PLACmipc_m with {'enz': 'rt7489'}\n",
      "Updating PLACmipc_r with {'enz': 'rt7489'}\n",
      "Updating PLACpail45p_c with {'enz': 'rt4487'}\n",
      "Updating PLACpail45p_n with {'enz': 'rt4487'}\n",
      "Updating PLACpail_c with {'enz': 'rt3387'}\n",
      "Updating PLACpg_mm with {'enz': 'rt3355'}\n",
      "Updating PLADpc_en with {'enz': 'rt5655'}\n",
      "Updating PMANM_c with {'enz': 'rt0873'}\n",
      "Updating PMDPHT_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating PMEMT_rm with {'enz': 'rt4380'}\n",
      "Updating PMEVK_c with {'enz': 'rt0334'}\n",
      "Updating PMI12346PS_c with {'enz': 'rt1395'}\n",
      "Updating PMI1346PS_c with {'enz': 'rt1395'}\n",
      "Updating PMPK_c with {'enz': 'rt4410'}\n",
      "Updating PMTCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating PMTCOAt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating PMTCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating PNP_c with {'enz': 'rt0004rt4016rt7908'}\n",
      "Updating PNTK_c with {'enz': 'rt5817'}\n",
      "Updating PNTOtps_e with {'enz': 'rt3985'}\n",
      "Updating POLYAO1_c with {'enz': 'rt7571'}\n",
      "Updating POLYAO2_c with {'enz': 'rt7571'}\n",
      "Updating POLYAO3_c with {'enz': 'rt7571'}\n",
      "Updating PPA_c with {'enz': 'rt7511'}\n",
      "Updating PPAt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating PPAt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PPBNGD_c with {'enz': 'rt6350'}\n",
      "Updating PPBNGS_1_c with {'enz': 'rt5026'}\n",
      "Updating PPCDC_c with {'enz': 'rt0168'}\n",
      "Updating PPCK_c with {'enz': 'rt2473'}\n",
      "Updating PPGPPDP_c with {'enz': 'rt8139'}\n",
      "Updating PPIt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating PPIt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PPIt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating PPIt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating PPIt_m_mm with {'gene_reaction_rule': ''}\n",
      "Updating PPM_c with {'enz': 'rt5343'}\n",
      "Updating PPNCL2_c with {'enz': 'rt0510'}\n",
      "Updating PPND2_c with {'enz': 'rt7857'}\n",
      "Updating PPNDH_c with {'enz': 'rt5827'}\n",
      "Updating PPPG9t_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PPPGO_m with {'enz': 'rt3499'}\n",
      "Updating PPTT_l with {'enz': 'rt4281'}\n",
      "Updating PPTT_m with {'enz': 'rt0532'}\n",
      "Updating PRAGSi_c with {'enz': 'rt5891'}\n",
      "Updating PRAIS_c with {'enz': 'rt5891'}\n",
      "Updating PRAIi_c with {'enz': 'rt8196'}\n",
      "Updating PRAMPC_c with {'enz': 'rt3278'}\n",
      "Updating PRASCSi_c with {'enz': 'rt0852'}\n",
      "Updating PRATPP_c with {'enz': 'rt3278'}\n",
      "Updating PREN9DPt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating PREN9DPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PRENT10_l with {'enz': 'rt4281'}\n",
      "Updating PRENT11_l with {'enz': 'rt4281'}\n",
      "Updating PRENT12_l with {'enz': 'rt4281'}\n",
      "Updating PRENT13_l with {'enz': 'rt4281'}\n",
      "Updating PRENT14_l with {'enz': 'rt4281'}\n",
      "Updating PRENT15_l with {'enz': 'rt4281'}\n",
      "Updating PRENT16_l with {'enz': 'rt4281'}\n",
      "Updating PRENT17_l with {'enz': 'rt4281'}\n",
      "Updating PRENT18_l with {'enz': 'rt4281'}\n",
      "Updating PRENT19_l with {'enz': 'rt4281'}\n",
      "Updating PRENT20_l with {'enz': 'rt4281'}\n",
      "Updating PRENT21_l with {'enz': 'rt4281'}\n",
      "Updating PRENT6_l with {'enz': 'rt4281'}\n",
      "Updating PRENT7_l with {'enz': 'rt4281'}\n",
      "Updating PRENT8_l with {'enz': 'rt4281'}\n",
      "Updating PRENT9_l with {'enz': 'rt4281'}\n",
      "Updating PRFGS_c with {'enz': 'rt6665'}\n",
      "Updating PRMICI_c with {'enz': 'rt3242'}\n",
      "Updating PROD2_m with {'enz': 'rt6234'}\n",
      "Updating PROTRS_c with {'enz': 'rt3118'}\n",
      "Updating PROt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PRPPS_c with {'enz': 'rt2777rt8224'}\n",
      "Updating PRPPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating PSD_gm with {'enz': 'rt0683'}\n",
      "Updating PSD_mm with {'enz': 'rt2136'}\n",
      "Updating PSD_vm with {'enz': 'rt0683'}\n",
      "Updating PSERT_c with {'enz': 'rt7036'}\n",
      "Updating PSPHPL_r with {'enz': 'rt3557'}\n",
      "Updating PSPHS_r with {'enz': 'rt6946'}\n",
      "Updating PSSA_rm with {'enz': 'rt3215'}\n",
      "Updating PSt_c_m with {'enz': 'rt4312'}\n",
      "Updating PSt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating PSt_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating PSt_gm_rm with {'gene_reaction_rule': ''}\n",
      "Updating PSt_mm_rm with {'gene_reaction_rule': ''}\n",
      "Updating PSt_rm_vm with {'gene_reaction_rule': ''}\n",
      "Updating PTPAT_c with {'enz': 'rt6481'}\n",
      "Updating PTRCAT1_c with {'enz': 'rt7678'}\n",
      "Updating PTRCtpa_e with {'enz': 'rt1939'}\n",
      "Updating PTRCtpa_v with {'enz': 'rt1939'}\n",
      "Updating PUNP3_c with {'enz': 'rt7908'}\n",
      "Updating PUNP3_m with {'enz': 'rt7908'}\n",
      "Updating PUNP5_c with {'enz': 'rt7908'}\n",
      "Updating PYAM5PO_c with {'enz': 'rt2312'}\n",
      "Updating PYDAMK_c with {'enz': 'rt6177'}\n",
      "Updating PYDX5PS_c with {'enz': 'rt2768rt2769'}\n",
      "Updating PYDX5Pt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating PYDXK_c with {'enz': 'rt6177'}\n",
      "Updating PYDXNK_c with {'enz': 'rt6177'}\n",
      "Updating PYDXNOR_c with {'enz': 'rt3182'}\n",
      "Updating PYDXNO_c with {'enz': 'rt2312'}\n",
      "Updating PYDXO_c with {'enz': 'rt2312'}\n",
      "Updating PYK_c with {'enz': 'rt4634'}\n",
      "Updating PYNP2r_c with {'gene_reaction_rule': ''}\n",
      "Updating PYRDC2_c with {'enz': 'rt7423'}\n",
      "Updating PYRDC_c with {'enz': 'rt7423'}\n",
      "Updating PYRtps_e with {'enz': 'rt1816'}\n",
      "Updating PYRtps_m with {'enz': 'rt7405rt8213'}\n",
      "Updating Q9H2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating Q9H2t_c_m with {'gene_reaction_rule': ''}\n",
      "Updating Q9MO_m with {'enz': 'rt7829'}\n",
      "Updating Q9MT1_m with {'enz': 'rt1845'}\n",
      "Updating Q9MT2_m with {'enz': 'rt3097'}\n",
      "Updating Q9OR_m with {'enz': 'rt1845rt3097rt7669rt7829rt8484'}\n",
      "Updating QUILSYN_c with {'gene_reaction_rule': ''}\n",
      "Updating QUINDH_c with {'enz': 'rt5341'}\n",
      "Updating QUINt_c_e with {'enz': 'rt5359'}\n",
      "Updating QULNt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating RADHx_m with {'enz': 'rt5058'}\n",
      "Updating RAFH_e with {'enz': 'rt3242'}\n",
      "Updating RBFK_c with {'enz': 'rt0930'}\n",
      "Updating RBFK_m with {'enz': 'rt0930'}\n",
      "Updating RBFSa_c with {'enz': 'rt0910'}\n",
      "Updating RBFSb_c with {'enz': 'rt4127'}\n",
      "Updating RBK_D_c with {'enz': 'rt6000'}\n",
      "Updating RBK_c with {'enz': 'rt2501'}\n",
      "Updating RDH_c with {'enz': 'rt5407'}\n",
      "Updating RHMND2_c with {'enz': 'rt1893'}\n",
      "Updating RIBFLVt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating RIBt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating RLFC2O_c with {'enz': 'rt0434rt4960'}\n",
      "Updating RLFC2O_m with {'enz': 'rt0434rt4871'}\n",
      "Updating RNDR1_c with {'enz': 'rt2922rt5869rt6971'}\n",
      "Updating RNDR1_n with {'enz': 'rt2922rt5869rt6971'}\n",
      "Updating RNDR2_c with {'enz': 'rt2922rt5869rt6971'}\n",
      "Updating RNDR2_n with {'enz': 'rt2922rt5869rt6971'}\n",
      "Updating RNDR3_c with {'enz': 'rt2922rt5869rt6971'}\n",
      "Updating RNDR3_n with {'enz': 'rt2922rt5869rt6971'}\n",
      "Updating RNDR4_c with {'enz': 'rt2922rt5869rt6971'}\n",
      "Updating RNDR4_n with {'enz': 'rt2922rt5869rt6971'}\n",
      "Updating RNMK_c with {'enz': 'rt6206'}\n",
      "Updating RPE_c with {'enz': 'rt3505'}\n",
      "Updating RPI_c with {'enz': 'rt4920'}\n",
      "Updating SACCD1_c with {'enz': 'rt8465'}\n",
      "Updating SACCD2_c with {'enz': 'rt5719'}\n",
      "Updating SADT_c with {'enz': 'rt5075'}\n",
      "Updating SALCMOX_x with {'enz': 'rt7337'}\n",
      "Updating SAM24MT_c with {'enz': 'rt4692'}\n",
      "Updating SBPP1_r with {'enz': 'rt5044'}\n",
      "Updating SBPP2_r with {'enz': 'rt5044'}\n",
      "Updating SBP_c with {'enz': 'rt4269'}\n",
      "Updating SBTD_D2_c with {'enz': 'rt8084'}\n",
      "Updating SBTDt_c with {'gene_reaction_rule': ''}\n",
      "Updating SBTLt_c with {'gene_reaction_rule': ''}\n",
      "Updating SDPDS_c with {'enz': 'rt3986'}\n",
      "Updating SERAT_c with {'enz': 'rt6880'}\n",
      "Updating SERD_L_c with {'enz': 'rt0848'}\n",
      "Updating SERPT_r with {'enz': 'rt1057rt1935'}\n",
      "Updating SERTRS_c with {'enz': 'rt1878'}\n",
      "Updating SERt_c_m with {'enz': 'rt7910'}\n",
      "Updating SERt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating SERt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating SERtps_e with {'enz': 'rt2976'}\n",
      "Updating SFGTHi_c with {'enz': 'rt1276'}\n",
      "Updating SHCHD2_c with {'enz': 'rt5201'}\n",
      "Updating SHCHF_c with {'enz': 'rt5201'}\n",
      "Updating SHK3Di_c with {'enz': 'rt5884'}\n",
      "Updating SHKK_c with {'enz': 'rt5884'}\n",
      "Updating SLCBK1_r with {'enz': 'rt2023'}\n",
      "Updating SO3R_c with {'enz': 'rt0376rt2006rt2682rt5201'}\n",
      "Updating SO3t_c_e with {'enz': 'rt4278'}\n",
      "Updating SO4t_c_e with {'enz': 'rt5751'}\n",
      "Updating SPHGLpK_r with {'enz': 'rt2023'}\n",
      "Updating SPHINGSt_c_e with {'enz': 'rt8393'}\n",
      "Updating SPHPL_r with {'enz': 'rt3557'}\n",
      "Updating SPMDAT_c with {'gene_reaction_rule': ''}\n",
      "Updating SPMDtpa_e with {'enz': 'rt7439'}\n",
      "Updating SPMS_c with {'enz': 'rt8465'}\n",
      "Updating SPRMAT_c with {'enz': 'rt1145'}\n",
      "Updating SPRMS_c with {'enz': 'rt4913'}\n",
      "Updating SPRMtpa_e with {'enz': 'rt6406'}\n",
      "Updating SPRMtpa_v with {'enz': 'rt2649'}\n",
      "Updating SQLEx_r with {'enz': 'rt5361'}\n",
      "Updating SQLEy_r with {'enz': 'rt5361'}\n",
      "Updating SQLS_c with {'enz': 'rt8223'}\n",
      "Updating SQLt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating SRBt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating SSALy_c with {'enz': 'rt0324'}\n",
      "Updating SSQ23EPXt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating STATg_rm with {'enz': 'rt3431'}\n",
      "Updating STCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating STCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating STEHz161_en with {'enz': 'rt5879'}\n",
      "Updating STEHz161_rm with {'enz': 'rt6249'}\n",
      "Updating STEHz181_en with {'enz': 'rt5879'}\n",
      "Updating STEHz181_rm with {'enz': 'rt6249'}\n",
      "Updating SUCCt2_m with {'enz': 'rt5142'}\n",
      "Updating SUCCt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating SUCDq9_m with {'enz': 'rt3261rt3434rt3609rt7913'}\n",
      "Updating SUCFUMt_c_m with {'enz': 'rt4237'}\n",
      "Updating SUCOAS_m with {'enz': 'rt7599rt7776'}\n",
      "Updating SUCR_e with {'enz': 'rt6458'}\n",
      "Updating SULFOX2_c with {'enz': 'rt8390'}\n",
      "Updating TAGL_l with {'enz': 'rt1378'}\n",
      "Updating TAGL_m with {'enz': 'rt5949'}\n",
      "Updating TAGt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating TAGt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating TAGt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating TALA_c with {'enz': 'rt1805'}\n",
      "Updating TAUDO_c with {'enz': 'rt0250'}\n",
      "Updating TAURt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating TCHOLAtabc_v with {'enz': 'rt2898'}\n",
      "Updating TDCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating TDCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating TDPDRR_c with {'enz': 'rt3425'}\n",
      "Updating THFGLUS_c with {'enz': 'rt6435'}\n",
      "Updating THIORDXi_m with {'enz': 'rt0553'}\n",
      "Updating THIORDXi_n with {'enz': 'rt6669rt6971'}\n",
      "Updating THIORDXi_x with {'enz': 'rt4894rt6971'}\n",
      "Updating THMP_c with {'enz': 'rt2517'}\n",
      "Updating THMtps_e with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating THRA2_c with {'enz': 'rt0854'}\n",
      "Updating THRA_c with {'enz': 'rt7814'}\n",
      "Updating THRD_L_m with {'enz': 'rt3541'}\n",
      "Updating THRS_c with {'enz': 'rt1374'}\n",
      "Updating THRTRS_c with {'enz': 'rt7480'}\n",
      "Updating THRt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating THRtps_e with {'enz': 'rt5861'}\n",
      "Updating THYMDt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating THZPSN1_c with {'gene_reaction_rule': ''}\n",
      "Updating THZPSN2_c with {'gene_reaction_rule': ''}\n",
      "Updating TKT1_c with {'enz': 'rt7263'}\n",
      "Updating TKT2_c with {'enz': 'rt7263'}\n",
      "Updating TMDK1_c with {'gene_reaction_rule': ''}\n",
      "Updating TMDPK_c with {'enz': 'rt8116'}\n",
      "Updating TMDPPK_c with {'enz': 'rt8116'}\n",
      "Updating TMDS_c with {'enz': 'rt2477'}\n",
      "Updating TMLYSOX_m with {'enz': 'rt0812'}\n",
      "Updating TMN_c with {'enz': 'rt4410'}\n",
      "Updating TMPK_c with {'gene_reaction_rule': ''}\n",
      "Updating TMPPP_c with {'enz': 'rt7845'}\n",
      "Updating TORULNOXi_c with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating TPI_c with {'enz': 'rt0932'}\n",
      "Updating TRDOXt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating TRDRDt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating TRDR_c with {'enz': 'rt1320'}\n",
      "Updating TRDR_m with {'enz': 'rt0553rt7114'}\n",
      "Updating TRE6PP_c with {'enz': 'rt3021rt3666'}\n",
      "Updating TRE6PS_c with {'enz': 'rt3021rt3666'}\n",
      "Updating TREH_v with {'enz': 'rt0212'}\n",
      "Updating TREt_c_v with {'gene_reaction_rule': ''}\n",
      "Updating TREtps_e with {'enz': 'rt0769'}\n",
      "Updating TROPNDH_c with {'enz': 'rt0537'}\n",
      "Updating TRPDC_c with {'enz': 'rt1736'}\n",
      "Updating TRPO2_c with {'enz': 'rt7555'}\n",
      "Updating TRPS1_c with {'enz': 'rt0894'}\n",
      "Updating TRPTRS_c with {'enz': 'rt6341'}\n",
      "Updating TRPTRS_m with {'enz': 'rt7165'}\n",
      "Updating TRPt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating TTCCOAt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating TTCCOAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating TTCCOAt_c_r with {'gene_reaction_rule': ''}\n",
      "Updating TTCCOAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating TTCOSAt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating TTCOSAt_c_l with {'gene_reaction_rule': ''}\n",
      "Updating TTCOSAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating TTDCAt_c_e with {'enz': 'rt1544rt2799'}\n",
      "Updating TTDCAt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating TTDCAt_c_x with {'enz': 'rt5693'}\n",
      "Updating TTDCAt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating TYRCBOX_x with {'enz': 'rt5309'}\n",
      "Updating TYRTA_m with {'enz': 'rt5562'}\n",
      "Updating TYRTRS_c with {'enz': 'rt8273'}\n",
      "Updating TYRTRS_m with {'enz': 'rt4769'}\n",
      "Updating TYRt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating TYRt_c_v with {'enz': 'rt6501'}\n",
      "Updating TYRt_c_x with {'gene_reaction_rule': ''}\n",
      "Updating TYRtps_v with {'enz': 'rt4375'}\n",
      "Updating UAGDP_c with {'enz': 'rt0711'}\n",
      "Updating UDPG4E_c with {'enz': 'rt0601'}\n",
      "Updating UDPGALt_c_g with {'enz': 'rt7574'}\n",
      "Updating UDPGLDC_c with {'enz': 'rt7050'}\n",
      "Updating UGLT_c with {'enz': 'rt2964'}\n",
      "Updating UMPK_c with {'enz': 'rt4822'}\n",
      "Updating UMPK_n with {'enz': 'rt4822'}\n",
      "Updating UMPt_c_n with {'gene_reaction_rule': ''}\n",
      "Updating UPP3MT_2_c with {'enz': 'rt2682'}\n",
      "Updating UPP3S_c with {'enz': 'rt2147'}\n",
      "Updating UPPDC1_c with {'enz': 'rt5811'}\n",
      "Updating UPPRT_c with {'enz': 'rt6727'}\n",
      "Updating URAtps_e with {'enz': 'rt4541'}\n",
      "Updating UREASE_c with {'enz': 'rt0958'}\n",
      "Updating UREA_c with {'enz': 'rt2839'}\n",
      "Updating URIDK2_n with {'enz': 'rt4822'}\n",
      "Updating URIDK2i_c with {'enz': 'rt4822'}\n",
      "Updating URIH_c with {'enz': 'rt4016'}\n",
      "Updating URIK1_c with {'enz': 'rt0265'}\n",
      "Updating URIK2_c with {'enz': 'rt0265'}\n",
      "Updating URItps_e with {'enz': 'rt1991'}\n",
      "Updating URO_c with {'enz': 'rt2202'}\n",
      "Updating UTPUMPtp_m with {'enz': 'rt4019'}\n",
      "Updating UTPt_c_m with {'enz': 'rt4019'}\n",
      "Updating VALTA_c with {'enz': 'rt6485'}\n",
      "Updating VALTA_m with {'enz': 'rt6242'}\n",
      "Updating VALTRS_c with {'enz': 'rt8160'}\n",
      "Updating VALTRS_m with {'enz': 'rt8160'}\n",
      "Updating VALt_c_m with {'gene_reaction_rule': ''}\n",
      "Updating VALtps_e with {'enz': 'rt5861'}\n",
      "Updating VANLLOX_x with {'enz': 'rt7726'}\n",
      "Updating XAND_c with {'enz': 'rt7594'}\n",
      "Updating XANt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating XTSNt_c_e with {'gene_reaction_rule': ''}\n",
      "Updating XU5PFGT_x with {'enz': 'rt7263'}\n",
      "Updating XYLI1_c with {'enz': 'rt0738'}\n",
      "Updating XYLI2_c with {'enz': 'rt0738'}\n",
      "Updating XYLK_c with {'enz': 'rt8482'}\n",
      "Updating XYLRy_c with {'enz': 'rt1406'}\n",
      "Updating XYLTOR_c with {'enz': 'rt4609'}\n",
      "Updating XYLTt_c_e with {'enz': 'rt8486'}\n",
      "Updating XYLURx_c with {'enz': 'rt0736'}\n",
      "Updating XYLURy_c with {'enz': 'rt0736'}\n",
      "Updating XYLt_c_e with {'enz': 'RT12976'}\n",
      "Updating YUMPS_c with {'enz': 'rt7720'}\n",
      "Updating ZEAEPOX1_c with {'enz': 'rt0235'}\n",
      "Updating ZEAEPOX2_c with {'enz': 'rt3595'}\n",
      "Updating ZN2t_c_e with {'gene_reaction_rule': ''}\n",
      "Updating ZYM161t_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating ZYM161t_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating ZYM181t_en_rm with {'gene_reaction_rule': ''}\n",
      "Updating ZYM181t_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating ZYMSTATer_RT_rm with {'enz': 'rt3431'}\n",
      "Updating ZYMSTt_c_en with {'gene_reaction_rule': ''}\n",
      "Updating ZYMSTt_c_rm with {'gene_reaction_rule': ''}\n",
      "Updating ZYMSTt_l_rm with {'gene_reaction_rule': ''}\n",
      "Updating compACYLCOA_c with {'gene_reaction_rule': ''}\n",
      "Updating compACYLCOA_l with {'gene_reaction_rule': ''}\n",
      "Updating compACYLCOA_m with {'gene_reaction_rule': ''}\n",
      "Updating compACYLCOA_r with {'gene_reaction_rule': ''}\n",
      "Updating compACYLCOA_rm with {'gene_reaction_rule': ''}\n",
      "Updating compACYLCOA_x with {'gene_reaction_rule': ''}\n",
      "Updating compCER_g with {'gene_reaction_rule': ''}\n",
      "Updating compCER_m with {'gene_reaction_rule': ''}\n",
      "Updating compCER_r with {'gene_reaction_rule': ''}\n",
      "Updating compFALPD_c with {'gene_reaction_rule': ''}\n",
      "Updating compFALPD_en with {'gene_reaction_rule': ''}\n",
      "Updating compFALPD_m with {'gene_reaction_rule': ''}\n",
      "Updating compFALPD_mm with {'gene_reaction_rule': ''}\n",
      "Updating compFALPD_rm with {'gene_reaction_rule': ''}\n",
      "Updating tpscu2_c_m with {'gene_reaction_rule': 'UNKNOWN'}\n",
      "Updating uniDOLDP_l with {'gene_reaction_rule': ''}\n"
     ]
    }
   ],
   "source": [
    "# check if all rxns in model are accounted for in df_enz_rxns, and if they have been automatically updated\n",
    "# make a list of all rxns in df_enz_rxns\n",
    "enz_rxns = []\n",
    "for index, row in df_enz_rxns.iterrows():\n",
    "\tenz_rxns += row['rxns'].split(',')\n",
    "enz_rxns = sorted(list(dict.fromkeys(enz_rxns)))\n",
    "# check if all rxns in model are in enz_rxns\n",
    "missing_rxns = []\n",
    "for rxn in model.reactions:\n",
    "\tif rxn.id not in rxns_not_needing_enzymes and rxn.id not in enz_rxns and rxn.id not in whole_biomass_rxns and rxn.gene_reaction_rule not in spont_rxn_dict.keys():\n",
    "\t\tmissing_rxns.append(rxn.id)\n",
    "if len(missing_rxns) > 0:\n",
    "\tprint(f'Warning: the following reactions are not accounted for in df_enz_rxns:')\n",
    "\tfor rxn in missing_rxns:\n",
    "\t\tprint(rxn)\n",
    "# remove rxns from rxn_updates if their value is {}\n",
    "for rxn in rxn_updates.copy():\n",
    "\tif not rxn_updates[rxn]:\n",
    "\t\tdel rxn_updates[rxn]\n",
    "\telse:\n",
    "\t\tprint(f'Updating {rxn} with {rxn_updates[rxn]}')\n",
    "# if rxn_updates isn't empty, print a warning\n",
    "# if len(rxn_updates) > 0:\n",
    "# \traise ValueError(f'Warning: the following reactions have been automatically updated:\\n{rxn_updates}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert df_enz_rxns to old ENZYME_stoich_curation file format from 2024 version of scRBA, for compatibility\n",
    "df_rxn_enz_pairs = pd.DataFrame(columns=['id','rxn_src','enz','protein_stoich'])\n",
    "catalyst_rxn_dict = {**spont_rxn_dict, **enz_rxn_dict}\n",
    "for k, v in catalyst_rxn_dict.items():\n",
    "\tif k in spont_rxn_dict:\n",
    "\t\tcatalyst_rxn_dict[k]['rxns'].extend(spont_rxn_dict[k]['rxns']) # adds rxns that were overrided by merger with enz_rxn_dict\n",
    "\t\tcatalyst_rxn_dict[k]['rxns'] = list(dict.fromkeys(catalyst_rxn_dict[k]['rxns']))\n",
    "for enz in catalyst_rxn_dict:\n",
    "\tenz_id = enz if enz != '' else 'SPONT'\n",
    "\t# don't seem to need MW of enzyme in prior RBA code\n",
    "\tprot_stoich_str = 'zeroCost' if enz in spont_GPRs + [''] else ','.join([prot+':'+str(catalyst_rxn_dict[enz]['protein_stoich'][prot]) for prot in catalyst_rxn_dict[enz]['protein_stoich']])\n",
    "\tfor rxn in catalyst_rxn_dict[enz]['rxns']:\n",
    "\t\tdirs = []\n",
    "\t\t# add FWD and REV to rxn_src if those directions are possible\n",
    "\t\tif model.reactions.get_by_id(rxn).lower_bound < 0:\n",
    "\t\t\tdirs.append('REV')\n",
    "\t\tif model.reactions.get_by_id(rxn).upper_bound > 0:\n",
    "\t\t\tdirs.append('FWD')\n",
    "\t\tfor dir in dirs:\n",
    "\t\t\t# print(enz,rxn)\n",
    "\t\t\t# df_rxn_enz_pairs = df_rxn_enz_pairs.append({'id':f'RXN-{rxn}_{dir}-{enz}','enz':enz,'gpr':'','protein_stoich':','.join([prot+':'+str(enz_rxn_dict[enz]['protein_stoich'][prot]) for prot in enz_rxn_dict[enz]['protein_stoich']]),'status':'automatically added'}, ignore_index=True)\n",
    "\t\t\tdf_rxn_enz_pairs.loc[len(df_rxn_enz_pairs)] = [f'RXN-{rxn}_{dir}-{enz_id}',rxn,enz,prot_stoich_str]\n",
    "\t\t\t# df_rxn_enz_pairs = pd.concat([df_rxn_enz_pairs,pd.DataFrame({'id':f'RXN-{rxn}_{dir}-{enz}','enz':enz,'protein_stoich':','.join([prot+':'+str(enz_rxn_dict[enz]['protein_stoich'][prot]) for prot in enz_rxn_dict[enz]['protein_stoich']])})], ignore_index=True)\n",
    "\n",
    "# sort by ID\n",
    "df_rxn_enz_pairs = df_rxn_enz_pairs.sort_values(by='id')\t\t\t\n",
    "df_rxn_enz_pairs.to_csv('./model/ENZYME_rxn_pairs.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_eqn = pd.DataFrame(columns=['id', 'type', 'coupling_type', 'coupling_species', 'reaction','tag','FBA_name','dir','enz_id'])\n",
    "c = -1\n",
    "medium = []\n",
    "\n",
    "### Metabolic network reaction\n",
    "# Exchange reactions\n",
    "for rxn in model.reactions:\n",
    "\t# check if rxn is in df_enz\n",
    "\t# if rxn.id not in df_enz.rxn_src.to_list():\n",
    "\tif rxn.id[:3] == 'EX_':\n",
    "\t\tenz_id = 'SPONT'\n",
    "\t\tmake_reversible = True\n",
    "# else:\n",
    "# \tgpr = rxn.gene_reaction_rule if rxn.gene_reaction_rule != \"\" else 'UNKNOWN'\n",
    "\n",
    "\t\tprotein_stoich = 'zeroCost' if enz_id in spont_GPRs else ''\n",
    "\t\t\n",
    "\t\tmet = [i for i in rxn.metabolites.keys()][0]\n",
    "\t\tc += 1\n",
    "\t\ttag = 'RXN'\n",
    "\t\tdir = 'FWD'\n",
    "\t\tnew_id = f'{tag}-{rxn.id}_{dir}-{enz_id}'\n",
    "\t\tdf_eqn.loc[c, 'id'] = new_id\n",
    "\t\tdf_eqn.loc[c, 'type'] = 'metabolic'\n",
    "\t\tdf_eqn.loc[c, 'reaction'] = 'MET-' + met.id + ' -->'\n",
    "\t\t# df_eqn.loc[c, ['FBA_name','dir']] = [rxn.id,'FWD']\n",
    "\t\tdf_eqn.loc[c, ['tag','FBA_name','dir','enz_id']] = [tag,rxn.id,dir,enz_id]\n",
    "\t\t\n",
    "\t\tc += 1\n",
    "\t\tdir = 'REV'\n",
    "\t\tnew_id = f'{tag}-{rxn.id}_{dir}-{enz_id}'\n",
    "\t\tdf_eqn.loc[c, 'id'] = new_id\n",
    "\t\tdf_eqn.loc[c, 'type'] = 'metabolic'\n",
    "\t\tdf_eqn.loc[c, 'reaction'] = '-->' + 'MET-' + met.id\n",
    "\t\t# df_eqn.loc[c, ['FBA_name','dir']] = [rxn.id,'REV']\n",
    "\t\tdf_eqn.loc[c, ['tag','FBA_name','dir','enz_id']] = [tag,rxn.id,dir,enz_id]\n",
    "\n",
    "# Reactions that are not exchange reactions\n",
    "for i in df_enz.index:\n",
    "\trxn_id = df_enz.id[i]\n",
    "\ttag,rxn_base_id,rxn_dir,enz_id = extract_details_from_rxnid(rxn_id)\n",
    "\t\n",
    "\tif rxn_base_id[:3] == 'EX_':\n",
    "\t\tcontinue\n",
    "\t\n",
    "\tc += 1\n",
    "\trxn_base = model.reactions.get_by_id(rxn_base_id)\n",
    "\t\n",
    "\tmet_dict = metabolites_dict_from_reaction_equation_RBA(rxn_base.reaction)\n",
    "\tmet_dict = {k:v for k,v in met_dict.items() if k != ''}\n",
    "\tmet_dict = {'MET-' + k:v for k,v in met_dict.items()}\n",
    "\tif rxn_dir == 'REV':\n",
    "\t\tmet_dict = {k:-v for k,v in met_dict.items()}\n",
    "\telif rxn_dir == 'FWD':\n",
    "\t\tNone\n",
    "\telse:\n",
    "\t\tprint(\"Unknown ID that indicate reaction direction, only accepting 'FWD' and 'REV'\")\n",
    "\t\n",
    "\tif enz_id not in spont_GPRs:\n",
    "\t\tdf_eqn.loc[c, 'coupling_type'] = 'rxn_enz'\n",
    "\t\tdf_eqn.loc[c, 'coupling_species'] = enz_id\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = rxn_id\n",
    "\tdf_eqn.loc[c, 'type'] = 'metabolic'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(met_dict, arrow='-->')\n",
    "\tdf_eqn.loc[c, ['tag','FBA_name','dir','enz_id']] = [tag,rxn_base_id,rxn_dir,enz_id]\n",
    "\t\n",
    "### Enzyme synthesis network reaction\n",
    "enz_stoich = OrderedDict()\n",
    "for i in df_enz.index:\n",
    "\tenz_stoich[df_enz.enz[i]] = df_enz.protein_stoich[i]\n",
    "\n",
    "c = df_eqn.shape[0] - 1\n",
    "for enz_id,prot_str in enz_stoich.items():\n",
    "\tif prot_str == 'zeroCost':\n",
    "\t\tcontinue\n",
    "\t\n",
    "\tc += 1\n",
    "\tprot_str = prot_str.split(',')\n",
    "\tcoeffs = OrderedDict({'PTMPRO-' + i.split(':')[0]:-int(i.split(':')[1]) for i in prot_str})\n",
    "\tcoeffs['ENZ-' + enz_id] = 1\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = 'ENZSYN-' + enz_id\n",
    "\tdf_eqn.loc[c, 'type'] = 'enzyme'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(coeffs, arrow='-->')\n",
    "\n",
    "#these ENZLOAD equations also need to have their MW listed in the enz_mw_g_per_mmol file\n",
    "#initialize an enzload sting\n",
    "enzload_str = \"\"\n",
    "for i in df_enz.index:\n",
    "\tif df_enz.protein_stoich[i] == 'zeroCost':\n",
    "\t\tcontinue\n",
    "\t\t\n",
    "\tc += 1\n",
    "\tcoeffs = OrderedDict()\n",
    "\tcoeffs['ENZ-' + df_enz.enz[i]] = -1\n",
    "\n",
    "\tenzload_id = 'ENZLOAD-' + df_enz.id[i][4:]\n",
    "\n",
    "\t#for debugging, comment out if unused\n",
    "\t#print(\"i: \",i,\"\\tenzload_id: \", enzload_id)\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = enzload_id\n",
    "\tdf_eqn.loc[c, 'type'] = 'enzymeRxnLoad'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(coeffs, arrow='-->')\n",
    "\n",
    "\t#get the associated mw\n",
    "\tenz_mw = getattr(df_enz,\"MW (g/mmol)\")[i]\n",
    "\n",
    "\tnew_line = enzload_id + \"\\t\" + str(enz_mw) + \"\\n\"\n",
    "\tenzload_str = enzload_str + new_line\n",
    "\n",
    "# write the file resulting from enzload_str\n",
    "with open('./model/enz_mw_g_per_mmol.txt', 'w') as f:\n",
    "\tf.write(enzload_str)\n",
    "with open('../input/enz_mw_g_per_mmol.txt', 'w') as f:\n",
    "\tf.write(enzload_str)\n",
    "\n",
    "### Ribosome\n",
    "ribo_subunits = {}\n",
    "c = df_eqn.shape[0] - 1\n",
    "for rna in df_rnas.index:\n",
    "\tc += 1\n",
    "\trna_stoich = OrderedDict({i:0 for i in ['MET-'+rna, 'MET-atp_c', 'MET-ctp_c',\n",
    "\t\t\t\t\t\t\t\t\t\t\t'MET-gtp_c', 'MET-utp_c', 'MET-ppi_c']})\n",
    "\trna_stoich['RIBO-'+rna] = 1\n",
    "\trna_stoich['MET-atp_c'] = -int(df_rnas.A[rna])\n",
    "\trna_stoich['MET-ctp_c'] = -int(df_rnas.C[rna])\n",
    "\trna_stoich['MET-gtp_c'] = -int(df_rnas.G[rna])\n",
    "\trna_stoich['MET-utp_c'] = -int(df_rnas.U[rna])\n",
    "\trna_stoich['MET-ppi_c'] = int(df_rnas.loc[rna, ['A','C','G','U']].sum())\n",
    "\trna_stoich['BIO-rrna'] = df_rnas.loc[rna, 'MW (g/mmol)']\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = 'RIBOSYN-' + rna\n",
    "\tdf_eqn.loc[c, 'type'] = 'ribosome'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(rna_stoich, arrow='-->')\n",
    "\tgene_expression_dict.setdefault(rna, []).append(df_eqn.loc[c, 'id'])\n",
    "with open('./model/prosyn_ribo_subunits.txt', 'w') as f, open('./model/pro_ribo_subunits.txt', 'w') as f2, open('./model/kribo.txt', 'w') as f4, open('./model/j_ribo.txt', 'w') as f5, open('./model/ribosomes.txt', 'w') as f6:\n",
    "\tfor file in [f,f2,f4,f5,f6]: file.write('/')\n",
    "\tfor loc, ribo in ribo_dict.items():\n",
    "\t\t## proteins into ribosome subunits (to account for paralogs)\n",
    "\t\tribo_stoich = OrderedDict()\n",
    "\t\t# if only 1 ribosome is present, all rRNAs are used\n",
    "\t\trnas = [item for item in df_rnas.index if item.endswith(loc) or len(ribo_dict.keys()) == 1]\n",
    "\t\t# assume 1 of each rRNA and protein is used per ribosome\n",
    "\t\tfor i in ribo.index:\n",
    "\t\t\tif ribo.id[i] in ribo_subunits:\n",
    "\t\t\t\tribo_subunits[ribo.id[i]].append(loc[1:])\n",
    "\t\t\telse:\n",
    "\t\t\t\tribo_subunits[ribo.id[i]] = [loc[1:]]\n",
    "\t\t\tif ribo.id[i] in rnas:\n",
    "\t\t\t\tribo_stoich['RIBO-' + ribo.id[i]] = -1\n",
    "\t\t\telse:\n",
    "\t\t\t\tf.write(\"\\n'\"+loc[1:]+\"'.'PROSYN-\"+ribo.id[i]+\"'\")\n",
    "\t\t\t\tf2.write(\"\\n'\"+loc[1:]+\"'.'\"+ribo.id[i]+\"'\")\n",
    "\t\t\t\t# if it's not a paralog of another protein, all paralogs between it and the next non-paralog are treated as paralogs\n",
    "\t\t\t\tif pd.isnull(ribo.paralog[i]):\n",
    "\t\t\t\t\tribosub_name = 'RIBOSUB-' + ribo.id[i]\n",
    "\t\t\t\t# add rxn converting the protein into the ribosome subunit (RIBOSUB prefix)\n",
    "\t\t\t\tc += 1\n",
    "\t\t\t\tribosub_stoich = OrderedDict()\n",
    "\t\t\t\tribosub_stoich['PTMPRO-' + ribo.id[i]] = -1\n",
    "\t\t\t\tribosub_stoich[ribosub_name] = 1\n",
    "\t\t\t\tdf_eqn.loc[c, 'id'] = 'RIBOSUBSYN-' + ribosub_name + '-FROM-' + ribo.id[i]\n",
    "\t\t\t\tdf_eqn.loc[c, 'type'] = 'ribosome-subunit'\n",
    "\t\t\t\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(ribosub_stoich, arrow='-->')\n",
    "\t\t\t\t# add the ribosome subunit to the ribosome\n",
    "\t\t\t\tribo_stoich[ribosub_name] = -1\n",
    "\t\t# making the ribosome itself from the subunits\n",
    "\t\tc += 1\n",
    "\t\tdf_eqn.loc[c, 'id'] = 'RIBOSYN-ribonuc' if loc == '_c' else 'RIBOSYN-ribomito' if loc == '_m' else 'RIBOSYN-ribo'+loc[1:]\n",
    "\t\t# add to ribosome files\n",
    "\t\tribo_id = \"'\"+loc[1:]+\"'\"\n",
    "\t\tf6.write('\\n'+ribo_id); f5.write('\\n'+ribo_id+\".'\"+df_eqn.loc[c, 'id']+\"'\"); f4.write('\\n'+ribo_id+' '+str(kribo_default))\n",
    "\t\tdf_eqn.loc[c, 'type'] = 'ribosome'\n",
    "\t\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(ribo_stoich, arrow='-->')\n",
    "\tfor file in [f,f2,f4,f5,f6]: file.write('\\n/')\n",
    "# write the ribosome subunit stoichiometry to a json file\n",
    "with open('./model/ribo_subunits.json', 'w') as f:\n",
    "\tjson.dump(ribo_subunits, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Build files related to MW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = getattr(df_pro,'MW (g/mmol)').to_dict()\n",
    "ids = getattr(df_pro,'id').to_list()\n",
    "\n",
    "#build the string to write\n",
    "prot_mw_str = \"\"\n",
    "\n",
    "#for each id\n",
    "for id in ids:\n",
    "\tnew_line = id + \"\\t\" + str(data[id]) + \"\\n\"\n",
    "\tprot_mw_str = prot_mw_str + new_line\n",
    "# add dummy proteins to the string\n",
    "for prot in dummy_list:\n",
    "\tnew_line = prot[\"name\"].replace('PROSYN-','') + \"\\t\" + str(prot[\"MW\"]) + \"\\n\"\n",
    "\tprot_mw_str = prot_mw_str + new_line\n",
    "\n",
    "#write to output\n",
    "with open('./model/pro_mw_g_per_mmol.txt', 'w') as f:\n",
    "\tf.write(prot_mw_str)\n",
    "with open('../input/pro_mw_g_per_mmol.txt', 'w') as f:\n",
    "\tf.write(prot_mw_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TEST: this whole block. get MW list\n",
    "mw_list = getattr(df_enz,\"MW (g/mmol)\").to_list()\n",
    "\n",
    "#get enz list\n",
    "enz_list = getattr(df_enz,\"enz\").to_list()\n",
    "\n",
    "#create string for writing output file\n",
    "enz_mw_str = \"enz_id\\tMW (g/mmol)\\n\"\n",
    "\n",
    "#add to the string for enzyme entries\n",
    "#need to initialize a counter\n",
    "c = 0\n",
    "\n",
    "#list to keep track of existing entries\n",
    "#predefined values prevent adding unknown\n",
    "existing_pro = ['nan','unknown']\n",
    "\n",
    "for enz in enz_list:\n",
    "\n",
    "\t#only add a new line if unique entry\n",
    "\tif not(str(enz) in existing_pro):\n",
    "\n",
    "\t\t#for debugging, comment out if unused\n",
    "\t\t#print(\"enz: |\",enz,\"|, type: \",type(enz))\n",
    "\t\n",
    "\t\tnew_line = str(enz) + \"\\t\" + str(mw_list[c]) + \"\\n\"\n",
    "\n",
    "\t\t#for debugging, comment out if unused\n",
    "\t\t#print(new_line)\n",
    "\n",
    "\t\tenz_mw_str = enz_mw_str + new_line\n",
    "\n",
    "\t\texisting_pro.append(enz)\n",
    "\n",
    "\tc = c + 1\n",
    "\n",
    "#write to output\n",
    "with open('../input/enz_mw_g_per_mmol_norxnmapped.txt', 'w') as f:\n",
    "\tf.write(enz_mw_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Protein file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Protein\n",
    "\n",
    "pro_lengths = dict()\n",
    "\n",
    "c = df_eqn.shape[0] - 1\n",
    "translation_loci = set()\n",
    "unknown_ribo_prosyn = set()\n",
    "protein_functions = dict()\n",
    "\n",
    "def edit_stoich(stoich:OrderedDict,edits):\n",
    "\tif not pd.isna(edits):\n",
    "\t\tcofs_st = edits.split(',')\n",
    "\t\tcofs_st = OrderedDict({c.split(':')[0]:c.split(':')[1] for c in cofs_st})\n",
    "\t\tfor k,v in cofs_st.items():\n",
    "\t\t\tstoich['MET-'+k] = -int(v)\n",
    "\treturn stoich\n",
    "\n",
    "for i in df_pro.index: \n",
    "\tc += 1\n",
    "\tpro_id = df_pro.loc[i, 'id'] if 'id' in df_pro.columns else i\n",
    "\tprepro = 'PRO-' + str(pro_id) # preprotein (from translation)\n",
    "\tprot = 'PTMPRO-' + str(pro_id) # protein after post-translational modifications (PTMs)\n",
    "\n",
    "\ttrans_st = OrderedDict()\n",
    "\tPTM_st = OrderedDict({prepro:-1,prot:1})\n",
    "\tfor met in ['MET-atp_c', 'MET-h2o_c',\n",
    "\t\t\t\t'MET-adp_c', 'MET-pi_c', 'MET-h_c', 'MET-gtp_c',\n",
    "\t\t\t\t'MET-gdp_c']:\n",
    "\t\ttrans_st[met] = 0\n",
    "\tseq = df_pro.sequence[i].replace(\"*\",\"\")\n",
    "\n",
    "\ttrans_st = edit_stoich(trans_st,df_pro.trans_stoich_edits[i])\n",
    "\tPTM_st = edit_stoich(PTM_st,df_pro.PTM_stoich_edits[i])\n",
    "\n",
    "\tpro_lengths[i] = len(seq)\n",
    "\tfor aa in df_aamap.index:\n",
    "\t\ttrans_st[df_aamap.tRNA_in[aa]] = -seq.count(aa)\n",
    "\t\ttrans_st[df_aamap.tRNA_out[aa]] = seq.count(aa)\n",
    "\t\t\t\t\n",
    "\ttrans_st[prepro] = 1\n",
    "\tdf_eqn.loc[c, 'coupling_type'] = 'prot_ribo'\n",
    "\tdf_eqn.loc[c, 'coupling_species'] = 'ribo'\n",
    "\t\n",
    "\t# protein will occupy cellular space in its specific compartment\n",
    "\tif df_pro.subloc_assigned[i] in mito_compartments:\n",
    "\t\ttrans_st['BIO-protmito'] = df_pro.loc[i, 'MW (g/mmol)']\n",
    "\telse:\n",
    "\t\ttrans_st['BIO-protcyt'] = df_pro.loc[i, 'MW (g/mmol)']\n",
    "\t\n",
    "\t# Cost: Initiation: 1 ATP + 2 GTP (initiate and bind Methionine)\n",
    "\tfor met in ['MET-atp_c', 'MET-h2o_c']:\n",
    "\t\ttrans_st[met] -= 1\n",
    "\tfor met in ['MET-adp_c', 'MET-pi_c', 'MET-h_c']:\n",
    "\t\ttrans_st[met] += 1\n",
    "\t\t\t\t\n",
    "\t# Elongation: 2 GTP / cycle (2(n-1) cycles for n amino acids + 2 GTP from initiation = 2n GTP consumed so far)\n",
    "\t# (elongation process excludes Methionine since it is already bound in initiation process)\n",
    "\t# Termination and recycling: 1 GTP + 1 ATP: https://doi.org/10.1534/genetics.115.186221\n",
    "\t# \tHowever, we exclude these costs since it's not always required (e.g., via \"leaky scanning\"): https://doi.org/10.1093/nar/gkp765\n",
    "\tfor met in ['MET-gtp_c', 'MET-h2o_c']:\n",
    "\t\ttrans_st[met] -= 2*len(seq)\n",
    "\tfor met in ['MET-gdp_c', 'MET-pi_c', 'MET-h_c']:\n",
    "\t\ttrans_st[met] += 2*len(seq)\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = 'PROSYN-' + df_pro.id[i]\n",
    "\n",
    "\t# add translation rxn to gene_expression_dict list, or create a new list if it doesn't exist\n",
    "\tgene_expression_dict.setdefault(df_pro.gene_src[i], []).append(df_eqn.loc[c, 'id'])\n",
    "\n",
    "\t# add to protein functions dictionary under its gene_src\n",
    "\tprotein_functions.setdefault(df_pro.gene_src[i], {})[df_pro.id[i]] = {}\n",
    "\tdf_eqn.loc[c, 'type'] = 'protein'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(trans_st, arrow='-->')\n",
    "\t# Set ribosome translating the protein\n",
    "\tif df_pro.loc[i, 'translation_loc'] in ['unknown','any']:\n",
    "\t\tunknown_ribo_prosyn.add(\"'\"+df_eqn.loc[c, 'id']+\"'\")\n",
    "\telse:\n",
    "\t\ttranslation_loci.add(\"'\"+df_pro.loc[i, 'translation_loc']+\"'.'\"+df_eqn.loc[c, 'id']+\"'\")\n",
    "\t# add PTM reaction\n",
    "\tc += 1\n",
    "\tdf_eqn.loc[c, 'id'] = 'PTM-' + df_pro.id[i]\n",
    "\tdf_eqn.loc[c, 'type'] = 'proteinPTM'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(PTM_st, arrow='-->')\n",
    "# write the protein length file\n",
    "len_str = \"/\\n\"\n",
    "for prot in pro_lengths:\n",
    "\tlen_str = len_str + \"'PROSYN-\" + prot + \"' \" + str(pro_lengths[prot]) + \"\\n\"\n",
    "for prot in dummy_list:\n",
    "\tlen_str += \"'\" + prot[\"name\"] + \"' \" + str(prot[\"medianL\"]) + \"\\n\"\n",
    "len_str = len_str + \"/\"\n",
    "with open('./model/RBA_proteinLength.txt', 'w') as f:\n",
    "\tf.write(len_str)\n",
    "\n",
    "### Dummy protein\n",
    "for prot in dummy_list:\n",
    "\tdummy_MW = prot[\"MW\"]\n",
    "\ttrans_st = make_dummy_protein_stoich(aa_standards_df=df_aamap, prot_df=prot[\"df\"], length=prot[\"medianL\"], rxn_name=prot[\"name\"], mw=dummy_MW)\n",
    "\n",
    "\tc += 1\n",
    "\t# prot_st[\"BIO-protdummy\"] = prot[\"MW\"]\n",
    "\tdf_eqn.loc[c, 'id'] = prot[\"name\"]\n",
    "\tdf_eqn.loc[c, 'coupling_type'] = 'prot_ribo'\n",
    "\tdf_eqn.loc[c, 'coupling_species'] = 'ribo' #TEST: Update for ribonuc, ribomito, and other ribo types\n",
    "\tdf_eqn.loc[c, 'type'] = 'protein'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(trans_st,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\tarrow='-->', floatdecimal=6)\n",
    "\tif prot[\"translation_loc\"] in ['unknown','any']:\n",
    "\t\tunknown_ribo_prosyn.add(\"'\"+df_eqn.loc[c, 'id']+\"'\")\n",
    "\telse:\n",
    "\t\ttranslation_loci.add(\"'\"+df_eqn.loc[c, 'id']+\"'.'\"+df_eqn.loc[c, 'id']+\"'\")\n",
    "\n",
    "# Protein waste reactions\n",
    "c = df_eqn.shape[0] - 1\n",
    "with open('./model/pro_syn_waste.txt', 'w') as f:\n",
    "\tf.write('/')\n",
    "\tfor i in df_pro.index: \n",
    "\t\tfor abbv,tag in {'':'PROSYN','PTM':'PTM'}.items():\n",
    "\t\t\tc += 1\n",
    "\t\t\tpro=abbv+'PRO-'+df_pro.id[i]\n",
    "\t\t\tnew_st = OrderedDict()\n",
    "\t\t\tnew_st[pro] = -1\n",
    "\t\t\t# add the protein's MW as the coefficient for BIO-protwasted\n",
    "\t\t\tnew_st['BIO-protwasted'] = df_pro.loc[i, 'MW (g/mmol)']\n",
    "\t\t\t\n",
    "\t\t\tpw_id = f\"{abbv}PROWASTE-{df_pro.id[i]}\"\n",
    "\t\t\tf.write(f\"\\n'{tag}-{df_pro.id[i]}'.'{pw_id}'\")\n",
    "\n",
    "\t\t\tdf_eqn.loc[c, 'id'] = pw_id\n",
    "\t\t\tdf_eqn.loc[c, 'type'] = 'proteinWaste'\n",
    "\t\t\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(new_st, arrow='-->')\n",
    "\n",
    "\tf.write('\\n/')\n",
    "# Pseudoreaction representing mass of waste protein (for kapps): PROWASTE-TOTALPROTEIN\n",
    "c += 1\n",
    "df_eqn.loc[c, 'id'] = 'PROWASTE-TOTALPROTEIN'\n",
    "df_eqn.loc[c, 'type'] = 'proteinWaste'\n",
    "df_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA({'BIO-protwasted':-1}, arrow='-->')\n",
    "for loc in ['modeled','dummy','cyt','mito']: # add PROWASTE rxns for biomass-related byproduct of PROSYN rxns, to support simulations at 0 growth\n",
    "\tc += 1\n",
    "\tdf_eqn.loc[c, 'id'] = 'PROWASTE-PROT'+loc.upper()\n",
    "\tdf_eqn.loc[c, 'type'] = 'proteinWaste'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA({'BIO-prot'+loc:-1}, arrow='-->')\n",
    "\n",
    "### Write translation localization lists\n",
    "with open('./model/translation.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(list(translation_loci)) + ['/']))\n",
    "with open('./model/translated_by_any_ribo.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(list(unknown_ribo_prosyn)) + ['/']))\n",
    "\n",
    "### Biomass\n",
    "for i in df_biom.index:\n",
    "\tc += 1\n",
    "\tif not pd.isna(df_biom.rxn_id[i]):\n",
    "\t\tdf_eqn.loc[c, 'id'] = df_biom.rxn_id[i]\n",
    "\t\tdf_eqn.loc[c, 'type'] = 'biomass'\n",
    "\t\tdf_eqn.loc[c, 'reaction'] = df_biom.rxn_equation[i]\n",
    "# make into file RBA_rxns_whole_biomass.txt\n",
    "with open('./model/RBA_rxns_whole_biomass.txt', 'w') as f:\n",
    "\t# find whole_biomass_rxns in df_eqn and write them to file\n",
    "\tf.write('\\n'.join(['/'] + [i for i in whole_biomass_rxns if i in df_eqn.id.to_list()] + ['/']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k,v in gene_expression_dict.items():\n",
    "\t# check for entries in v starting with \"'RIBO\"\n",
    "\tif any([i.startswith(\"'RIBO\") for i in v]):\n",
    "\t\tprint(f\"Warning: {k} has entries starting with 'RIBO' in gene_expression_dict\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save gene list to file\n",
    "with open('./model/genes.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + [f\"'{i}'\" for i in genes] + ['/']))\n",
    "for gene in genes:\n",
    "\tif gene not in gene_expression_dict:\n",
    "\t\tgene_expression_dict[gene] = []\n",
    "# remove duplicates from gene_expression_dict, preserve order\n",
    "gene_expression_dict = {k: list(dict.fromkeys([f\"'{i}'\" for i in v])) for k, v in gene_expression_dict.items()}\n",
    "# save gene-protein mapping to file\n",
    "with open('./model/genes_expression.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + [f\"'{gene}'.({','.join(gene_expression_dict[gene])})\" for gene in genes if gene_expression_dict[gene]] + ['/']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make extra kapp calculation files\n",
    "directory = '../GAMS/parameterization/enz_from_proteome'\n",
    "if not os.path.exists(directory):\n",
    "\tos.makedirs(directory)\n",
    "\n",
    "with open(directory + '/pro_and_enz.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(list(set([\"'ENZ-\" + i + \"'\" for i in enz_list] + [\"'PTMPRO-\" + i + \"'\" for i in df_pro.index]))) + ['/']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save excel files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_eqn.to_csv('./model/RBA_stoichiometry.tsv', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build GAMS Sij and flux bounds (originally A02_build_GAMS_Sij_and_fluxBounds.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_eqn = read_spreadsheet('./model/RBA_stoichiometry.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assemble list of species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "met_list_raw = []\n",
    "for i in df_eqn.index:\n",
    "\tmet_dict = metabolites_dict_from_reaction_equation_RBA(df_eqn.reaction[i])\n",
    "\tmet_list_raw += met_dict.keys()\n",
    "\n",
    "met_list = sorted(list(set(met_list_raw)))\n",
    "met_list = [i for i in met_list if i != '']\n",
    "with open('./model/RBA_species.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + [\"'\" + i + \"'\" for i in met_list] + ['/']))\n",
    "# add list of demand rxns for each metabolite (for testing purposes)\n",
    "demand_list = [\"DM-\" + i for i in met_list]\n",
    "# WIP: add to df_eqn\n",
    "with open('./model/RBA_rxns_demand.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + [\"'DM-\" + i + \"'\" for i in met_list] + ['/']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assemble list of reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = df_eqn.id.to_list()\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of uptake reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag,rxn_base_id,rxn_dir,enz_id = extract_details_from_rxnid(i)\n",
    "\tif tag == 'RXN' and rxn_base_id[:3] == 'EX_' and rxn_dir == 'REV':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_EXREV.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of secretion reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag,rxn_base_id,rxn_dir,enz_id = extract_details_from_rxnid(i)\n",
    "\tif tag == 'RXN' and rxn_base_id[:3] == 'EX_' and rxn_dir == 'FWD':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_EXFWD.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of protein waste reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag = i.split('-')[0]\n",
    "\tif tag in ['PROWASTE','PTMPROWASTE']:\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_prowaste.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of enzyme syn reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag = i.split('-')[0]\n",
    "\tif tag == 'ENZSYN':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_enzsyn.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of enzyme load reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag = i.split('-')[0]\n",
    "\tif tag == 'ENZLOAD':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "# list created so GAMS knows which ENZLOAD corresponds to which rxn\n",
    "rxn_enzload_coupling_list = [i + '.' + i.replace('ENZLOAD-','RXN-') + ' 1' for i in rxn_list]\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_enzload.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))\n",
    "with open('./model/RBA_rxn_enzload_coupling.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + rxn_enzload_coupling_list + ['/']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of metabolic network reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag = i.split('-')[0]\n",
    "\tif tag == 'RXN':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_rxnmetabolicnetwork.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make k_app (enzyme turnover rate) file draft using placeholder values\n",
    "kapp_list = []\n",
    "kapp_list = [i + \" 360000\" for i in rxn_list if i != '/']\n",
    "kapp_list = ['/'] + kapp_list + ['/']\n",
    "with open('./model/kapp_placeholders.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(kapp_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assemble stoichiometry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make extra kapp calculation files\n",
    "directory = '../GAMS/parameterization/enz_from_proteome/'\n",
    "if not os.path.exists(directory):\n",
    "\tos.makedirs(directory)\n",
    "\n",
    "sij = []\n",
    "rxns_enz = []\n",
    "rxns_proenz = []\n",
    "sij_proenz = []\n",
    "\n",
    "for i in df_eqn.index:\n",
    "\tx = metabolites_dict_from_reaction_equation_RBA(df_eqn.reaction[i])\n",
    "\tmet_dict = dict()\n",
    "\tfor k,v in x.items():\n",
    "\t\tif k == '':\n",
    "\t\t\tcontinue\n",
    "\t\t# note when a protein is used in a reaction other than its PROWASTE and production reactions\n",
    "\t\tmet = k.split('-',1)[1]\n",
    "\t\tif k.startswith('PTMPRO-') and df_eqn.id[i] not in ['PTMPROWASTE-' + met,'PTM-' + met]:\n",
    "\t\t\tfor gene,prots in protein_functions.items():\n",
    "\t\t\t\tif met in prots:\n",
    "\t\t\t\t\tprotein_functions[gene][met].update({df_eqn.id[i]:v})\n",
    "\t\t\t\t\tbreak\n",
    "\t\tif v.is_integer():\n",
    "\t\t\tmet_dict[k] = int(v)\n",
    "\t\telse:\n",
    "\t\t\tmet_dict[k] = v\n",
    "\t\t\t\n",
    "\trxn_prefixes = ('ENZ', 'PTM-')\n",
    "\tif df_eqn.id[i].startswith(rxn_prefixes):\n",
    "\t\tif df_eqn.id[i].startswith(\"PTM-\"):\n",
    "\t\t\trxns_proenz.append(\"'\" + df_eqn.id[i].replace('PTM-','PROIN-') + \"'\")\n",
    "\t\telse:\n",
    "\t\t\trxns_proenz.append(\"'\" + df_eqn.id[i] + \"'\")\n",
    "\t\t\trxns_enz.append(\"'\" + df_eqn.id[i] + \"'\")\n",
    "\t\t# sij.append(\"'\" + df_eqn.id[i] + \"'.'\" + df_eqn.id[i] + \"' 1\")\n",
    "\t\tfor k,v in met_dict.items():\n",
    "\t\t\tsij.append(\"'\" + k + \"'.'\" + df_eqn.id[i] + \"' \" + str(v))\n",
    "\t\t\tif k.startswith('PTMPRO-') or k.startswith('ENZ-'):\n",
    "\t\t\t\tsij_proenz.append(\"'\" + k + \"'.'\" + df_eqn.id[i].replace('PTM-','PROIN-') + \"' \" + str(v))\n",
    "\telse:\n",
    "\t\tfor k,v in met_dict.items():\n",
    "\t\t\tsij.append(\"'\" + k + \"'.'\" + df_eqn.id[i] + \"' \" + str(v))\n",
    "\t\n",
    "sij = ['/'] + sij + ['/']\n",
    "with open('./model/RBA_sij.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(sij))\n",
    "# write protein functions to file\n",
    "with open('./protein_functions.json', 'w') as f:\n",
    "\tjson.dump(protein_functions, f)\n",
    "# make set of all proteins in protein_functions with no modeled functions\n",
    "proteins_without_modeled_functions = set()\n",
    "# identify them by checking if all values within values are empty\n",
    "for gene,prots in protein_functions.items():\n",
    "\tfor prot,funcs in prots.items():\n",
    "\t\t# if all values are empty, add to set\n",
    "\t\tif not any(funcs.values()):\n",
    "\t\t\tproteins_without_modeled_functions.add(prot)\n",
    "with open('./proteins_without_modeled_functions.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(sorted(list(proteins_without_modeled_functions))))\n",
    "with open(directory + 'rxns_enz.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(rxns_enz) + ['/']))\n",
    "with open(directory + 'rxns_pro_and_enz.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(rxns_proenz) + ['/']))\n",
    "with open(directory + 'sij_pro_and_enz.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sij_proenz + ['/']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build GAMS RBA constraints (originally A03_build_GAMS_RBA_constraints.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### LOAD INPUTS AND PARAMETERS\n",
    "copy_to_gams = True # if True, copy the model to the gams folder\n",
    "\n",
    "# Stoichiometry\n",
    "df_stoich = read_spreadsheet('./model/RBA_stoichiometry.tsv')\n",
    "df_stoich.index = df_stoich.id.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Write prosyn reaction\n",
    "idx = [i for i in df_stoich.index if i[:7] == 'PROSYN-']\n",
    "prosyn = [\"'\" + i + \"'\" for i in idx]\n",
    "prosyn = ['/'] + prosyn + ['/']\n",
    "with open('./model/RBA_rxns_prosyn.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(prosyn))\n",
    "## write ribosyn set\n",
    "idx = [i for i in df_stoich.index if i[:8] == 'RIBOSYN-']\n",
    "ribosyn = [\"'\" + i + \"'\" for i in idx]\n",
    "ribosyn = ['/'] + ribosyn + ['/']\n",
    "with open('./model/RBA_rxns_ribosyn.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(ribosyn))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make fwd and reversible rxn list for metabolic network and all rxns (not needed but helpful for testing)\n",
    "# Also, make flux coupling analysis file in case you use that for testing\n",
    "\n",
    "# Using Patrick's FCA code\n",
    "rxn_types = {'irrev': 0, 'reversible-fwd-half': 1, 'reversible-rev-half': 2, 'pseudoreaction': 3, 'exchange-fwd-half': 4, 'exchange-rev-half': 5}\n",
    "\n",
    "rev_rxn_list = []; rev_list = []\n",
    "fwd_rxn_list = []; fwd_list = []\n",
    "fca_list = []\n",
    "\n",
    "for i in df_stoich.id:\n",
    "\ttag,rxn_base_id,rxn_dir,enz_id = extract_details_from_rxnid(i)\n",
    "\t#print(\"'\"+i+\"' \"+str(rxn_types['irrev']))\n",
    "\t#print(rxn_base_id)\n",
    "\t#print(i,'\\t',tag,'\\t',rxn_dir)\n",
    "\tif rxn_dir == 'FWD':\n",
    "\t\tfwd_list.append(i)\n",
    "\t\tif tag == 'RXN':\n",
    "\t\t\tfwd_rxn_list.append(i)\n",
    "\t\tif rxn_base_id[:3] == 'EX_':\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['exchange-fwd-half']))\n",
    "\t\t# Add as irreversible if no counterpart found in other direction\n",
    "\t\telif i.replace(\"_FWD-\",\"_REV-\") not in df_stoich.id:\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['irrev'])) \n",
    "\t\telse:\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['reversible-fwd-half']))\n",
    "\telif rxn_dir == 'REV':\n",
    "\t\trev_list.append(i)\n",
    "\t\tif tag == 'RXN':\n",
    "\t\t\trev_rxn_list.append(i)\n",
    "\t\tif rxn_base_id[:3] == 'EX_':\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['exchange-rev-half']))\n",
    "\t\t# Add as irreversible if no counterpart found in other direction\n",
    "\t\telif i.replace(\"_REV-\",\"_FWD-\") not in df_stoich.id:\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['irrev'])) \n",
    "\t\telse:\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['reversible-rev-half']))\n",
    "\telse:\n",
    "\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['irrev'])) \n",
    "\n",
    "fwd_rxn_list = [\"'\" + i + \"'\" for i in fwd_rxn_list if i != '/']\n",
    "fwd_rxn_list = ['/'] + fwd_rxn_list + ['/']\n",
    "with open('./model/RBA_rxns_rxnmetabolicnetworkFWD.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(fwd_rxn_list))\n",
    "rev_rxn_list = [\"'\" + i + \"'\" for i in rev_rxn_list if i != '/']\n",
    "rev_rxn_list = ['/'] + rev_rxn_list + ['/']\n",
    "with open('./model/RBA_rxns_rxnmetabolicnetworkREV.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rev_rxn_list))\n",
    "\n",
    "fwd_list = [\"'\" + i + \"'\" for i in fwd_list if i != '/']\n",
    "fwd_list = ['/'] + fwd_list + ['/']\n",
    "with open('./model/RBA_rxns_FWD.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(fwd_list))\n",
    "rev_list = [\"'\" + i + \"'\" for i in rev_list if i != '/']\n",
    "rev_list = ['/'] + rev_list + ['/']\n",
    "with open('./model/RBA_rxns_REV.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rev_list))\n",
    "\n",
    "fca_list = ['/'] + fca_list + ['/']\n",
    "with open('./model/rxntype.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(fca_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating set pairing SM values w/ RBA ones, for use in converting flux constraints from SM format to RBA\n",
    "\n",
    "# filter out rxns with no FBA_name\n",
    "df_sm = df_stoich[df_stoich.FBA_name.notnull()]\n",
    "# get SM rxn IDs from FBA_name column\n",
    "sm_rxn_ids = df_sm.FBA_name.to_list()\n",
    "# get RBA rxn IDs from index\n",
    "rba_rxn_ids = df_sm.index.to_list()\n",
    "rba_rxn_dirs = [extract_details_from_rxnid(i)[2] for i in rba_rxn_ids]\n",
    "# convert dir to -1 if rev, 1 otherwise\n",
    "rba_rxn_dirs = [-1 if i == 'REV' else 1 for i in rba_rxn_dirs]\n",
    "# create list of strings pairing SM rxn IDs with RBA rxn IDs\n",
    "sm_rba_rxn_pairs_dict = {rba_rxn_ids[i]:sm_rxn_ids[i] for i in range(0, len(sm_rxn_ids))}\n",
    "# save as JSON file\n",
    "with open('./model/SM_RBA_rxn_pairs.json', 'w') as f:\n",
    "\tjson.dump(sm_rba_rxn_pairs_dict, f)\n",
    "sm_rba_rxn_pairs = [\"'\"+sm_rxn_ids[i]+\"'.'\"+rba_rxn_ids[i]+\"'\"+' '+str(int(rba_rxn_dirs[i])) for i in range(0, len(sm_rxn_ids))]\n",
    "# convert SM rxn bounds to RBA format; for each sm rxn, get the corresponding rba rxn and its direction\n",
    "with open('./model/SM_rxn_bounds.txt', 'w') as f:\n",
    "\tdelim = ' '\n",
    "\tsm_rxns_checked = []\n",
    "\t# find longest rxn ID for formatting purposes\n",
    "\tmax_sm_rxn_id_len = max([len(i)+2 for i in sm_rxn_ids])+1\n",
    "\tmax_lb_len = max([len(str(rxn.lower_bound)) for rxn in model.reactions if rxn.id in sm_rxn_ids])+1\n",
    "\tmax_ub_len = max([len(str(rxn.upper_bound)) for rxn in model.reactions if rxn.id in sm_rxn_ids])\n",
    "\t# write headers to file, with ljust\n",
    "\tf.write(f\"{'':>{max_sm_rxn_id_len}}{'lo':>{max_lb_len}}{' up':<{max_ub_len}}\")\n",
    "\tfor rba_rxn,sm_rxn in sm_rba_rxn_pairs_dict.items():\n",
    "\t\t# find bounds of sm rxn in model\n",
    "\t\trxn = model.reactions.get_by_id(sm_rxn)\n",
    "\t\tub = rxn.upper_bound\n",
    "\t\tlb = rxn.lower_bound\n",
    "\n",
    "\t\tif sm_rxn not in sm_rxns_checked:\n",
    "\t\t\tsm_rxns_checked.append(sm_rxn)\n",
    "\t\t\tname=f\"'{sm_rxn}'\"\n",
    "\t\t\t# make string with SM rxn ID, lb, and ub, using max lengths for formatting\n",
    "\t\t\tf.write(f\"\\n{name:>{max_sm_rxn_id_len}}{str(lb):>{max_lb_len}} {str(ub):<{max_ub_len}}\")\n",
    "# write to file\n",
    "with open('./model/SM_RBA_rxn_pairs.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(sm_rba_rxn_pairs) + ['/']))\n",
    "# make file w/ all FBA IDs\n",
    "with open('./model/SM_rxn_ids.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted([\"'\"+i+\"'\" for i in set(sm_rxn_ids)]) + ['/']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Applications/anaconda3/envs/rba/lib/python3.9/site-packages/cobra/core/metabolite.py:191: UserWarning: The element 'R' does not appear in the periodic table\n",
      "  warn(f\"The element {e} does not appear in the periodic table\")\n",
      "/Applications/anaconda3/envs/rba/lib/python3.9/site-packages/cobra/core/metabolite.py:191: UserWarning: The element 'Tr' does not appear in the periodic table\n",
      "  warn(f\"The element {e} does not appear in the periodic table\")\n",
      "/Applications/anaconda3/envs/rba/lib/python3.9/site-packages/cobra/core/metabolite.py:191: UserWarning: The element 'Fd' does not appear in the periodic table\n",
      "  warn(f\"The element {e} does not appear in the periodic table\")\n",
      "/Applications/anaconda3/envs/rba/lib/python3.9/site-packages/cobra/core/metabolite.py:191: UserWarning: The element 'M' does not appear in the periodic table\n",
      "  warn(f\"The element {e} does not appear in the periodic table\")\n"
     ]
    }
   ],
   "source": [
    "# make file for all formula weights (some may be inaccurate due to R groups and other non-standard groups)\n",
    "with open('./model/formula_weights.txt','w') as f:\n",
    "\tf.write('/\\n')\n",
    "\tfor met in model.metabolites:\n",
    "\t\t# write met name, formula, and formula weight\n",
    "\t\ttry:\n",
    "\t\t\tformula = str(met.formula)\n",
    "\t\t\tmw = str(met.formula_weight)\n",
    "\t\texcept:\n",
    "\t\t\tmw = 'unknown'\n",
    "\t\t\tformula = 'unknown'\n",
    "\t\tf.write(str(met.id) + '\\t' + formula + '\\t' + mw + '\\n')\n",
    "\tf.write('/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for AA in df_aamap, find all mets in met_list that match its name (without the location)\n",
    "with open('./model/aa.txt', 'w') as f, open('./model/tRNAs.txt', 'w') as f2:\n",
    "\tfis = [f,f2]\n",
    "\tfor fi in fis: fi.write('/')\n",
    "\tfor aa in df_aamap.index:\n",
    "\t\taa_mets = [f\"'{met}'\" for met in met_list if met.rsplit('_',1)[0] == df_aamap.tRNA_in[aa].rsplit('_',1)[0]]\n",
    "\t\tf.write(f\"\\n'{aa}'\")\n",
    "\t\tif aa_mets:\n",
    "\t\t\tf2.write(f\"\\n'{aa}'.({','.join(aa_mets)})\")\n",
    "\tfor fi in fis: fi.write('\\n/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write combinations of gene_src and protein IDs in form 'gene_src'.('protein1','protein2',...)\n",
    "with open('./model/proteins_and_locations.txt', 'w') as f, open('./model/proteins.txt', 'w') as f2:\n",
    "\tfis = [f,f2]\n",
    "\tfor fi in fis: fi.write('/')\n",
    "\tfor gene,prots in protein_functions.items():\n",
    "\t\tprots_str = ','.join([f\"'PROSYN-{prot}'\" for prot in prots])\n",
    "\t\tf.write(f\"\\n'{gene}'.({prots_str})\")\n",
    "\t\tf2.write(f\"\\n'{gene}'\")\n",
    "\tfor fi in fis: fi.write('\\n/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "if copy_to_gams: # automatically copies the model folder to the GAMS folder (removing the previous version), if you choose to\n",
    "\timport shutil\n",
    "\t# copy \"model\" folder to GAMS folder\n",
    "\t# Source folder path\n",
    "\tsource_folder = './model'\n",
    "\n",
    "\t# Destination folder path\n",
    "\tdestination_folder = '../GAMS/model'\n",
    "\n",
    "\t# Remove the existing destination folder\n",
    "\tshutil.rmtree(destination_folder)\n",
    "\n",
    "\t# Copy the folder\n",
    "\tshutil.copytree(source_folder, destination_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Enzyme PDHCPLX has cofactor thmpp_c counted multiple times, including subunits ['rt5580'] and rt5262.\n",
      "Warning: Enzyme MET1213 has cofactor fad_c counted multiple times, including subunits ['rt0876'] and rt7115.\n",
      "Enzyme rt2671 has cofactors {'fad_m'} in its reactions.\n",
      "  - G3PDf_m has cofactor fad_m\n",
      "Warning: Enzyme FECOORCPLX3 has cofactor cu2_m counted multiple times, including subunits ['rtmCOX1'] and rtmCOX2.\n",
      "Warning: Enzyme FECOORCPLX4 has cofactor cu2_m counted multiple times, including subunits ['rtmCOX1'] and rtmCOX2.\n",
      "Enzyme rt2040 has cofactors {'fad_x'} in its reactions.\n",
      "  - ACOAD162f_x has cofactor fad_x\n",
      "  - ACOAD100f_x has cofactor fad_x\n",
      "  - ACOAD143f_x has cofactor fad_x\n",
      "  - ACOAD121af_x has cofactor fad_x\n",
      "  - ACOAD182f_x has cofactor fad_x\n",
      "  - ACOAD142f_x has cofactor fad_x\n",
      "  - ACOAD101df_x has cofactor fad_x\n",
      "  - ACOAD82ef_x has cofactor fad_x\n",
      "  - ACOAD140f_x has cofactor fad_x\n",
      "  - ACOAD181af_x has cofactor fad_x\n",
      "  - ACOAD60f_x has cofactor fad_x\n",
      "  - ACOAD160f_x has cofactor fad_x\n",
      "  - ACOAD161bf_x has cofactor fad_x\n",
      "  - ACOAD120f_x has cofactor fad_x\n",
      "  - ACOAD40f_x has cofactor fad_x\n",
      "  - ACOAD102ef_x has cofactor fad_x\n",
      "  - ACOAD161af_x has cofactor fad_x\n",
      "  - ACOAD80f_x has cofactor fad_x\n",
      "  - ACOAD141af_x has cofactor fad_x\n",
      "  - ACOAD141bf_x has cofactor fad_x\n",
      "  - ACOAD180f_x has cofactor fad_x\n",
      "  - ACOAD183f_x has cofactor fad_x\n",
      "  - ACOAD163f_x has cofactor fad_x\n",
      "Enzyme rt1644 has cofactors {'fad_m'} in its reactions.\n",
      "  - ACOADm50f_m has cofactor fad_m\n",
      "Enzyme rt3621 has cofactors {'fad_m'} in its reactions.\n",
      "  - GLUTCOADH_m has cofactor fad_m\n",
      "Enzyme rt6234 has cofactors {'fad_m'} in its reactions.\n",
      "  - PROD2_m has cofactor fad_m\n",
      "Enzyme rt5702 has cofactors {'fad_m'} in its reactions.\n",
      "  - ACOADm40f_m has cofactor fad_m\n",
      "  - ACOAD40f_m has cofactor fad_m\n",
      "  - ACOAD60f_m has cofactor fad_m\n",
      "  - ACOADm30f_m has cofactor fad_m\n",
      "Warning: Enzyme FECOORCPLX2 has cofactor cu2_m counted multiple times, including subunits ['rtmCOX1'] and rtmCOX2.\n",
      "Enzyme rt4202 has cofactors {'fad_m'} in its reactions.\n",
      "  - ACOAD120f_m has cofactor fad_m\n",
      "  - ACOAD100f_m has cofactor fad_m\n",
      "  - ACOAD80f_m has cofactor fad_m\n",
      "Warning: Enzyme FECOORCPLX1 has cofactor cu2_m counted multiple times, including subunits ['rtmCOX1'] and rtmCOX2.\n"
     ]
    }
   ],
   "source": [
    "# check if any enzymes likely have cofactors counted multiple times\n",
    "for enz in set(df_enz['enz']):\n",
    "\t# find 1st row where the enz matches\n",
    "\tprot_stoich_str = df_enz[df_enz['enz'] == enz].iloc[0]['protein_stoich']\n",
    "\tif pd.isna(prot_stoich_str):\n",
    "\t\tcontinue\n",
    "\t# parse the protein_stoich string into a dictionary\n",
    "\t# print(enz, prot_stoich_str)\n",
    "\tprot_stoich = {item.split(':')[0]: float(item.split(':')[1]) for item in prot_stoich_str.split(',') if ':' in item and not pd.isna(item)}\n",
    "\tif not prot_stoich:\n",
    "\t\tcontinue\n",
    "\trxns = set(df_enz[df_enz['enz'] == enz]['rxn_src'].to_list())\n",
    "\t# remove blocked reactions\n",
    "\trxns = [rxn for rxn in rxns if model.reactions.get_by_id(rxn).bounds != (0, 0)]\n",
    "\t# find the trans_stoich_edits and ptm_stoich_edits for all proteins comprising it\n",
    "\tpossible_cofactors = dict()\n",
    "\tfor pro in prot_stoich.keys():\n",
    "\t\tif pro in df_pro.index:\n",
    "\t\t\tmet_dict = dict()\n",
    "\t\t\t# Collect edits, filter out NaN, and convert to string\n",
    "\t\t\tedits = [df_pro.loc[pro,'trans_stoich_edits'], df_pro.loc[pro,'PTM_stoich_edits']]\n",
    "\t\t\tedits = [str(e) for e in edits if pd.notna(e)]\n",
    "\t\t\tif not edits:\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tmet_items = set(','.join(edits).split(','))\n",
    "\t\t\tmet_dict = {i.split(':')[0]: float(i.split(':')[1]) for i in met_items if not pd.isna(i) and ':' in i}\n",
    "\t\t\tfor k,v in met_dict.items():\n",
    "\t\t\t\tif v > 0: # i.e., if it's consumed\n",
    "\t\t\t\t\tif k in possible_cofactors:\n",
    "\t\t\t\t\t\tprint(f\"Warning: Enzyme {enz} has cofactor {k} counted multiple times, including subunits {possible_cofactors[k]} and {pro}.\")\n",
    "\t\t\t\t\tpossible_cofactors.setdefault(k, []).append(pro)\n",
    "\tif possible_cofactors:\n",
    "\t\t# print(enz,possible_cofactors)\n",
    "\t\t# find the reactions that have the same cofactors\n",
    "\t\tcofactors_in_rxns = set()\n",
    "\t\tfor rxn in rxns:\n",
    "\t\t\trxn_cofactors = set()\n",
    "\t\t\t# get the stoichiometry of the reaction\n",
    "\t\t\trxn_stoich_dict = model.reactions.get_by_id(rxn).metabolites\n",
    "\t\t\tif rxn_stoich_dict is None:\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tfor met, coef in rxn_stoich_dict.items():\n",
    "\t\t\t\t# if coef < 0 and met.id in [i for i in possible_cofactors.keys()]:\n",
    "\t\t\t\tif coef < 0 and met.id.rsplit('_',1)[0] in [i.rsplit('_',1)[0] for i in possible_cofactors.keys()]:\n",
    "\t\t\t\t\trxn_cofactors.add(met.id)\n",
    "\t\t\t\t\tcofactors_in_rxns.add(met.id)\n",
    "\t\tif cofactors_in_rxns:\n",
    "\t\t\tprint(f\"Enzyme {enz} has cofactors {cofactors_in_rxns} in its reactions.\")\n",
    "\t\t\t# print the reactions that have the same cofactors\n",
    "\t\t\tfor rxn in rxns:\n",
    "\t\t\t\tfor cofactor in cofactors_in_rxns:\n",
    "\t\t\t\t\tif cofactor in [met.id for met in model.reactions.get_by_id(rxn).metabolites.keys()]:\n",
    "\t\t\t\t\t\tprint(f\"  - {rxn} has cofactor {cofactor}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rba",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
